// effect/dist/esm/TSubscriptionRef.js
import * as internal from "./internal/stm/tSubscriptionRef.js";
/**
 * @since 3.10.0
 * @category symbols
 */
export const TSubscriptionRefTypeId = internal.TSubscriptionRefTypeId;
/**
 * @since 3.10.0
 * @category mutations
 */
export const get = internal.get;
/**
 * @since 3.10.0
 * @category mutations
 */
export const getAndSet = internal.getAndSet;
/**
 * @since 3.10.0
 * @category mutations
 */
export const getAndUpdate = internal.getAndUpdate;
/**
 * @since 3.10.0
 * @category mutations
 */
export const getAndUpdateSome = internal.getAndUpdateSome;
/**
 * @since 3.10.0
 * @category constructors
 */
export const make = internal.make;
/**
 * @since 3.10.0
 * @category mutations
 */
export const modify = internal.modify;
/**
 * @since 3.10.0
 * @category mutations
 */
export const modifySome = internal.modifySome;
/**
 * @since 3.10.0
 * @category mutations
 */
export const set = internal.set;
/**
 * @since 3.10.0
 * @category mutations
 */
export const setAndGet = internal.setAndGet;
/**
 * @since 3.10.0
 * @category mutations
 */
export const update = internal.update;
/**
 * @since 3.10.0
 * @category mutations
 */
export const updateAndGet = internal.updateAndGet;
/**
 * @since 3.10.0
 * @category mutations
 */
export const updateSome = internal.updateSome;
/**
 * @since 3.10.0
 * @category mutations
 */
export const updateSomeAndGet = internal.updateSomeAndGet;
/**
 * @since 3.10.0
 * @category mutations
 */
export const changesScoped = internal.changesScoped;
/**
 * @since 3.10.0
 * @category mutations
 */
export const changesStream = internal.changesStream;
/**
 * @since 3.10.0
 * @category mutations
 */
export const changes = self => self.changes;
