// effect/dist/cjs/MergeDecision.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.match = exports.isMergeDecision = exports.MergeDecisionTypeId = exports.Done = exports.AwaitConst = exports.Await = void 0;
var internal = _interopRequireWildcard(require("./internal/channel/mergeDecision.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 2.0.0
 * @category symbols
 */
const MergeDecisionTypeId = exports.MergeDecisionTypeId = internal.MergeDecisionTypeId;
/**
 * @since 2.0.0
 * @category constructors
 */
const Done = exports.Done = internal.Done;
/**
 * @since 2.0.0
 * @category constructors
 */
const Await = exports.Await = internal.Await;
/**
 * @since 2.0.0
 * @category constructors
 */
const AwaitConst = exports.AwaitConst = internal.AwaitConst;
/**
 * Returns `true` if the specified value is a `MergeDecision`, `false`
 * otherwise.
 *
 * @since 2.0.0
 * @category refinements
 */
const isMergeDecision = exports.isMergeDecision = internal.isMergeDecision;
/**
 * @since 2.0.0
 * @category folding
 */
const match = exports.match = internal.match;
