// effect/dist/cjs/MetricState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.summary = exports.isSummaryState = exports.isMetricState = exports.isHistogramState = exports.isGaugeState = exports.isFrequencyState = exports.isCounterState = exports.histogram = exports.gauge = exports.frequency = exports.counter = exports.SummaryStateTypeId = exports.MetricStateTypeId = exports.HistogramStateTypeId = exports.GaugeStateTypeId = exports.FrequencyStateTypeId = exports.CounterStateTypeId = void 0;
var internal = _interopRequireWildcard(require("./internal/metric/state.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 2.0.0
 * @category symbols
 */
const MetricStateTypeId = exports.MetricStateTypeId = internal.MetricStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const CounterStateTypeId = exports.CounterStateTypeId = internal.CounterStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const FrequencyStateTypeId = exports.FrequencyStateTypeId = internal.FrequencyStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const GaugeStateTypeId = exports.GaugeStateTypeId = internal.GaugeStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const HistogramStateTypeId = exports.HistogramStateTypeId = internal.HistogramStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const SummaryStateTypeId = exports.SummaryStateTypeId = internal.SummaryStateTypeId;
/**
 * @since 2.0.0
 * @category constructors
 */
const counter = exports.counter = internal.counter;
/**
 * @since 2.0.0
 * @category constructors
 */
const frequency = exports.frequency = internal.frequency;
/**
 * @since 2.0.0
 * @category constructors
 */
const gauge = exports.gauge = internal.gauge;
/**
 * @since 2.0.0
 * @category constructors
 */
const histogram = exports.histogram = internal.histogram;
/**
 * @since 2.0.0
 * @category constructors
 */
const summary = exports.summary = internal.summary;
/**
 * @since 2.0.0
 * @category refinements
 */
const isMetricState = exports.isMetricState = internal.isMetricState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isCounterState = exports.isCounterState = internal.isCounterState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isFrequencyState = exports.isFrequencyState = internal.isFrequencyState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isGaugeState = exports.isGaugeState = internal.isGaugeState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isHistogramState = exports.isHistogramState = internal.isHistogramState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isSummaryState = exports.isSummaryState = internal.isSummaryState;
