// effect/dist/cjs/Secret.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.value = exports.unsafeWipe = exports.make = exports.isSecret = exports.fromString = exports.fromIterable = exports.SecretTypeId = void 0;
var InternalSecret = _interopRequireWildcard(require("./internal/secret.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 2.0.0
 * @category symbols
 * @deprecated
 */
const SecretTypeId = exports.SecretTypeId = InternalSecret.SecretTypeId;
/**
 * @since 2.0.0
 * @category refinements
 * @deprecated
 */
const isSecret = exports.isSecret = InternalSecret.isSecret;
/**
 * @since 2.0.0
 * @category constructors
 * @deprecated
 */
const make = exports.make = InternalSecret.make;
/**
 * @since 2.0.0
 * @category constructors
 * @deprecated
 */
const fromIterable = exports.fromIterable = InternalSecret.fromIterable;
/**
 * @since 2.0.0
 * @category constructors
 * @deprecated
 */
const fromString = exports.fromString = InternalSecret.fromString;
/**
 * @since 2.0.0
 * @category getters
 * @deprecated
 */
const value = exports.value = InternalSecret.value;
/**
 * @since 2.0.0
 * @category unsafe
 * @deprecated
 */
const unsafeWipe = exports.unsafeWipe = InternalSecret.unsafeWipe;
