// @lucide/svelte/dist/aliases/aliases.js
/**
 * @license @lucide/svelte v0.534.0 - ISC
 *
 * ISC License
 * 
 * Copyright (c) for portions of Lucide are held by Cole Bemis 2013-2022 as part of Feather (MIT). All other copyright (c) for Lucide are held by Lucide Contributors 2022.
 * 
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 * 
 */
// AlarmClockCheck aliases
export { default as AlarmCheck } from '../icons/alarm-check.js';
// AlarmClockMinus aliases
export { default as AlarmMinus } from '../icons/alarm-minus.js';
// AlarmClockPlus aliases
export { default as AlarmPlus } from '../icons/alarm-plus.js';
// ArrowDownAZ aliases
export { default as ArrowDownAz } from '../icons/arrow-down-az.js';
// ArrowDownZA aliases
export { default as ArrowDownZa } from '../icons/arrow-down-za.js';
// ArrowDownWideNarrow aliases
export { default as SortDesc } from '../icons/sort-desc.js';
// ArrowUpAZ aliases
export { default as ArrowUpAz } from '../icons/arrow-up-az.js';
// ArrowUpNarrowWide aliases
export { default as SortAsc } from '../icons/sort-asc.js';
// ArrowUpZA aliases
export { default as ArrowUpZa } from '../icons/arrow-up-za.js';
// Axis3d aliases
export { default as Axis3D } from '../icons/axis-3-d.js';
// BadgeCheck aliases
export { default as Verified } from '../icons/verified.js';
// BadgeQuestionMark aliases
export { default as BadgeHelp } from '../icons/badge-help.js';
// BetweenHorizontalStart aliases
export { 
/** @deprecated Renamed because of typo, use {@link BetweenHorizontalStart} instead. This alias will be removed in v1.0 */
default as BetweenHorizonalStart } from '../icons/between-horizonal-start.js';
// BetweenHorizontalEnd aliases
export { 
/** @deprecated Renamed because of typo, use {@link BetweenHorizontalEnd} instead. This alias will be removed in v1.0 */
default as BetweenHorizonalEnd } from '../icons/between-horizonal-end.js';
// BookDashed aliases
export { default as BookTemplate } from '../icons/book-template.js';
// Braces aliases
export { default as CurlyBraces } from '../icons/curly-braces.js';
// Captions aliases
export { default as Subtitles } from '../icons/subtitles.js';
// ChartArea aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartArea} instead. This alias will be removed in v1.0 */
default as AreaChart } from '../icons/area-chart.js';
// ChartBarBig aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartBarBig} instead. This alias will be removed in v1.0 */
default as BarChartHorizontalBig } from '../icons/bar-chart-horizontal-big.js';
// ChartBar aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartBar} instead. This alias will be removed in v1.0 */
default as BarChartHorizontal } from '../icons/bar-chart-horizontal.js';
// ChartCandlestick aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartCandlestick} instead. This alias will be removed in v1.0 */
default as CandlestickChart } from '../icons/candlestick-chart.js';
// ChartColumnBig aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartColumnBig} instead. This alias will be removed in v1.0 */
default as BarChartBig } from '../icons/bar-chart-big.js';
// ChartColumnIncreasing aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartColumnIncreasing} instead. This alias will be removed in v1.0 */
default as BarChart4 } from '../icons/bar-chart-4.js';
// ChartColumn aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartColumn} instead. This alias will be removed in v1.0 */
default as BarChart3 } from '../icons/bar-chart-3.js';
// ChartLine aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartLine} instead. This alias will be removed in v1.0 */
default as LineChart } from '../icons/line-chart.js';
// ChartNoAxesColumnIncreasing aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartNoAxesColumnIncreasing} instead. This alias will be removed in v1.0 */
default as BarChart } from '../icons/bar-chart.js';
// ChartNoAxesColumn aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartNoAxesColumn} instead. This alias will be removed in v1.0 */
default as BarChart2 } from '../icons/bar-chart-2.js';
// ChartNoAxesGantt aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartNoAxesGantt} instead. This alias will be removed in v1.0 */
default as GanttChart } from '../icons/gantt-chart.js';
// ChartPie aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartPie} instead. This alias will be removed in v1.0 */
default as PieChart } from '../icons/pie-chart.js';
// ChartScatter aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartScatter} instead. This alias will be removed in v1.0 */
default as ScatterChart } from '../icons/scatter-chart.js';
// CircleAlert aliases
export { default as AlertCircle } from '../icons/alert-circle.js';
// CircleArrowDown aliases
export { default as ArrowDownCircle } from '../icons/arrow-down-circle.js';
// CircleArrowLeft aliases
export { default as ArrowLeftCircle } from '../icons/arrow-left-circle.js';
// CircleArrowOutDownLeft aliases
export { default as ArrowDownLeftFromCircle } from '../icons/arrow-down-left-from-circle.js';
// CircleArrowOutDownRight aliases
export { default as ArrowDownRightFromCircle } from '../icons/arrow-down-right-from-circle.js';
// CircleArrowOutUpLeft aliases
export { default as ArrowUpLeftFromCircle } from '../icons/arrow-up-left-from-circle.js';
// CircleArrowOutUpRight aliases
export { default as ArrowUpRightFromCircle } from '../icons/arrow-up-right-from-circle.js';
// CircleArrowRight aliases
export { default as ArrowRightCircle } from '../icons/arrow-right-circle.js';
// CircleArrowUp aliases
export { default as ArrowUpCircle } from '../icons/arrow-up-circle.js';
// CircleCheckBig aliases
export { default as CheckCircle } from '../icons/check-circle.js';
// CircleCheck aliases
export { default as CheckCircle2 } from '../icons/check-circle-2.js';
// CircleChevronDown aliases
export { default as ChevronDownCircle } from '../icons/chevron-down-circle.js';
// CircleChevronLeft aliases
export { default as ChevronLeftCircle } from '../icons/chevron-left-circle.js';
// CircleChevronRight aliases
export { default as ChevronRightCircle } from '../icons/chevron-right-circle.js';
// CircleChevronUp aliases
export { default as ChevronUpCircle } from '../icons/chevron-up-circle.js';
// CircleDivide aliases
export { default as DivideCircle } from '../icons/divide-circle.js';
// CircleGauge aliases
export { default as GaugeCircle } from '../icons/gauge-circle.js';
// CircleMinus aliases
export { default as MinusCircle } from '../icons/minus-circle.js';
// CircleParkingOff aliases
export { default as ParkingCircleOff } from '../icons/parking-circle-off.js';
// CircleParking aliases
export { default as ParkingCircle } from '../icons/parking-circle.js';
// CirclePause aliases
export { default as PauseCircle } from '../icons/pause-circle.js';
// CirclePercent aliases
export { default as PercentCircle } from '../icons/percent-circle.js';
// CirclePlay aliases
export { default as PlayCircle } from '../icons/play-circle.js';
// CirclePlus aliases
export { default as PlusCircle } from '../icons/plus-circle.js';
// CirclePower aliases
export { default as PowerCircle } from '../icons/power-circle.js';
// CircleQuestionMark aliases
export { default as HelpCircle } from '../icons/help-circle.js';
// CircleQuestionMark aliases
export { default as CircleHelp } from '../icons/circle-help.js';
// CircleSlash2 aliases
export { default as CircleSlashed } from '../icons/circle-slashed.js';
// CircleStop aliases
export { default as StopCircle } from '../icons/stop-circle.js';
// CircleUserRound aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link CircleUserRound} instead. This alias will be removed in v1.0 */
default as UserCircle2 } from '../icons/user-circle-2.js';
// CircleUser aliases
export { default as UserCircle } from '../icons/user-circle.js';
// CircleX aliases
export { default as XCircle } from '../icons/x-circle.js';
// ClipboardPenLine aliases
export { default as ClipboardSignature } from '../icons/clipboard-signature.js';
// ClipboardPen aliases
export { default as ClipboardEdit } from '../icons/clipboard-edit.js';
// CloudDownload aliases
export { default as DownloadCloud } from '../icons/download-cloud.js';
// CloudUpload aliases
export { default as UploadCloud } from '../icons/upload-cloud.js';
// CodeXml aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link CodeXml} instead. This alias will be removed in v1.0 */
default as Code2 } from '../icons/code-2.js';
// Columns2 aliases
export { default as Columns } from '../icons/columns.js';
// Columns3Cog aliases
export { default as ColumnsSettings } from '../icons/columns-settings.js';
// Columns3Cog aliases
export { default as TableConfig } from '../icons/table-config.js';
// Columns3 aliases
export { default as PanelsLeftRight } from '../icons/panels-left-right.js';
// ContactRound aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ContactRound} instead. This alias will be removed in v1.0 */
default as Contact2 } from '../icons/contact-2.js';
// DiamondPercent aliases
export { default as PercentDiamond } from '../icons/percent-diamond.js';
// Earth aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link Earth} instead. This alias will be removed in v1.0 */
default as Globe2 } from '../icons/globe-2.js';
// EllipsisVertical aliases
export { default as MoreVertical } from '../icons/more-vertical.js';
// Ellipsis aliases
export { default as MoreHorizontal } from '../icons/more-horizontal.js';
// FileAxis3d aliases
export { default as FileAxis3D } from '../icons/file-axis-3-d.js';
// FileChartColumnIncreasing aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileChartColumnIncreasing} instead. This alias will be removed in v1.0 */
default as FileBarChart } from '../icons/file-bar-chart.js';
// FileChartColumn aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileChartColumn} instead. This alias will be removed in v1.0 */
default as FileBarChart2 } from '../icons/file-bar-chart-2.js';
// FileChartLine aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileChartLine} instead. This alias will be removed in v1.0 */
default as FileLineChart } from '../icons/file-line-chart.js';
// FileChartPie aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileChartPie} instead. This alias will be removed in v1.0 */
default as FilePieChart } from '../icons/file-pie-chart.js';
// FileCog aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileCog} instead. This alias will be removed in v1.0 */
default as FileCog2 } from '../icons/file-cog-2.js';
// FilePenLine aliases
export { default as FileSignature } from '../icons/file-signature.js';
// FilePlay aliases
export { default as FileVideo } from '../icons/file-video.js';
// FilePen aliases
export { default as FileEdit } from '../icons/file-edit.js';
// FileQuestionMark aliases
export { default as FileQuestion } from '../icons/file-question.js';
// FileVideoCamera aliases
export { default as FileVideo2 } from '../icons/file-video-2.js';
// FolderCog aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FolderCog} instead. This alias will be removed in v1.0 */
default as FolderCog2 } from '../icons/folder-cog-2.js';
// FolderPen aliases
export { default as FolderEdit } from '../icons/folder-edit.js';
// FunnelX aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FunnelX} instead. This alias will be removed in v1.0 */
default as FilterX } from '../icons/filter-x.js';
// Funnel aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link Funnel} instead. This alias will be removed in v1.0 */
default as Filter } from '../icons/filter.js';
// GitCommitHorizontal aliases
export { default as GitCommit } from '../icons/git-commit.js';
// Grid2x2Check aliases
export { default as Grid2X2Check } from '../icons/grid-2-x-2-check.js';
// Grid2x2Plus aliases
export { default as Grid2X2Plus } from '../icons/grid-2-x-2-plus.js';
// Grid2x2X aliases
export { default as Grid2X2X } from '../icons/grid-2-x-2-x.js';
// Grid2x2 aliases
export { default as Grid2X2 } from '../icons/grid-2-x-2.js';
// Grid3x3 aliases
export { default as Grid } from '../icons/grid.js';
// Grid3x3 aliases
export { default as Grid3X3 } from '../icons/grid-3-x-3.js';
// HandHelping aliases
export { default as HelpingHand } from '../icons/helping-hand.js';
// House aliases
export { default as Home } from '../icons/home.js';
// IceCreamBowl aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link IceCreamBowl} instead. This alias will be removed in v1.0 */
default as IceCream2 } from '../icons/ice-cream-2.js';
// IceCreamCone aliases
export { default as IceCream } from '../icons/ice-cream.js';
// IndentDecrease aliases
export { default as Outdent } from '../icons/outdent.js';
// IndentIncrease aliases
export { default as Indent } from '../icons/indent.js';
// LaptopMinimal aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link LaptopMinimal} instead. This alias will be removed in v1.0 */
default as Laptop2 } from '../icons/laptop-2.js';
// Layers aliases
export { default as Layers3 } from '../icons/layers-3.js';
// LoaderCircle aliases
export { default as Loader2 } from '../icons/loader-2.js';
// LockKeyholeOpen aliases
export { default as UnlockKeyhole } from '../icons/unlock-keyhole.js';
// LockOpen aliases
export { default as Unlock } from '../icons/unlock.js';
// MailQuestionMark aliases
export { default as MailQuestion } from '../icons/mail-question.js';
// MapPinPen aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link MapPinPen} instead. This alias will be removed in v1.0 */
default as LocationEdit } from '../icons/location-edit.js';
// MessageCircleQuestionMark aliases
export { default as MessageCircleQuestion } from '../icons/message-circle-question.js';
// MicVocal aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link MicVocal} instead. This alias will be removed in v1.0 */
default as Mic2 } from '../icons/mic-2.js';
// Move3d aliases
export { default as Move3D } from '../icons/move-3-d.js';
// OctagonAlert aliases
export { default as AlertOctagon } from '../icons/alert-octagon.js';
// OctagonPause aliases
export { default as PauseOctagon } from '../icons/pause-octagon.js';
// OctagonX aliases
export { default as XOctagon } from '../icons/x-octagon.js';
// PaintbrushVertical aliases
export { default as Paintbrush2 } from '../icons/paintbrush-2.js';
// PanelBottomDashed aliases
export { default as PanelBottomInactive } from '../icons/panel-bottom-inactive.js';
// PanelLeftClose aliases
export { default as SidebarClose } from '../icons/sidebar-close.js';
// PanelLeftDashed aliases
export { default as PanelLeftInactive } from '../icons/panel-left-inactive.js';
// PanelLeftOpen aliases
export { default as SidebarOpen } from '../icons/sidebar-open.js';
// PanelLeft aliases
export { default as Sidebar } from '../icons/sidebar.js';
// PanelRightDashed aliases
export { default as PanelRightInactive } from '../icons/panel-right-inactive.js';
// PanelTopDashed aliases
export { default as PanelTopInactive } from '../icons/panel-top-inactive.js';
// PanelsTopLeft aliases
export { default as Layout } from '../icons/layout.js';
// PenLine aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link PenLine} instead. This alias will be removed in v1.0 */
default as Edit3 } from '../icons/edit-3.js';
// Pen aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link Pen} instead. This alias will be removed in v1.0 */
default as Edit2 } from '../icons/edit-2.js';
// PlugZap aliases
export { default as PlugZap2 } from '../icons/plug-zap-2.js';
// RectangleEllipsis aliases
export { default as FormInput } from '../icons/form-input.js';
// Rotate3d aliases
export { default as Rotate3D } from '../icons/rotate-3-d.js';
// Rows2 aliases
export { default as Rows } from '../icons/rows.js';
// Rows3 aliases
export { default as PanelsTopBottom } from '../icons/panels-top-bottom.js';
// Scale3d aliases
export { default as Scale3D } from '../icons/scale-3-d.js';
// SendHorizontal aliases
export { 
/** @deprecated Renamed because of typo, use {@link SendHorizontal} instead. This alias will be removed in v1.0 */
default as SendHorizonal } from '../icons/send-horizonal.js';
// ShieldQuestionMark aliases
export { default as ShieldQuestion } from '../icons/shield-question.js';
// ShieldX aliases
export { default as ShieldClose } from '../icons/shield-close.js';
// SlidersVertical aliases
export { default as Sliders } from '../icons/sliders.js';
// Sparkles aliases
export { default as Stars } from '../icons/stars.js';
// SquareActivity aliases
export { default as ActivitySquare } from '../icons/activity-square.js';
// SquareArrowDownLeft aliases
export { default as ArrowDownLeftSquare } from '../icons/arrow-down-left-square.js';
// SquareArrowDownRight aliases
export { default as ArrowDownRightSquare } from '../icons/arrow-down-right-square.js';
// SquareArrowDown aliases
export { default as ArrowDownSquare } from '../icons/arrow-down-square.js';
// SquareArrowLeft aliases
export { default as ArrowLeftSquare } from '../icons/arrow-left-square.js';
// SquareArrowOutDownLeft aliases
export { default as ArrowDownLeftFromSquare } from '../icons/arrow-down-left-from-square.js';
// SquareArrowOutDownRight aliases
export { default as ArrowDownRightFromSquare } from '../icons/arrow-down-right-from-square.js';
// SquareArrowOutUpLeft aliases
export { default as ArrowUpLeftFromSquare } from '../icons/arrow-up-left-from-square.js';
// SquareArrowOutUpRight aliases
export { default as ArrowUpRightFromSquare } from '../icons/arrow-up-right-from-square.js';
// SquareArrowRight aliases
export { default as ArrowRightSquare } from '../icons/arrow-right-square.js';
// SquareArrowUpLeft aliases
export { default as ArrowUpLeftSquare } from '../icons/arrow-up-left-square.js';
// SquareArrowUpRight aliases
export { default as ArrowUpRightSquare } from '../icons/arrow-up-right-square.js';
// SquareArrowUp aliases
export { default as ArrowUpSquare } from '../icons/arrow-up-square.js';
// SquareAsterisk aliases
export { default as AsteriskSquare } from '../icons/asterisk-square.js';
// SquareBottomDashedScissors aliases
export { default as ScissorsSquareDashedBottom } from '../icons/scissors-square-dashed-bottom.js';
// SquareChartGantt aliases
export { default as GanttChartSquare } from '../icons/gantt-chart-square.js';
// SquareChartGantt aliases
export { default as SquareGanttChart } from '../icons/square-gantt-chart.js';
// SquareCheckBig aliases
export { default as CheckSquare } from '../icons/check-square.js';
// SquareCheck aliases
export { default as CheckSquare2 } from '../icons/check-square-2.js';
// SquareChevronDown aliases
export { default as ChevronDownSquare } from '../icons/chevron-down-square.js';
// SquareChevronLeft aliases
export { default as ChevronLeftSquare } from '../icons/chevron-left-square.js';
// SquareChevronRight aliases
export { default as ChevronRightSquare } from '../icons/chevron-right-square.js';
// SquareChevronUp aliases
export { default as ChevronUpSquare } from '../icons/chevron-up-square.js';
// SquareCode aliases
export { default as CodeSquare } from '../icons/code-square.js';
// SquareDashedKanban aliases
export { default as KanbanSquareDashed } from '../icons/kanban-square-dashed.js';
// SquareDashedMousePointer aliases
export { default as MousePointerSquareDashed } from '../icons/mouse-pointer-square-dashed.js';
// SquareDashed aliases
export { default as BoxSelect } from '../icons/box-select.js';
// SquareDivide aliases
export { default as DivideSquare } from '../icons/divide-square.js';
// SquareDot aliases
export { default as DotSquare } from '../icons/dot-square.js';
// SquareEqual aliases
export { default as EqualSquare } from '../icons/equal-square.js';
// SquareFunction aliases
export { default as FunctionSquare } from '../icons/function-square.js';
// SquareKanban aliases
export { default as KanbanSquare } from '../icons/kanban-square.js';
// SquareLibrary aliases
export { default as LibrarySquare } from '../icons/library-square.js';
// SquareM aliases
export { default as MSquare } from '../icons/m-square.js';
// SquareMenu aliases
export { default as MenuSquare } from '../icons/menu-square.js';
// SquareMousePointer aliases
export { default as Inspect } from '../icons/inspect.js';
// SquareMinus aliases
export { default as MinusSquare } from '../icons/minus-square.js';
// SquareParkingOff aliases
export { default as ParkingSquareOff } from '../icons/parking-square-off.js';
// SquareParking aliases
export { default as ParkingSquare } from '../icons/parking-square.js';
// SquarePen aliases
export { default as PenBox } from '../icons/pen-box.js';
// SquarePen aliases
export { default as Edit } from '../icons/edit.js';
// SquarePen aliases
export { default as PenSquare } from '../icons/pen-square.js';
// SquarePercent aliases
export { default as PercentSquare } from '../icons/percent-square.js';
// SquarePi aliases
export { default as PiSquare } from '../icons/pi-square.js';
// SquarePilcrow aliases
export { default as PilcrowSquare } from '../icons/pilcrow-square.js';
// SquarePlay aliases
export { default as PlaySquare } from '../icons/play-square.js';
// SquarePlus aliases
export { default as PlusSquare } from '../icons/plus-square.js';
// SquarePower aliases
export { default as PowerSquare } from '../icons/power-square.js';
// SquareScissors aliases
export { default as ScissorsSquare } from '../icons/scissors-square.js';
// SquareSigma aliases
export { default as SigmaSquare } from '../icons/sigma-square.js';
// SquareSlash aliases
export { default as SlashSquare } from '../icons/slash-square.js';
// SquareSplitHorizontal aliases
export { default as SplitSquareHorizontal } from '../icons/split-square-horizontal.js';
// SquareSplitVertical aliases
export { default as SplitSquareVertical } from '../icons/split-square-vertical.js';
// SquareUserRound aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link SquareUserRound} instead. This alias will be removed in v1.0 */
default as UserSquare2 } from '../icons/user-square-2.js';
// SquareTerminal aliases
export { default as TerminalSquare } from '../icons/terminal-square.js';
// SquareUser aliases
export { default as UserSquare } from '../icons/user-square.js';
// SquareX aliases
export { default as XSquare } from '../icons/x-square.js';
// TestTubeDiagonal aliases
export { default as TestTube2 } from '../icons/test-tube-2.js';
// TextSelect aliases
export { default as TextSelection } from '../icons/text-selection.js';
// TramFront aliases
export { default as Train } from '../icons/train.js';
// TreePalm aliases
export { default as Palmtree } from '../icons/palmtree.js';
// TriangleAlert aliases
export { default as AlertTriangle } from '../icons/alert-triangle.js';
// TvMinimal aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link TvMinimal} instead. This alias will be removed in v1.0 */
default as Tv2 } from '../icons/tv-2.js';
// University aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link University} instead. This alias will be removed in v1.0 */
default as School2 } from '../icons/school-2.js';
// UserRoundCheck aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundCheck} instead. This alias will be removed in v1.0 */
default as UserCheck2 } from '../icons/user-check-2.js';
// UserRoundCog aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundCog} instead. This alias will be removed in v1.0 */
default as UserCog2 } from '../icons/user-cog-2.js';
// UserRoundMinus aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundMinus} instead. This alias will be removed in v1.0 */
default as UserMinus2 } from '../icons/user-minus-2.js';
// UserRoundPlus aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundPlus} instead. This alias will be removed in v1.0 */
default as UserPlus2 } from '../icons/user-plus-2.js';
// UserRoundX aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundX} instead. This alias will be removed in v1.0 */
default as UserX2 } from '../icons/user-x-2.js';
// UserRound aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRound} instead. This alias will be removed in v1.0 */
default as User2 } from '../icons/user-2.js';
// UsersRound aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UsersRound} instead. This alias will be removed in v1.0 */
default as Users2 } from '../icons/users-2.js';
// UtensilsCrossed aliases
export { default as ForkKnifeCrossed } from '../icons/fork-knife-crossed.js';
// Utensils aliases
export { default as ForkKnife } from '../icons/fork-knife.js';
// WalletMinimal aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link WalletMinimal} instead. This alias will be removed in v1.0 */
default as Wallet2 } from '../icons/wallet-2.js';
// WandSparkles aliases
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link WandSparkles} instead. This alias will be removed in v1.0 */
default as Wand2 } from '../icons/wand-2.js';
