// @lucide/svelte/dist/aliases/suffixed.js
/**
 * @license @lucide/svelte v0.534.0 - ISC
 *
 * ISC License
 * 
 * Copyright (c) for portions of Lucide are held by Cole Bemis 2013-2022 as part of Feather (MIT). All other copyright (c) for Lucide are held by Lucide Contributors 2022.
 * 
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 * 
 */
// AArrowUp aliases
export { default as AArrowUpIcon } from '../icons/a-arrow-up.js';
// ALargeSmall aliases
export { default as ALargeSmallIcon } from '../icons/a-large-small.js';
// Accessibility aliases
export { default as AccessibilityIcon } from '../icons/accessibility.js';
// Activity aliases
export { default as ActivityIcon } from '../icons/activity.js';
// AirVent aliases
export { default as AirVentIcon } from '../icons/air-vent.js';
// Airplay aliases
export { default as AirplayIcon } from '../icons/airplay.js';
// AlarmClockOff aliases
export { default as AlarmClockOffIcon } from '../icons/alarm-clock-off.js';
// AlarmClock aliases
export { default as AlarmClockIcon } from '../icons/alarm-clock.js';
// AlarmSmoke aliases
export { default as AlarmSmokeIcon } from '../icons/alarm-smoke.js';
// AlignCenterHorizontal aliases
export { default as AlignCenterHorizontalIcon } from '../icons/align-center-horizontal.js';
// Album aliases
export { default as AlbumIcon } from '../icons/album.js';
// AlignCenterVertical aliases
export { default as AlignCenterVerticalIcon } from '../icons/align-center-vertical.js';
// AlignCenter aliases
export { default as AlignCenterIcon } from '../icons/align-center.js';
// AlignEndHorizontal aliases
export { default as AlignEndHorizontalIcon } from '../icons/align-end-horizontal.js';
// AlignEndVertical aliases
export { default as AlignEndVerticalIcon } from '../icons/align-end-vertical.js';
// AlignHorizontalDistributeCenter aliases
export { default as AlignHorizontalDistributeCenterIcon } from '../icons/align-horizontal-distribute-center.js';
// AlignHorizontalDistributeEnd aliases
export { default as AlignHorizontalDistributeEndIcon } from '../icons/align-horizontal-distribute-end.js';
// AlignHorizontalDistributeStart aliases
export { default as AlignHorizontalDistributeStartIcon } from '../icons/align-horizontal-distribute-start.js';
// AlignHorizontalJustifyCenter aliases
export { default as AlignHorizontalJustifyCenterIcon } from '../icons/align-horizontal-justify-center.js';
// AlignHorizontalJustifyStart aliases
export { default as AlignHorizontalJustifyStartIcon } from '../icons/align-horizontal-justify-start.js';
// AlignHorizontalJustifyEnd aliases
export { default as AlignHorizontalJustifyEndIcon } from '../icons/align-horizontal-justify-end.js';
// AlignHorizontalSpaceBetween aliases
export { default as AlignHorizontalSpaceBetweenIcon } from '../icons/align-horizontal-space-between.js';
// AlignJustify aliases
export { default as AlignJustifyIcon } from '../icons/align-justify.js';
// AlignHorizontalSpaceAround aliases
export { default as AlignHorizontalSpaceAroundIcon } from '../icons/align-horizontal-space-around.js';
// AlignLeft aliases
export { default as AlignLeftIcon } from '../icons/align-left.js';
// AlignRight aliases
export { default as AlignRightIcon } from '../icons/align-right.js';
// AlignStartHorizontal aliases
export { default as AlignStartHorizontalIcon } from '../icons/align-start-horizontal.js';
// AlignVerticalDistributeCenter aliases
export { default as AlignVerticalDistributeCenterIcon } from '../icons/align-vertical-distribute-center.js';
// AlignStartVertical aliases
export { default as AlignStartVerticalIcon } from '../icons/align-start-vertical.js';
// AlignVerticalDistributeEnd aliases
export { default as AlignVerticalDistributeEndIcon } from '../icons/align-vertical-distribute-end.js';
// AlignVerticalDistributeStart aliases
export { default as AlignVerticalDistributeStartIcon } from '../icons/align-vertical-distribute-start.js';
// AlignVerticalJustifyCenter aliases
export { default as AlignVerticalJustifyCenterIcon } from '../icons/align-vertical-justify-center.js';
// AlignVerticalJustifyEnd aliases
export { default as AlignVerticalJustifyEndIcon } from '../icons/align-vertical-justify-end.js';
// AlignVerticalSpaceAround aliases
export { default as AlignVerticalSpaceAroundIcon } from '../icons/align-vertical-space-around.js';
// AlignVerticalJustifyStart aliases
export { default as AlignVerticalJustifyStartIcon } from '../icons/align-vertical-justify-start.js';
// AlignVerticalSpaceBetween aliases
export { default as AlignVerticalSpaceBetweenIcon } from '../icons/align-vertical-space-between.js';
// Ambulance aliases
export { default as AmbulanceIcon } from '../icons/ambulance.js';
// Ampersand aliases
export { default as AmpersandIcon } from '../icons/ampersand.js';
// Amphora aliases
export { default as AmphoraIcon } from '../icons/amphora.js';
// Ampersands aliases
export { default as AmpersandsIcon } from '../icons/ampersands.js';
// Anchor aliases
export { default as AnchorIcon } from '../icons/anchor.js';
// Angry aliases
export { default as AngryIcon } from '../icons/angry.js';
// Annoyed aliases
export { default as AnnoyedIcon } from '../icons/annoyed.js';
// Anvil aliases
export { default as AnvilIcon } from '../icons/anvil.js';
// Antenna aliases
export { default as AntennaIcon } from '../icons/antenna.js';
// Aperture aliases
export { default as ApertureIcon } from '../icons/aperture.js';
// AppWindowMac aliases
export { default as AppWindowMacIcon } from '../icons/app-window-mac.js';
// AppWindow aliases
export { default as AppWindowIcon } from '../icons/app-window.js';
// Apple aliases
export { default as AppleIcon } from '../icons/apple.js';
// ArchiveRestore aliases
export { default as ArchiveRestoreIcon } from '../icons/archive-restore.js';
// ArchiveX aliases
export { default as ArchiveXIcon } from '../icons/archive-x.js';
// Archive aliases
export { default as ArchiveIcon } from '../icons/archive.js';
// Armchair aliases
export { default as ArmchairIcon } from '../icons/armchair.js';
// ArrowBigDownDash aliases
export { default as ArrowBigDownDashIcon } from '../icons/arrow-big-down-dash.js';
// ArrowBigDown aliases
export { default as ArrowBigDownIcon } from '../icons/arrow-big-down.js';
// ArrowBigLeftDash aliases
export { default as ArrowBigLeftDashIcon } from '../icons/arrow-big-left-dash.js';
// ArrowBigRightDash aliases
export { default as ArrowBigRightDashIcon } from '../icons/arrow-big-right-dash.js';
// ArrowBigLeft aliases
export { default as ArrowBigLeftIcon } from '../icons/arrow-big-left.js';
// ArrowBigRight aliases
export { default as ArrowBigRightIcon } from '../icons/arrow-big-right.js';
// ArrowBigUpDash aliases
export { default as ArrowBigUpDashIcon } from '../icons/arrow-big-up-dash.js';
// ArrowBigUp aliases
export { default as ArrowBigUpIcon } from '../icons/arrow-big-up.js';
// ArrowDownFromLine aliases
export { default as ArrowDownFromLineIcon } from '../icons/arrow-down-from-line.js';
// ArrowDownLeft aliases
export { default as ArrowDownLeftIcon } from '../icons/arrow-down-left.js';
// ArrowDownNarrowWide aliases
export { default as ArrowDownNarrowWideIcon } from '../icons/arrow-down-narrow-wide.js';
// ArrowDownRight aliases
export { default as ArrowDownRightIcon } from '../icons/arrow-down-right.js';
// ArrowDownToDot aliases
export { default as ArrowDownToDotIcon } from '../icons/arrow-down-to-dot.js';
// ArrowDownToLine aliases
export { default as ArrowDownToLineIcon } from '../icons/arrow-down-to-line.js';
// ArrowDownUp aliases
export { default as ArrowDownUpIcon } from '../icons/arrow-down-up.js';
// ArrowDown aliases
export { default as ArrowDownIcon } from '../icons/arrow-down.js';
// ArrowLeftFromLine aliases
export { default as ArrowLeftFromLineIcon } from '../icons/arrow-left-from-line.js';
// ArrowLeftToLine aliases
export { default as ArrowLeftToLineIcon } from '../icons/arrow-left-to-line.js';
// ArrowLeftRight aliases
export { default as ArrowLeftRightIcon } from '../icons/arrow-left-right.js';
// ArrowLeft aliases
export { default as ArrowLeftIcon } from '../icons/arrow-left.js';
// ArrowRightFromLine aliases
export { default as ArrowRightFromLineIcon } from '../icons/arrow-right-from-line.js';
// ArrowRightToLine aliases
export { default as ArrowRightToLineIcon } from '../icons/arrow-right-to-line.js';
// ArrowRightLeft aliases
export { default as ArrowRightLeftIcon } from '../icons/arrow-right-left.js';
// ArrowRight aliases
export { default as ArrowRightIcon } from '../icons/arrow-right.js';
// ArrowUpFromDot aliases
export { default as ArrowUpFromDotIcon } from '../icons/arrow-up-from-dot.js';
// ArrowUpDown aliases
export { default as ArrowUpDownIcon } from '../icons/arrow-up-down.js';
// ArrowUpFromLine aliases
export { default as ArrowUpFromLineIcon } from '../icons/arrow-up-from-line.js';
// ArrowUpLeft aliases
export { default as ArrowUpLeftIcon } from '../icons/arrow-up-left.js';
// ArrowUpRight aliases
export { default as ArrowUpRightIcon } from '../icons/arrow-up-right.js';
// ArrowUpToLine aliases
export { default as ArrowUpToLineIcon } from '../icons/arrow-up-to-line.js';
// ArrowUpWideNarrow aliases
export { default as ArrowUpWideNarrowIcon } from '../icons/arrow-up-wide-narrow.js';
// ArrowUp aliases
export { default as ArrowUpIcon } from '../icons/arrow-up.js';
// Asterisk aliases
export { default as AsteriskIcon } from '../icons/asterisk.js';
// ArrowsUpFromLine aliases
export { default as ArrowsUpFromLineIcon } from '../icons/arrows-up-from-line.js';
// AtSign aliases
export { default as AtSignIcon } from '../icons/at-sign.js';
// AudioLines aliases
export { default as AudioLinesIcon } from '../icons/audio-lines.js';
// AudioWaveform aliases
export { default as AudioWaveformIcon } from '../icons/audio-waveform.js';
// Atom aliases
export { default as AtomIcon } from '../icons/atom.js';
// Award aliases
export { default as AwardIcon } from '../icons/award.js';
// Axe aliases
export { default as AxeIcon } from '../icons/axe.js';
// Baby aliases
export { default as BabyIcon } from '../icons/baby.js';
// Backpack aliases
export { default as BackpackIcon } from '../icons/backpack.js';
// BadgeAlert aliases
export { default as BadgeAlertIcon } from '../icons/badge-alert.js';
// BadgeCent aliases
export { default as BadgeCentIcon } from '../icons/badge-cent.js';
// BadgeEuro aliases
export { default as BadgeEuroIcon } from '../icons/badge-euro.js';
// BadgeDollarSign aliases
export { default as BadgeDollarSignIcon } from '../icons/badge-dollar-sign.js';
// BadgeIndianRupee aliases
export { default as BadgeIndianRupeeIcon } from '../icons/badge-indian-rupee.js';
// BadgeInfo aliases
export { default as BadgeInfoIcon } from '../icons/badge-info.js';
// BadgeJapaneseYen aliases
export { default as BadgeJapaneseYenIcon } from '../icons/badge-japanese-yen.js';
// BadgeMinus aliases
export { default as BadgeMinusIcon } from '../icons/badge-minus.js';
// BadgePercent aliases
export { default as BadgePercentIcon } from '../icons/badge-percent.js';
// BadgePlus aliases
export { default as BadgePlusIcon } from '../icons/badge-plus.js';
// BadgePoundSterling aliases
export { default as BadgePoundSterlingIcon } from '../icons/badge-pound-sterling.js';
// BadgeRussianRuble aliases
export { default as BadgeRussianRubleIcon } from '../icons/badge-russian-ruble.js';
// BadgeSwissFranc aliases
export { default as BadgeSwissFrancIcon } from '../icons/badge-swiss-franc.js';
// BadgeTurkishLira aliases
export { default as BadgeTurkishLiraIcon } from '../icons/badge-turkish-lira.js';
// BadgeX aliases
export { default as BadgeXIcon } from '../icons/badge-x.js';
// Badge aliases
export { default as BadgeIcon } from '../icons/badge.js';
// BaggageClaim aliases
export { default as BaggageClaimIcon } from '../icons/baggage-claim.js';
// Ban aliases
export { default as BanIcon } from '../icons/ban.js';
// Banana aliases
export { default as BananaIcon } from '../icons/banana.js';
// Bandage aliases
export { default as BandageIcon } from '../icons/bandage.js';
// BanknoteArrowDown aliases
export { default as BanknoteArrowDownIcon } from '../icons/banknote-arrow-down.js';
// BanknoteArrowUp aliases
export { default as BanknoteArrowUpIcon } from '../icons/banknote-arrow-up.js';
// BanknoteX aliases
export { default as BanknoteXIcon } from '../icons/banknote-x.js';
// Banknote aliases
export { default as BanknoteIcon } from '../icons/banknote.js';
// Barcode aliases
export { default as BarcodeIcon } from '../icons/barcode.js';
// Baseline aliases
export { default as BaselineIcon } from '../icons/baseline.js';
// Barrel aliases
export { default as BarrelIcon } from '../icons/barrel.js';
// Bath aliases
export { default as BathIcon } from '../icons/bath.js';
// BatteryCharging aliases
export { default as BatteryChargingIcon } from '../icons/battery-charging.js';
// BatteryFull aliases
export { default as BatteryFullIcon } from '../icons/battery-full.js';
// BatteryLow aliases
export { default as BatteryLowIcon } from '../icons/battery-low.js';
// BatteryPlus aliases
export { default as BatteryPlusIcon } from '../icons/battery-plus.js';
// BatteryMedium aliases
export { default as BatteryMediumIcon } from '../icons/battery-medium.js';
// BatteryWarning aliases
export { default as BatteryWarningIcon } from '../icons/battery-warning.js';
// Battery aliases
export { default as BatteryIcon } from '../icons/battery.js';
// Beaker aliases
export { default as BeakerIcon } from '../icons/beaker.js';
// BedDouble aliases
export { default as BedDoubleIcon } from '../icons/bed-double.js';
// BeanOff aliases
export { default as BeanOffIcon } from '../icons/bean-off.js';
// BedSingle aliases
export { default as BedSingleIcon } from '../icons/bed-single.js';
// AArrowDown aliases
export { default as AArrowDownIcon } from '../icons/a-arrow-down.js';
// Bean aliases
export { default as BeanIcon } from '../icons/bean.js';
// Bed aliases
export { default as BedIcon } from '../icons/bed.js';
// Beef aliases
export { default as BeefIcon } from '../icons/beef.js';
// BeerOff aliases
export { default as BeerOffIcon } from '../icons/beer-off.js';
// Beer aliases
export { default as BeerIcon } from '../icons/beer.js';
// BellElectric aliases
export { default as BellElectricIcon } from '../icons/bell-electric.js';
// BellDot aliases
export { default as BellDotIcon } from '../icons/bell-dot.js';
// BellMinus aliases
export { default as BellMinusIcon } from '../icons/bell-minus.js';
// BellOff aliases
export { default as BellOffIcon } from '../icons/bell-off.js';
// BellPlus aliases
export { default as BellPlusIcon } from '../icons/bell-plus.js';
// BellRing aliases
export { default as BellRingIcon } from '../icons/bell-ring.js';
// Bell aliases
export { default as BellIcon } from '../icons/bell.js';
// BetweenVerticalEnd aliases
export { default as BetweenVerticalEndIcon } from '../icons/between-vertical-end.js';
// BetweenVerticalStart aliases
export { default as BetweenVerticalStartIcon } from '../icons/between-vertical-start.js';
// BicepsFlexed aliases
export { default as BicepsFlexedIcon } from '../icons/biceps-flexed.js';
// Bike aliases
export { default as BikeIcon } from '../icons/bike.js';
// Binoculars aliases
export { default as BinocularsIcon } from '../icons/binoculars.js';
// Binary aliases
export { default as BinaryIcon } from '../icons/binary.js';
// Bird aliases
export { default as BirdIcon } from '../icons/bird.js';
// Bitcoin aliases
export { default as BitcoinIcon } from '../icons/bitcoin.js';
// Biohazard aliases
export { default as BiohazardIcon } from '../icons/biohazard.js';
// Blend aliases
export { default as BlendIcon } from '../icons/blend.js';
// Blocks aliases
export { default as BlocksIcon } from '../icons/blocks.js';
// BluetoothConnected aliases
export { default as BluetoothConnectedIcon } from '../icons/bluetooth-connected.js';
// Blinds aliases
export { default as BlindsIcon } from '../icons/blinds.js';
// BluetoothOff aliases
export { default as BluetoothOffIcon } from '../icons/bluetooth-off.js';
// BluetoothSearching aliases
export { default as BluetoothSearchingIcon } from '../icons/bluetooth-searching.js';
// Bold aliases
export { default as BoldIcon } from '../icons/bold.js';
// Bluetooth aliases
export { default as BluetoothIcon } from '../icons/bluetooth.js';
// Bolt aliases
export { default as BoltIcon } from '../icons/bolt.js';
// Bomb aliases
export { default as BombIcon } from '../icons/bomb.js';
// Bone aliases
export { default as BoneIcon } from '../icons/bone.js';
// BookAlert aliases
export { default as BookAlertIcon } from '../icons/book-alert.js';
// BookA aliases
export { default as BookAIcon } from '../icons/book-a.js';
// BookAudio aliases
export { default as BookAudioIcon } from '../icons/book-audio.js';
// BookCheck aliases
export { default as BookCheckIcon } from '../icons/book-check.js';
// BookCopy aliases
export { default as BookCopyIcon } from '../icons/book-copy.js';
// BookDown aliases
export { default as BookDownIcon } from '../icons/book-down.js';
// BookHeadphones aliases
export { default as BookHeadphonesIcon } from '../icons/book-headphones.js';
// BookHeart aliases
export { default as BookHeartIcon } from '../icons/book-heart.js';
// BookImage aliases
export { default as BookImageIcon } from '../icons/book-image.js';
// BookKey aliases
export { default as BookKeyIcon } from '../icons/book-key.js';
// BookLock aliases
export { default as BookLockIcon } from '../icons/book-lock.js';
// BookMarked aliases
export { default as BookMarkedIcon } from '../icons/book-marked.js';
// BookMinus aliases
export { default as BookMinusIcon } from '../icons/book-minus.js';
// BookOpenCheck aliases
export { default as BookOpenCheckIcon } from '../icons/book-open-check.js';
// BookOpenText aliases
export { default as BookOpenTextIcon } from '../icons/book-open-text.js';
// BookOpen aliases
export { default as BookOpenIcon } from '../icons/book-open.js';
// BookPlus aliases
export { default as BookPlusIcon } from '../icons/book-plus.js';
// BookText aliases
export { default as BookTextIcon } from '../icons/book-text.js';
// BookType aliases
export { default as BookTypeIcon } from '../icons/book-type.js';
// BookUp2 aliases
export { default as BookUp2Icon } from '../icons/book-up-2.js';
// BookUp aliases
export { default as BookUpIcon } from '../icons/book-up.js';
// BookUser aliases
export { default as BookUserIcon } from '../icons/book-user.js';
// BookX aliases
export { default as BookXIcon } from '../icons/book-x.js';
// Book aliases
export { default as BookIcon } from '../icons/book.js';
// BookmarkCheck aliases
export { default as BookmarkCheckIcon } from '../icons/bookmark-check.js';
// BookmarkMinus aliases
export { default as BookmarkMinusIcon } from '../icons/bookmark-minus.js';
// BookmarkPlus aliases
export { default as BookmarkPlusIcon } from '../icons/bookmark-plus.js';
// BookmarkX aliases
export { default as BookmarkXIcon } from '../icons/bookmark-x.js';
// Bookmark aliases
export { default as BookmarkIcon } from '../icons/bookmark.js';
// BoomBox aliases
export { default as BoomBoxIcon } from '../icons/boom-box.js';
// BotMessageSquare aliases
export { default as BotMessageSquareIcon } from '../icons/bot-message-square.js';
// BotOff aliases
export { default as BotOffIcon } from '../icons/bot-off.js';
// Bot aliases
export { default as BotIcon } from '../icons/bot.js';
// BottleWine aliases
export { default as BottleWineIcon } from '../icons/bottle-wine.js';
// BowArrow aliases
export { default as BowArrowIcon } from '../icons/bow-arrow.js';
// Box aliases
export { default as BoxIcon } from '../icons/box.js';
// Boxes aliases
export { default as BoxesIcon } from '../icons/boxes.js';
// Brackets aliases
export { default as BracketsIcon } from '../icons/brackets.js';
// BrainCircuit aliases
export { default as BrainCircuitIcon } from '../icons/brain-circuit.js';
// BrainCog aliases
export { default as BrainCogIcon } from '../icons/brain-cog.js';
// Brain aliases
export { default as BrainIcon } from '../icons/brain.js';
// BrickWallFire aliases
export { default as BrickWallFireIcon } from '../icons/brick-wall-fire.js';
// BriefcaseBusiness aliases
export { default as BriefcaseBusinessIcon } from '../icons/briefcase-business.js';
// BrickWall aliases
export { default as BrickWallIcon } from '../icons/brick-wall.js';
// BriefcaseConveyorBelt aliases
export { default as BriefcaseConveyorBeltIcon } from '../icons/briefcase-conveyor-belt.js';
// BriefcaseMedical aliases
export { default as BriefcaseMedicalIcon } from '../icons/briefcase-medical.js';
// Briefcase aliases
export { default as BriefcaseIcon } from '../icons/briefcase.js';
// BringToFront aliases
export { default as BringToFrontIcon } from '../icons/bring-to-front.js';
// BrushCleaning aliases
export { default as BrushCleaningIcon } from '../icons/brush-cleaning.js';
// Brush aliases
export { default as BrushIcon } from '../icons/brush.js';
// Bubbles aliases
export { default as BubblesIcon } from '../icons/bubbles.js';
// BugOff aliases
export { default as BugOffIcon } from '../icons/bug-off.js';
// BugPlay aliases
export { default as BugPlayIcon } from '../icons/bug-play.js';
// Bug aliases
export { default as BugIcon } from '../icons/bug.js';
// Building2 aliases
export { default as Building2Icon } from '../icons/building-2.js';
// Building aliases
export { default as BuildingIcon } from '../icons/building.js';
// BusFront aliases
export { default as BusFrontIcon } from '../icons/bus-front.js';
// Bus aliases
export { default as BusIcon } from '../icons/bus.js';
// CableCar aliases
export { default as CableCarIcon } from '../icons/cable-car.js';
// Cable aliases
export { default as CableIcon } from '../icons/cable.js';
// CakeSlice aliases
export { default as CakeSliceIcon } from '../icons/cake-slice.js';
// Cake aliases
export { default as CakeIcon } from '../icons/cake.js';
// Calculator aliases
export { default as CalculatorIcon } from '../icons/calculator.js';
// Calendar1 aliases
export { default as Calendar1Icon } from '../icons/calendar-1.js';
// CalendarArrowDown aliases
export { default as CalendarArrowDownIcon } from '../icons/calendar-arrow-down.js';
// CalendarArrowUp aliases
export { default as CalendarArrowUpIcon } from '../icons/calendar-arrow-up.js';
// CalendarCheck2 aliases
export { default as CalendarCheck2Icon } from '../icons/calendar-check-2.js';
// CalendarClock aliases
export { default as CalendarClockIcon } from '../icons/calendar-clock.js';
// CalendarCheck aliases
export { default as CalendarCheckIcon } from '../icons/calendar-check.js';
// CalendarCog aliases
export { default as CalendarCogIcon } from '../icons/calendar-cog.js';
// CalendarDays aliases
export { default as CalendarDaysIcon } from '../icons/calendar-days.js';
// CalendarFold aliases
export { default as CalendarFoldIcon } from '../icons/calendar-fold.js';
// CalendarHeart aliases
export { default as CalendarHeartIcon } from '../icons/calendar-heart.js';
// CalendarMinus2 aliases
export { default as CalendarMinus2Icon } from '../icons/calendar-minus-2.js';
// CalendarMinus aliases
export { default as CalendarMinusIcon } from '../icons/calendar-minus.js';
// CalendarOff aliases
export { default as CalendarOffIcon } from '../icons/calendar-off.js';
// CalendarPlus2 aliases
export { default as CalendarPlus2Icon } from '../icons/calendar-plus-2.js';
// CalendarPlus aliases
export { default as CalendarPlusIcon } from '../icons/calendar-plus.js';
// CalendarRange aliases
export { default as CalendarRangeIcon } from '../icons/calendar-range.js';
// CalendarSearch aliases
export { default as CalendarSearchIcon } from '../icons/calendar-search.js';
// CalendarSync aliases
export { default as CalendarSyncIcon } from '../icons/calendar-sync.js';
// CalendarX2 aliases
export { default as CalendarX2Icon } from '../icons/calendar-x-2.js';
// CalendarX aliases
export { default as CalendarXIcon } from '../icons/calendar-x.js';
// Calendar aliases
export { default as CalendarIcon } from '../icons/calendar.js';
// CameraOff aliases
export { default as CameraOffIcon } from '../icons/camera-off.js';
// Camera aliases
export { default as CameraIcon } from '../icons/camera.js';
// CandyCane aliases
export { default as CandyCaneIcon } from '../icons/candy-cane.js';
// CandyOff aliases
export { default as CandyOffIcon } from '../icons/candy-off.js';
// Candy aliases
export { default as CandyIcon } from '../icons/candy.js';
// Cannabis aliases
export { default as CannabisIcon } from '../icons/cannabis.js';
// CaptionsOff aliases
export { default as CaptionsOffIcon } from '../icons/captions-off.js';
// CarFront aliases
export { default as CarFrontIcon } from '../icons/car-front.js';
// CarTaxiFront aliases
export { default as CarTaxiFrontIcon } from '../icons/car-taxi-front.js';
// Car aliases
export { default as CarIcon } from '../icons/car.js';
// CardSim aliases
export { default as CardSimIcon } from '../icons/card-sim.js';
// Caravan aliases
export { default as CaravanIcon } from '../icons/caravan.js';
// Carrot aliases
export { default as CarrotIcon } from '../icons/carrot.js';
// CaseLower aliases
export { default as CaseLowerIcon } from '../icons/case-lower.js';
// CaseSensitive aliases
export { default as CaseSensitiveIcon } from '../icons/case-sensitive.js';
// CaseUpper aliases
export { default as CaseUpperIcon } from '../icons/case-upper.js';
// CassetteTape aliases
export { default as CassetteTapeIcon } from '../icons/cassette-tape.js';
// Cast aliases
export { default as CastIcon } from '../icons/cast.js';
// Castle aliases
export { default as CastleIcon } from '../icons/castle.js';
// Cat aliases
export { default as CatIcon } from '../icons/cat.js';
// Cctv aliases
export { default as CctvIcon } from '../icons/cctv.js';
// ChartBarDecreasing aliases
export { default as ChartBarDecreasingIcon } from '../icons/chart-bar-decreasing.js';
// ChartBarIncreasing aliases
export { default as ChartBarIncreasingIcon } from '../icons/chart-bar-increasing.js';
// ChartBarStacked aliases
export { default as ChartBarStackedIcon } from '../icons/chart-bar-stacked.js';
// ChartColumnDecreasing aliases
export { default as ChartColumnDecreasingIcon } from '../icons/chart-column-decreasing.js';
// ChartColumnStacked aliases
export { default as ChartColumnStackedIcon } from '../icons/chart-column-stacked.js';
// ChartGantt aliases
export { default as ChartGanttIcon } from '../icons/chart-gantt.js';
// ChartNetwork aliases
export { default as ChartNetworkIcon } from '../icons/chart-network.js';
// ChartNoAxesColumnDecreasing aliases
export { default as ChartNoAxesColumnDecreasingIcon } from '../icons/chart-no-axes-column-decreasing.js';
// ChartNoAxesCombined aliases
export { default as ChartNoAxesCombinedIcon } from '../icons/chart-no-axes-combined.js';
// ChartSpline aliases
export { default as ChartSplineIcon } from '../icons/chart-spline.js';
// CheckCheck aliases
export { default as CheckCheckIcon } from '../icons/check-check.js';
// Check aliases
export { default as CheckIcon } from '../icons/check.js';
// CheckLine aliases
export { default as CheckLineIcon } from '../icons/check-line.js';
// ChefHat aliases
export { default as ChefHatIcon } from '../icons/chef-hat.js';
// Cherry aliases
export { default as CherryIcon } from '../icons/cherry.js';
// ChevronDown aliases
export { default as ChevronDownIcon } from '../icons/chevron-down.js';
// ChevronFirst aliases
export { default as ChevronFirstIcon } from '../icons/chevron-first.js';
// ChevronLast aliases
export { default as ChevronLastIcon } from '../icons/chevron-last.js';
// ChevronLeft aliases
export { default as ChevronLeftIcon } from '../icons/chevron-left.js';
// ChevronRight aliases
export { default as ChevronRightIcon } from '../icons/chevron-right.js';
// ChevronUp aliases
export { default as ChevronUpIcon } from '../icons/chevron-up.js';
// ChevronsDownUp aliases
export { default as ChevronsDownUpIcon } from '../icons/chevrons-down-up.js';
// ChevronsDown aliases
export { default as ChevronsDownIcon } from '../icons/chevrons-down.js';
// ChevronsLeftRightEllipsis aliases
export { default as ChevronsLeftRightEllipsisIcon } from '../icons/chevrons-left-right-ellipsis.js';
// ChevronsLeftRight aliases
export { default as ChevronsLeftRightIcon } from '../icons/chevrons-left-right.js';
// ChevronsLeft aliases
export { default as ChevronsLeftIcon } from '../icons/chevrons-left.js';
// ChevronsRightLeft aliases
export { default as ChevronsRightLeftIcon } from '../icons/chevrons-right-left.js';
// ChevronsRight aliases
export { default as ChevronsRightIcon } from '../icons/chevrons-right.js';
// ChevronsUpDown aliases
export { default as ChevronsUpDownIcon } from '../icons/chevrons-up-down.js';
// ChevronsUp aliases
export { default as ChevronsUpIcon } from '../icons/chevrons-up.js';
// Chrome aliases
export { default as ChromeIcon } from '../icons/chrome.js';
// Church aliases
export { default as ChurchIcon } from '../icons/church.js';
// CigaretteOff aliases
export { default as CigaretteOffIcon } from '../icons/cigarette-off.js';
// Cigarette aliases
export { default as CigaretteIcon } from '../icons/cigarette.js';
// CircleDashed aliases
export { default as CircleDashedIcon } from '../icons/circle-dashed.js';
// CircleDollarSign aliases
export { default as CircleDollarSignIcon } from '../icons/circle-dollar-sign.js';
// CircleDotDashed aliases
export { default as CircleDotDashedIcon } from '../icons/circle-dot-dashed.js';
// CircleDot aliases
export { default as CircleDotIcon } from '../icons/circle-dot.js';
// CircleEllipsis aliases
export { default as CircleEllipsisIcon } from '../icons/circle-ellipsis.js';
// CircleEqual aliases
export { default as CircleEqualIcon } from '../icons/circle-equal.js';
// CircleFadingArrowUp aliases
export { default as CircleFadingArrowUpIcon } from '../icons/circle-fading-arrow-up.js';
// CircleFadingPlus aliases
export { default as CircleFadingPlusIcon } from '../icons/circle-fading-plus.js';
// CircleOff aliases
export { default as CircleOffIcon } from '../icons/circle-off.js';
// CirclePoundSterling aliases
export { default as CirclePoundSterlingIcon } from '../icons/circle-pound-sterling.js';
// CircleSlash aliases
export { default as CircleSlashIcon } from '../icons/circle-slash.js';
// CircleSmall aliases
export { default as CircleSmallIcon } from '../icons/circle-small.js';
// Circle aliases
export { default as CircleIcon } from '../icons/circle.js';
// CircuitBoard aliases
export { default as CircuitBoardIcon } from '../icons/circuit-board.js';
// Clapperboard aliases
export { default as ClapperboardIcon } from '../icons/clapperboard.js';
// Citrus aliases
export { default as CitrusIcon } from '../icons/citrus.js';
// ClipboardCheck aliases
export { default as ClipboardCheckIcon } from '../icons/clipboard-check.js';
// ClipboardCopy aliases
export { default as ClipboardCopyIcon } from '../icons/clipboard-copy.js';
// ClipboardClock aliases
export { default as ClipboardClockIcon } from '../icons/clipboard-clock.js';
// ClipboardList aliases
export { default as ClipboardListIcon } from '../icons/clipboard-list.js';
// ClipboardMinus aliases
export { default as ClipboardMinusIcon } from '../icons/clipboard-minus.js';
// ClipboardPaste aliases
export { default as ClipboardPasteIcon } from '../icons/clipboard-paste.js';
// ClipboardPlus aliases
export { default as ClipboardPlusIcon } from '../icons/clipboard-plus.js';
// ClipboardType aliases
export { default as ClipboardTypeIcon } from '../icons/clipboard-type.js';
// ClipboardX aliases
export { default as ClipboardXIcon } from '../icons/clipboard-x.js';
// Clipboard aliases
export { default as ClipboardIcon } from '../icons/clipboard.js';
// Clock1 aliases
export { default as Clock1Icon } from '../icons/clock-1.js';
// Clock10 aliases
export { default as Clock10Icon } from '../icons/clock-10.js';
// Clock12 aliases
export { default as Clock12Icon } from '../icons/clock-12.js';
// Clock11 aliases
export { default as Clock11Icon } from '../icons/clock-11.js';
// Clock2 aliases
export { default as Clock2Icon } from '../icons/clock-2.js';
// Clock3 aliases
export { default as Clock3Icon } from '../icons/clock-3.js';
// Clock4 aliases
export { default as Clock4Icon } from '../icons/clock-4.js';
// Clock6 aliases
export { default as Clock6Icon } from '../icons/clock-6.js';
// Clock5 aliases
export { default as Clock5Icon } from '../icons/clock-5.js';
// Clock7 aliases
export { default as Clock7Icon } from '../icons/clock-7.js';
// Clock8 aliases
export { default as Clock8Icon } from '../icons/clock-8.js';
// Clock9 aliases
export { default as Clock9Icon } from '../icons/clock-9.js';
// ClockAlert aliases
export { default as ClockAlertIcon } from '../icons/clock-alert.js';
// ClockArrowDown aliases
export { default as ClockArrowDownIcon } from '../icons/clock-arrow-down.js';
// ClockArrowUp aliases
export { default as ClockArrowUpIcon } from '../icons/clock-arrow-up.js';
// ClockFading aliases
export { default as ClockFadingIcon } from '../icons/clock-fading.js';
// ClockPlus aliases
export { default as ClockPlusIcon } from '../icons/clock-plus.js';
// ClosedCaption aliases
export { default as ClosedCaptionIcon } from '../icons/closed-caption.js';
// Clock aliases
export { default as ClockIcon } from '../icons/clock.js';
// CloudAlert aliases
export { default as CloudAlertIcon } from '../icons/cloud-alert.js';
// CloudCheck aliases
export { default as CloudCheckIcon } from '../icons/cloud-check.js';
// CloudCog aliases
export { default as CloudCogIcon } from '../icons/cloud-cog.js';
// CloudDrizzle aliases
export { default as CloudDrizzleIcon } from '../icons/cloud-drizzle.js';
// CloudFog aliases
export { default as CloudFogIcon } from '../icons/cloud-fog.js';
// CloudHail aliases
export { default as CloudHailIcon } from '../icons/cloud-hail.js';
// CloudLightning aliases
export { default as CloudLightningIcon } from '../icons/cloud-lightning.js';
// CloudMoonRain aliases
export { default as CloudMoonRainIcon } from '../icons/cloud-moon-rain.js';
// CloudMoon aliases
export { default as CloudMoonIcon } from '../icons/cloud-moon.js';
// CloudOff aliases
export { default as CloudOffIcon } from '../icons/cloud-off.js';
// CloudRainWind aliases
export { default as CloudRainWindIcon } from '../icons/cloud-rain-wind.js';
// CloudRain aliases
export { default as CloudRainIcon } from '../icons/cloud-rain.js';
// CloudSnow aliases
export { default as CloudSnowIcon } from '../icons/cloud-snow.js';
// CloudSun aliases
export { default as CloudSunIcon } from '../icons/cloud-sun.js';
// CloudSunRain aliases
export { default as CloudSunRainIcon } from '../icons/cloud-sun-rain.js';
// Cloud aliases
export { default as CloudIcon } from '../icons/cloud.js';
// Cloudy aliases
export { default as CloudyIcon } from '../icons/cloudy.js';
// Clover aliases
export { default as CloverIcon } from '../icons/clover.js';
// Club aliases
export { default as ClubIcon } from '../icons/club.js';
// Code aliases
export { default as CodeIcon } from '../icons/code.js';
// Codepen aliases
export { default as CodepenIcon } from '../icons/codepen.js';
// Codesandbox aliases
export { default as CodesandboxIcon } from '../icons/codesandbox.js';
// Coffee aliases
export { default as CoffeeIcon } from '../icons/coffee.js';
// Cog aliases
export { default as CogIcon } from '../icons/cog.js';
// Coins aliases
export { default as CoinsIcon } from '../icons/coins.js';
// Columns4 aliases
export { default as Columns4Icon } from '../icons/columns-4.js';
// Combine aliases
export { default as CombineIcon } from '../icons/combine.js';
// Command aliases
export { default as CommandIcon } from '../icons/command.js';
// Compass aliases
export { default as CompassIcon } from '../icons/compass.js';
// Component aliases
export { default as ComponentIcon } from '../icons/component.js';
// ConciergeBell aliases
export { default as ConciergeBellIcon } from '../icons/concierge-bell.js';
// Computer aliases
export { default as ComputerIcon } from '../icons/computer.js';
// Cone aliases
export { default as ConeIcon } from '../icons/cone.js';
// Construction aliases
export { default as ConstructionIcon } from '../icons/construction.js';
// Contact aliases
export { default as ContactIcon } from '../icons/contact.js';
// Container aliases
export { default as ContainerIcon } from '../icons/container.js';
// Contrast aliases
export { default as ContrastIcon } from '../icons/contrast.js';
// Cookie aliases
export { default as CookieIcon } from '../icons/cookie.js';
// CookingPot aliases
export { default as CookingPotIcon } from '../icons/cooking-pot.js';
// CopyCheck aliases
export { default as CopyCheckIcon } from '../icons/copy-check.js';
// CopyMinus aliases
export { default as CopyMinusIcon } from '../icons/copy-minus.js';
// CopySlash aliases
export { default as CopySlashIcon } from '../icons/copy-slash.js';
// CopyPlus aliases
export { default as CopyPlusIcon } from '../icons/copy-plus.js';
// CopyX aliases
export { default as CopyXIcon } from '../icons/copy-x.js';
// Copy aliases
export { default as CopyIcon } from '../icons/copy.js';
// Copyleft aliases
export { default as CopyleftIcon } from '../icons/copyleft.js';
// Copyright aliases
export { default as CopyrightIcon } from '../icons/copyright.js';
// CornerDownRight aliases
export { default as CornerDownRightIcon } from '../icons/corner-down-right.js';
// CornerDownLeft aliases
export { default as CornerDownLeftIcon } from '../icons/corner-down-left.js';
// CornerLeftDown aliases
export { default as CornerLeftDownIcon } from '../icons/corner-left-down.js';
// CornerLeftUp aliases
export { default as CornerLeftUpIcon } from '../icons/corner-left-up.js';
// CornerRightDown aliases
export { default as CornerRightDownIcon } from '../icons/corner-right-down.js';
// CornerRightUp aliases
export { default as CornerRightUpIcon } from '../icons/corner-right-up.js';
// CornerUpLeft aliases
export { default as CornerUpLeftIcon } from '../icons/corner-up-left.js';
// Cpu aliases
export { default as CpuIcon } from '../icons/cpu.js';
// CornerUpRight aliases
export { default as CornerUpRightIcon } from '../icons/corner-up-right.js';
// CreativeCommons aliases
export { default as CreativeCommonsIcon } from '../icons/creative-commons.js';
// CreditCard aliases
export { default as CreditCardIcon } from '../icons/credit-card.js';
// Crop aliases
export { default as CropIcon } from '../icons/crop.js';
// Cross aliases
export { default as CrossIcon } from '../icons/cross.js';
// Crosshair aliases
export { default as CrosshairIcon } from '../icons/crosshair.js';
// Croissant aliases
export { default as CroissantIcon } from '../icons/croissant.js';
// Crown aliases
export { default as CrownIcon } from '../icons/crown.js';
// Cuboid aliases
export { default as CuboidIcon } from '../icons/cuboid.js';
// CupSoda aliases
export { default as CupSodaIcon } from '../icons/cup-soda.js';
// Cylinder aliases
export { default as CylinderIcon } from '../icons/cylinder.js';
// Currency aliases
export { default as CurrencyIcon } from '../icons/currency.js';
// Dam aliases
export { default as DamIcon } from '../icons/dam.js';
// DatabaseBackup aliases
export { default as DatabaseBackupIcon } from '../icons/database-backup.js';
// DatabaseZap aliases
export { default as DatabaseZapIcon } from '../icons/database-zap.js';
// Database aliases
export { default as DatabaseIcon } from '../icons/database.js';
// DecimalsArrowLeft aliases
export { default as DecimalsArrowLeftIcon } from '../icons/decimals-arrow-left.js';
// DecimalsArrowRight aliases
export { default as DecimalsArrowRightIcon } from '../icons/decimals-arrow-right.js';
// Delete aliases
export { default as DeleteIcon } from '../icons/delete.js';
// Dessert aliases
export { default as DessertIcon } from '../icons/dessert.js';
// Diameter aliases
export { default as DiameterIcon } from '../icons/diameter.js';
// DiamondMinus aliases
export { default as DiamondMinusIcon } from '../icons/diamond-minus.js';
// Diamond aliases
export { default as DiamondIcon } from '../icons/diamond.js';
// DiamondPlus aliases
export { default as DiamondPlusIcon } from '../icons/diamond-plus.js';
// Dice1 aliases
export { default as Dice1Icon } from '../icons/dice-1.js';
// Dice2 aliases
export { default as Dice2Icon } from '../icons/dice-2.js';
// Dice4 aliases
export { default as Dice4Icon } from '../icons/dice-4.js';
// Dice5 aliases
export { default as Dice5Icon } from '../icons/dice-5.js';
// Dice3 aliases
export { default as Dice3Icon } from '../icons/dice-3.js';
// Dice6 aliases
export { default as Dice6Icon } from '../icons/dice-6.js';
// Dices aliases
export { default as DicesIcon } from '../icons/dices.js';
// Diff aliases
export { default as DiffIcon } from '../icons/diff.js';
// Disc3 aliases
export { default as Disc3Icon } from '../icons/disc-3.js';
// Disc2 aliases
export { default as Disc2Icon } from '../icons/disc-2.js';
// DiscAlbum aliases
export { default as DiscAlbumIcon } from '../icons/disc-album.js';
// Disc aliases
export { default as DiscIcon } from '../icons/disc.js';
// Divide aliases
export { default as DivideIcon } from '../icons/divide.js';
// DnaOff aliases
export { default as DnaOffIcon } from '../icons/dna-off.js';
// Dna aliases
export { default as DnaIcon } from '../icons/dna.js';
// Dock aliases
export { default as DockIcon } from '../icons/dock.js';
// Dog aliases
export { default as DogIcon } from '../icons/dog.js';
// Donut aliases
export { default as DonutIcon } from '../icons/donut.js';
// DollarSign aliases
export { default as DollarSignIcon } from '../icons/dollar-sign.js';
// DoorClosedLocked aliases
export { default as DoorClosedLockedIcon } from '../icons/door-closed-locked.js';
// DoorClosed aliases
export { default as DoorClosedIcon } from '../icons/door-closed.js';
// DoorOpen aliases
export { default as DoorOpenIcon } from '../icons/door-open.js';
// Dot aliases
export { default as DotIcon } from '../icons/dot.js';
// Download aliases
export { default as DownloadIcon } from '../icons/download.js';
// DraftingCompass aliases
export { default as DraftingCompassIcon } from '../icons/drafting-compass.js';
// Drama aliases
export { default as DramaIcon } from '../icons/drama.js';
// Drill aliases
export { default as DrillIcon } from '../icons/drill.js';
// Dribbble aliases
export { default as DribbbleIcon } from '../icons/dribbble.js';
// Drone aliases
export { default as DroneIcon } from '../icons/drone.js';
// DropletOff aliases
export { default as DropletOffIcon } from '../icons/droplet-off.js';
// Droplets aliases
export { default as DropletsIcon } from '../icons/droplets.js';
// Droplet aliases
export { default as DropletIcon } from '../icons/droplet.js';
// Drum aliases
export { default as DrumIcon } from '../icons/drum.js';
// Drumstick aliases
export { default as DrumstickIcon } from '../icons/drumstick.js';
// Dumbbell aliases
export { default as DumbbellIcon } from '../icons/dumbbell.js';
// EarOff aliases
export { default as EarOffIcon } from '../icons/ear-off.js';
// Ear aliases
export { default as EarIcon } from '../icons/ear.js';
// EarthLock aliases
export { default as EarthLockIcon } from '../icons/earth-lock.js';
// Eclipse aliases
export { default as EclipseIcon } from '../icons/eclipse.js';
// EggFried aliases
export { default as EggFriedIcon } from '../icons/egg-fried.js';
// EggOff aliases
export { default as EggOffIcon } from '../icons/egg-off.js';
// EqualApproximately aliases
export { default as EqualApproximatelyIcon } from '../icons/equal-approximately.js';
// Egg aliases
export { default as EggIcon } from '../icons/egg.js';
// EqualNot aliases
export { default as EqualNotIcon } from '../icons/equal-not.js';
// Equal aliases
export { default as EqualIcon } from '../icons/equal.js';
// Eraser aliases
export { default as EraserIcon } from '../icons/eraser.js';
// EthernetPort aliases
export { default as EthernetPortIcon } from '../icons/ethernet-port.js';
// Euro aliases
export { default as EuroIcon } from '../icons/euro.js';
// Expand aliases
export { default as ExpandIcon } from '../icons/expand.js';
// ExternalLink aliases
export { default as ExternalLinkIcon } from '../icons/external-link.js';
// EyeClosed aliases
export { default as EyeClosedIcon } from '../icons/eye-closed.js';
// Eye aliases
export { default as EyeIcon } from '../icons/eye.js';
// Facebook aliases
export { default as FacebookIcon } from '../icons/facebook.js';
// Factory aliases
export { default as FactoryIcon } from '../icons/factory.js';
// EyeOff aliases
export { default as EyeOffIcon } from '../icons/eye-off.js';
// FastForward aliases
export { default as FastForwardIcon } from '../icons/fast-forward.js';
// Fan aliases
export { default as FanIcon } from '../icons/fan.js';
// Feather aliases
export { default as FeatherIcon } from '../icons/feather.js';
// Fence aliases
export { default as FenceIcon } from '../icons/fence.js';
// FerrisWheel aliases
export { default as FerrisWheelIcon } from '../icons/ferris-wheel.js';
// Figma aliases
export { default as FigmaIcon } from '../icons/figma.js';
// FileArchive aliases
export { default as FileArchiveIcon } from '../icons/file-archive.js';
// FileAudio2 aliases
export { default as FileAudio2Icon } from '../icons/file-audio-2.js';
// FileAudio aliases
export { default as FileAudioIcon } from '../icons/file-audio.js';
// FileBadge2 aliases
export { default as FileBadge2Icon } from '../icons/file-badge-2.js';
// FileBadge aliases
export { default as FileBadgeIcon } from '../icons/file-badge.js';
// FileBox aliases
export { default as FileBoxIcon } from '../icons/file-box.js';
// FileCheck2 aliases
export { default as FileCheck2Icon } from '../icons/file-check-2.js';
// FileCheck aliases
export { default as FileCheckIcon } from '../icons/file-check.js';
// FileClock aliases
export { default as FileClockIcon } from '../icons/file-clock.js';
// FileCode2 aliases
export { default as FileCode2Icon } from '../icons/file-code-2.js';
// FileCode aliases
export { default as FileCodeIcon } from '../icons/file-code.js';
// FileDigit aliases
export { default as FileDigitIcon } from '../icons/file-digit.js';
// FileDiff aliases
export { default as FileDiffIcon } from '../icons/file-diff.js';
// FileDown aliases
export { default as FileDownIcon } from '../icons/file-down.js';
// FileHeart aliases
export { default as FileHeartIcon } from '../icons/file-heart.js';
// FileImage aliases
export { default as FileImageIcon } from '../icons/file-image.js';
// FileInput aliases
export { default as FileInputIcon } from '../icons/file-input.js';
// FileJson2 aliases
export { default as FileJson2Icon } from '../icons/file-json-2.js';
// FileKey2 aliases
export { default as FileKey2Icon } from '../icons/file-key-2.js';
// FileJson aliases
export { default as FileJsonIcon } from '../icons/file-json.js';
// FileKey aliases
export { default as FileKeyIcon } from '../icons/file-key.js';
// FileLock2 aliases
export { default as FileLock2Icon } from '../icons/file-lock-2.js';
// FileLock aliases
export { default as FileLockIcon } from '../icons/file-lock.js';
// FileMinus2 aliases
export { default as FileMinus2Icon } from '../icons/file-minus-2.js';
// FileMinus aliases
export { default as FileMinusIcon } from '../icons/file-minus.js';
// FileMusic aliases
export { default as FileMusicIcon } from '../icons/file-music.js';
// FileOutput aliases
export { default as FileOutputIcon } from '../icons/file-output.js';
// FilePlus aliases
export { default as FilePlusIcon } from '../icons/file-plus.js';
// FileScan aliases
export { default as FileScanIcon } from '../icons/file-scan.js';
// FilePlus2 aliases
export { default as FilePlus2Icon } from '../icons/file-plus-2.js';
// FileSearch2 aliases
export { default as FileSearch2Icon } from '../icons/file-search-2.js';
// FileSearch aliases
export { default as FileSearchIcon } from '../icons/file-search.js';
// FileSliders aliases
export { default as FileSlidersIcon } from '../icons/file-sliders.js';
// FileStack aliases
export { default as FileStackIcon } from '../icons/file-stack.js';
// FileSpreadsheet aliases
export { default as FileSpreadsheetIcon } from '../icons/file-spreadsheet.js';
// FileSymlink aliases
export { default as FileSymlinkIcon } from '../icons/file-symlink.js';
// FileText aliases
export { default as FileTextIcon } from '../icons/file-text.js';
// FileTerminal aliases
export { default as FileTerminalIcon } from '../icons/file-terminal.js';
// FileType2 aliases
export { default as FileType2Icon } from '../icons/file-type-2.js';
// FileType aliases
export { default as FileTypeIcon } from '../icons/file-type.js';
// FileUp aliases
export { default as FileUpIcon } from '../icons/file-up.js';
// FileVolume2 aliases
export { default as FileVolume2Icon } from '../icons/file-volume-2.js';
// FileUser aliases
export { default as FileUserIcon } from '../icons/file-user.js';
// FileVolume aliases
export { default as FileVolumeIcon } from '../icons/file-volume.js';
// FileWarning aliases
export { default as FileWarningIcon } from '../icons/file-warning.js';
// FileX2 aliases
export { default as FileX2Icon } from '../icons/file-x-2.js';
// File aliases
export { default as FileIcon } from '../icons/file.js';
// FileX aliases
export { default as FileXIcon } from '../icons/file-x.js';
// Files aliases
export { default as FilesIcon } from '../icons/files.js';
// Film aliases
export { default as FilmIcon } from '../icons/film.js';
// Fingerprint aliases
export { default as FingerprintIcon } from '../icons/fingerprint.js';
// FireExtinguisher aliases
export { default as FireExtinguisherIcon } from '../icons/fire-extinguisher.js';
// FishSymbol aliases
export { default as FishSymbolIcon } from '../icons/fish-symbol.js';
// FishOff aliases
export { default as FishOffIcon } from '../icons/fish-off.js';
// Fish aliases
export { default as FishIcon } from '../icons/fish.js';
// FlagOff aliases
export { default as FlagOffIcon } from '../icons/flag-off.js';
// FlagTriangleLeft aliases
export { default as FlagTriangleLeftIcon } from '../icons/flag-triangle-left.js';
// Flag aliases
export { default as FlagIcon } from '../icons/flag.js';
// FlagTriangleRight aliases
export { default as FlagTriangleRightIcon } from '../icons/flag-triangle-right.js';
// Flame aliases
export { default as FlameIcon } from '../icons/flame.js';
// FlameKindling aliases
export { default as FlameKindlingIcon } from '../icons/flame-kindling.js';
// FlaskConicalOff aliases
export { default as FlaskConicalOffIcon } from '../icons/flask-conical-off.js';
// FlashlightOff aliases
export { default as FlashlightOffIcon } from '../icons/flashlight-off.js';
// Flashlight aliases
export { default as FlashlightIcon } from '../icons/flashlight.js';
// FlaskConical aliases
export { default as FlaskConicalIcon } from '../icons/flask-conical.js';
// FlaskRound aliases
export { default as FlaskRoundIcon } from '../icons/flask-round.js';
// FlipHorizontal2 aliases
export { default as FlipHorizontal2Icon } from '../icons/flip-horizontal-2.js';
// FlipHorizontal aliases
export { default as FlipHorizontalIcon } from '../icons/flip-horizontal.js';
// FlipVertical2 aliases
export { default as FlipVertical2Icon } from '../icons/flip-vertical-2.js';
// Flower2 aliases
export { default as Flower2Icon } from '../icons/flower-2.js';
// FlipVertical aliases
export { default as FlipVerticalIcon } from '../icons/flip-vertical.js';
// Flower aliases
export { default as FlowerIcon } from '../icons/flower.js';
// Focus aliases
export { default as FocusIcon } from '../icons/focus.js';
// FoldVertical aliases
export { default as FoldVerticalIcon } from '../icons/fold-vertical.js';
// FoldHorizontal aliases
export { default as FoldHorizontalIcon } from '../icons/fold-horizontal.js';
// FolderArchive aliases
export { default as FolderArchiveIcon } from '../icons/folder-archive.js';
// FolderClock aliases
export { default as FolderClockIcon } from '../icons/folder-clock.js';
// FolderCode aliases
export { default as FolderCodeIcon } from '../icons/folder-code.js';
// FolderCheck aliases
export { default as FolderCheckIcon } from '../icons/folder-check.js';
// FolderClosed aliases
export { default as FolderClosedIcon } from '../icons/folder-closed.js';
// FolderDot aliases
export { default as FolderDotIcon } from '../icons/folder-dot.js';
// FolderDown aliases
export { default as FolderDownIcon } from '../icons/folder-down.js';
// FolderGit2 aliases
export { default as FolderGit2Icon } from '../icons/folder-git-2.js';
// FolderGit aliases
export { default as FolderGitIcon } from '../icons/folder-git.js';
// FolderInput aliases
export { default as FolderInputIcon } from '../icons/folder-input.js';
// FolderHeart aliases
export { default as FolderHeartIcon } from '../icons/folder-heart.js';
// FolderKanban aliases
export { default as FolderKanbanIcon } from '../icons/folder-kanban.js';
// FolderKey aliases
export { default as FolderKeyIcon } from '../icons/folder-key.js';
// FolderLock aliases
export { default as FolderLockIcon } from '../icons/folder-lock.js';
// FolderMinus aliases
export { default as FolderMinusIcon } from '../icons/folder-minus.js';
// FolderOpenDot aliases
export { default as FolderOpenDotIcon } from '../icons/folder-open-dot.js';
// FolderOpen aliases
export { default as FolderOpenIcon } from '../icons/folder-open.js';
// FolderOutput aliases
export { default as FolderOutputIcon } from '../icons/folder-output.js';
// FolderPlus aliases
export { default as FolderPlusIcon } from '../icons/folder-plus.js';
// FolderRoot aliases
export { default as FolderRootIcon } from '../icons/folder-root.js';
// FolderSearch2 aliases
export { default as FolderSearch2Icon } from '../icons/folder-search-2.js';
// FolderSearch aliases
export { default as FolderSearchIcon } from '../icons/folder-search.js';
// FolderSync aliases
export { default as FolderSyncIcon } from '../icons/folder-sync.js';
// FolderSymlink aliases
export { default as FolderSymlinkIcon } from '../icons/folder-symlink.js';
// FolderUp aliases
export { default as FolderUpIcon } from '../icons/folder-up.js';
// FolderTree aliases
export { default as FolderTreeIcon } from '../icons/folder-tree.js';
// FolderX aliases
export { default as FolderXIcon } from '../icons/folder-x.js';
// Folder aliases
export { default as FolderIcon } from '../icons/folder.js';
// Folders aliases
export { default as FoldersIcon } from '../icons/folders.js';
// Footprints aliases
export { default as FootprintsIcon } from '../icons/footprints.js';
// Forklift aliases
export { default as ForkliftIcon } from '../icons/forklift.js';
// Forward aliases
export { default as ForwardIcon } from '../icons/forward.js';
// Frame aliases
export { default as FrameIcon } from '../icons/frame.js';
// Framer aliases
export { default as FramerIcon } from '../icons/framer.js';
// Frown aliases
export { default as FrownIcon } from '../icons/frown.js';
// Fuel aliases
export { default as FuelIcon } from '../icons/fuel.js';
// Fullscreen aliases
export { default as FullscreenIcon } from '../icons/fullscreen.js';
// FunnelPlus aliases
export { default as FunnelPlusIcon } from '../icons/funnel-plus.js';
// GalleryHorizontalEnd aliases
export { default as GalleryHorizontalEndIcon } from '../icons/gallery-horizontal-end.js';
// GalleryHorizontal aliases
export { default as GalleryHorizontalIcon } from '../icons/gallery-horizontal.js';
// GalleryThumbnails aliases
export { default as GalleryThumbnailsIcon } from '../icons/gallery-thumbnails.js';
// GalleryVerticalEnd aliases
export { default as GalleryVerticalEndIcon } from '../icons/gallery-vertical-end.js';
// GalleryVertical aliases
export { default as GalleryVerticalIcon } from '../icons/gallery-vertical.js';
// Gamepad2 aliases
export { default as Gamepad2Icon } from '../icons/gamepad-2.js';
// Gauge aliases
export { default as GaugeIcon } from '../icons/gauge.js';
// Gamepad aliases
export { default as GamepadIcon } from '../icons/gamepad.js';
// Gavel aliases
export { default as GavelIcon } from '../icons/gavel.js';
// Gem aliases
export { default as GemIcon } from '../icons/gem.js';
// GeorgianLari aliases
export { default as GeorgianLariIcon } from '../icons/georgian-lari.js';
// Ghost aliases
export { default as GhostIcon } from '../icons/ghost.js';
// GitBranchPlus aliases
export { default as GitBranchPlusIcon } from '../icons/git-branch-plus.js';
// Gift aliases
export { default as GiftIcon } from '../icons/gift.js';
// GitBranch aliases
export { default as GitBranchIcon } from '../icons/git-branch.js';
// GitCommitVertical aliases
export { default as GitCommitVerticalIcon } from '../icons/git-commit-vertical.js';
// GitCompareArrows aliases
export { default as GitCompareArrowsIcon } from '../icons/git-compare-arrows.js';
// GitCompare aliases
export { default as GitCompareIcon } from '../icons/git-compare.js';
// GitFork aliases
export { default as GitForkIcon } from '../icons/git-fork.js';
// GitMerge aliases
export { default as GitMergeIcon } from '../icons/git-merge.js';
// GitPullRequestArrow aliases
export { default as GitPullRequestArrowIcon } from '../icons/git-pull-request-arrow.js';
// GitGraph aliases
export { default as GitGraphIcon } from '../icons/git-graph.js';
// GitPullRequestClosed aliases
export { default as GitPullRequestClosedIcon } from '../icons/git-pull-request-closed.js';
// GitPullRequestCreateArrow aliases
export { default as GitPullRequestCreateArrowIcon } from '../icons/git-pull-request-create-arrow.js';
// GitPullRequestDraft aliases
export { default as GitPullRequestDraftIcon } from '../icons/git-pull-request-draft.js';
// GitPullRequestCreate aliases
export { default as GitPullRequestCreateIcon } from '../icons/git-pull-request-create.js';
// Github aliases
export { default as GithubIcon } from '../icons/github.js';
// GitPullRequest aliases
export { default as GitPullRequestIcon } from '../icons/git-pull-request.js';
// Gitlab aliases
export { default as GitlabIcon } from '../icons/gitlab.js';
// GlassWater aliases
export { default as GlassWaterIcon } from '../icons/glass-water.js';
// GlobeLock aliases
export { default as GlobeLockIcon } from '../icons/globe-lock.js';
// Glasses aliases
export { default as GlassesIcon } from '../icons/glasses.js';
// Globe aliases
export { default as GlobeIcon } from '../icons/globe.js';
// Goal aliases
export { default as GoalIcon } from '../icons/goal.js';
// Gpu aliases
export { default as GpuIcon } from '../icons/gpu.js';
// GraduationCap aliases
export { default as GraduationCapIcon } from '../icons/graduation-cap.js';
// Grab aliases
export { default as GrabIcon } from '../icons/grab.js';
// Grape aliases
export { default as GrapeIcon } from '../icons/grape.js';
// Grid3x2 aliases
export { default as Grid3x2Icon } from '../icons/grid-3x2.js';
// GripHorizontal aliases
export { default as GripHorizontalIcon } from '../icons/grip-horizontal.js';
// GripVertical aliases
export { default as GripVerticalIcon } from '../icons/grip-vertical.js';
// Grip aliases
export { default as GripIcon } from '../icons/grip.js';
// Group aliases
export { default as GroupIcon } from '../icons/group.js';
// Guitar aliases
export { default as GuitarIcon } from '../icons/guitar.js';
// Hamburger aliases
export { default as HamburgerIcon } from '../icons/hamburger.js';
// Ham aliases
export { default as HamIcon } from '../icons/ham.js';
// Hammer aliases
export { default as HammerIcon } from '../icons/hammer.js';
// HandCoins aliases
export { default as HandCoinsIcon } from '../icons/hand-coins.js';
// HandHeart aliases
export { default as HandHeartIcon } from '../icons/hand-heart.js';
// HandMetal aliases
export { default as HandMetalIcon } from '../icons/hand-metal.js';
// HandPlatter aliases
export { default as HandPlatterIcon } from '../icons/hand-platter.js';
// Hand aliases
export { default as HandIcon } from '../icons/hand.js';
// Handbag aliases
export { default as HandbagIcon } from '../icons/handbag.js';
// Handshake aliases
export { default as HandshakeIcon } from '../icons/handshake.js';
// HardDriveUpload aliases
export { default as HardDriveUploadIcon } from '../icons/hard-drive-upload.js';
// HardDriveDownload aliases
export { default as HardDriveDownloadIcon } from '../icons/hard-drive-download.js';
// HardDrive aliases
export { default as HardDriveIcon } from '../icons/hard-drive.js';
// HardHat aliases
export { default as HardHatIcon } from '../icons/hard-hat.js';
// Hash aliases
export { default as HashIcon } from '../icons/hash.js';
// HatGlasses aliases
export { default as HatGlassesIcon } from '../icons/hat-glasses.js';
// Haze aliases
export { default as HazeIcon } from '../icons/haze.js';
// HdmiPort aliases
export { default as HdmiPortIcon } from '../icons/hdmi-port.js';
// Heading1 aliases
export { default as Heading1Icon } from '../icons/heading-1.js';
// Heading2 aliases
export { default as Heading2Icon } from '../icons/heading-2.js';
// Heading3 aliases
export { default as Heading3Icon } from '../icons/heading-3.js';
// Heading4 aliases
export { default as Heading4Icon } from '../icons/heading-4.js';
// Heading5 aliases
export { default as Heading5Icon } from '../icons/heading-5.js';
// Heading6 aliases
export { default as Heading6Icon } from '../icons/heading-6.js';
// Heading aliases
export { default as HeadingIcon } from '../icons/heading.js';
// HeadphoneOff aliases
export { default as HeadphoneOffIcon } from '../icons/headphone-off.js';
// Headphones aliases
export { default as HeadphonesIcon } from '../icons/headphones.js';
// Headset aliases
export { default as HeadsetIcon } from '../icons/headset.js';
// HeartCrack aliases
export { default as HeartCrackIcon } from '../icons/heart-crack.js';
// HeartHandshake aliases
export { default as HeartHandshakeIcon } from '../icons/heart-handshake.js';
// HeartMinus aliases
export { default as HeartMinusIcon } from '../icons/heart-minus.js';
// HeartOff aliases
export { default as HeartOffIcon } from '../icons/heart-off.js';
// HeartPlus aliases
export { default as HeartPlusIcon } from '../icons/heart-plus.js';
// Heater aliases
export { default as HeaterIcon } from '../icons/heater.js';
// HeartPulse aliases
export { default as HeartPulseIcon } from '../icons/heart-pulse.js';
// Hexagon aliases
export { default as HexagonIcon } from '../icons/hexagon.js';
// Heart aliases
export { default as HeartIcon } from '../icons/heart.js';
// Highlighter aliases
export { default as HighlighterIcon } from '../icons/highlighter.js';
// History aliases
export { default as HistoryIcon } from '../icons/history.js';
// Hop aliases
export { default as HopIcon } from '../icons/hop.js';
// Hospital aliases
export { default as HospitalIcon } from '../icons/hospital.js';
// HopOff aliases
export { default as HopOffIcon } from '../icons/hop-off.js';
// Hotel aliases
export { default as HotelIcon } from '../icons/hotel.js';
// Hourglass aliases
export { default as HourglassIcon } from '../icons/hourglass.js';
// HousePlus aliases
export { default as HousePlusIcon } from '../icons/house-plus.js';
// HousePlug aliases
export { default as HousePlugIcon } from '../icons/house-plug.js';
// HouseWifi aliases
export { default as HouseWifiIcon } from '../icons/house-wifi.js';
// IdCardLanyard aliases
export { default as IdCardLanyardIcon } from '../icons/id-card-lanyard.js';
// IdCard aliases
export { default as IdCardIcon } from '../icons/id-card.js';
// ImageMinus aliases
export { default as ImageMinusIcon } from '../icons/image-minus.js';
// ImageDown aliases
export { default as ImageDownIcon } from '../icons/image-down.js';
// ImageOff aliases
export { default as ImageOffIcon } from '../icons/image-off.js';
// ImagePlus aliases
export { default as ImagePlusIcon } from '../icons/image-plus.js';
// ImagePlay aliases
export { default as ImagePlayIcon } from '../icons/image-play.js';
// ImageUp aliases
export { default as ImageUpIcon } from '../icons/image-up.js';
// ImageUpscale aliases
export { default as ImageUpscaleIcon } from '../icons/image-upscale.js';
// Image aliases
export { default as ImageIcon } from '../icons/image.js';
// Import aliases
export { default as ImportIcon } from '../icons/import.js';
// Images aliases
export { default as ImagesIcon } from '../icons/images.js';
// Inbox aliases
export { default as InboxIcon } from '../icons/inbox.js';
// IndianRupee aliases
export { default as IndianRupeeIcon } from '../icons/indian-rupee.js';
// Infinity aliases
export { default as InfinityIcon } from '../icons/infinity.js';
// Info aliases
export { default as InfoIcon } from '../icons/info.js';
// Instagram aliases
export { default as InstagramIcon } from '../icons/instagram.js';
// InspectionPanel aliases
export { default as InspectionPanelIcon } from '../icons/inspection-panel.js';
// IterationCcw aliases
export { default as IterationCcwIcon } from '../icons/iteration-ccw.js';
// Italic aliases
export { default as ItalicIcon } from '../icons/italic.js';
// IterationCw aliases
export { default as IterationCwIcon } from '../icons/iteration-cw.js';
// JapaneseYen aliases
export { default as JapaneseYenIcon } from '../icons/japanese-yen.js';
// Joystick aliases
export { default as JoystickIcon } from '../icons/joystick.js';
// Kanban aliases
export { default as KanbanIcon } from '../icons/kanban.js';
// KeySquare aliases
export { default as KeySquareIcon } from '../icons/key-square.js';
// KeyRound aliases
export { default as KeyRoundIcon } from '../icons/key-round.js';
// Key aliases
export { default as KeyIcon } from '../icons/key.js';
// KeyboardMusic aliases
export { default as KeyboardMusicIcon } from '../icons/keyboard-music.js';
// KeyboardOff aliases
export { default as KeyboardOffIcon } from '../icons/keyboard-off.js';
// LampCeiling aliases
export { default as LampCeilingIcon } from '../icons/lamp-ceiling.js';
// Keyboard aliases
export { default as KeyboardIcon } from '../icons/keyboard.js';
// LampDesk aliases
export { default as LampDeskIcon } from '../icons/lamp-desk.js';
// LampWallDown aliases
export { default as LampWallDownIcon } from '../icons/lamp-wall-down.js';
// LampFloor aliases
export { default as LampFloorIcon } from '../icons/lamp-floor.js';
// LampWallUp aliases
export { default as LampWallUpIcon } from '../icons/lamp-wall-up.js';
// Lamp aliases
export { default as LampIcon } from '../icons/lamp.js';
// Landmark aliases
export { default as LandmarkIcon } from '../icons/landmark.js';
// LandPlot aliases
export { default as LandPlotIcon } from '../icons/land-plot.js';
// Languages aliases
export { default as LanguagesIcon } from '../icons/languages.js';
// LaptopMinimalCheck aliases
export { default as LaptopMinimalCheckIcon } from '../icons/laptop-minimal-check.js';
// LassoSelect aliases
export { default as LassoSelectIcon } from '../icons/lasso-select.js';
// Laptop aliases
export { default as LaptopIcon } from '../icons/laptop.js';
// Lasso aliases
export { default as LassoIcon } from '../icons/lasso.js';
// Laugh aliases
export { default as LaughIcon } from '../icons/laugh.js';
// Layers2 aliases
export { default as Layers2Icon } from '../icons/layers-2.js';
// LayoutDashboard aliases
export { default as LayoutDashboardIcon } from '../icons/layout-dashboard.js';
// LayoutGrid aliases
export { default as LayoutGridIcon } from '../icons/layout-grid.js';
// LayoutList aliases
export { default as LayoutListIcon } from '../icons/layout-list.js';
// LayoutPanelTop aliases
export { default as LayoutPanelTopIcon } from '../icons/layout-panel-top.js';
// LayoutPanelLeft aliases
export { default as LayoutPanelLeftIcon } from '../icons/layout-panel-left.js';
// LayoutTemplate aliases
export { default as LayoutTemplateIcon } from '../icons/layout-template.js';
// Leaf aliases
export { default as LeafIcon } from '../icons/leaf.js';
// LeafyGreen aliases
export { default as LeafyGreenIcon } from '../icons/leafy-green.js';
// Lectern aliases
export { default as LecternIcon } from '../icons/lectern.js';
// LetterText aliases
export { default as LetterTextIcon } from '../icons/letter-text.js';
// Library aliases
export { default as LibraryIcon } from '../icons/library.js';
// LibraryBig aliases
export { default as LibraryBigIcon } from '../icons/library-big.js';
// Ligature aliases
export { default as LigatureIcon } from '../icons/ligature.js';
// LifeBuoy aliases
export { default as LifeBuoyIcon } from '../icons/life-buoy.js';
// LightbulbOff aliases
export { default as LightbulbOffIcon } from '../icons/lightbulb-off.js';
// LineSquiggle aliases
export { default as LineSquiggleIcon } from '../icons/line-squiggle.js';
// Lightbulb aliases
export { default as LightbulbIcon } from '../icons/lightbulb.js';
// Link2Off aliases
export { default as Link2OffIcon } from '../icons/link-2-off.js';
// Link2 aliases
export { default as Link2Icon } from '../icons/link-2.js';
// Link aliases
export { default as LinkIcon } from '../icons/link.js';
// Linkedin aliases
export { default as LinkedinIcon } from '../icons/linkedin.js';
// ListCheck aliases
export { default as ListCheckIcon } from '../icons/list-check.js';
// ListChecks aliases
export { default as ListChecksIcon } from '../icons/list-checks.js';
// ListEnd aliases
export { default as ListEndIcon } from '../icons/list-end.js';
// ListCollapse aliases
export { default as ListCollapseIcon } from '../icons/list-collapse.js';
// ListFilter aliases
export { default as ListFilterIcon } from '../icons/list-filter.js';
// ListFilterPlus aliases
export { default as ListFilterPlusIcon } from '../icons/list-filter-plus.js';
// ListMinus aliases
export { default as ListMinusIcon } from '../icons/list-minus.js';
// ListMusic aliases
export { default as ListMusicIcon } from '../icons/list-music.js';
// ListOrdered aliases
export { default as ListOrderedIcon } from '../icons/list-ordered.js';
// ListPlus aliases
export { default as ListPlusIcon } from '../icons/list-plus.js';
// ListRestart aliases
export { default as ListRestartIcon } from '../icons/list-restart.js';
// ListStart aliases
export { default as ListStartIcon } from '../icons/list-start.js';
// ListTree aliases
export { default as ListTreeIcon } from '../icons/list-tree.js';
// ListTodo aliases
export { default as ListTodoIcon } from '../icons/list-todo.js';
// ListVideo aliases
export { default as ListVideoIcon } from '../icons/list-video.js';
// List aliases
export { default as ListIcon } from '../icons/list.js';
// ListX aliases
export { default as ListXIcon } from '../icons/list-x.js';
// LoaderPinwheel aliases
export { default as LoaderPinwheelIcon } from '../icons/loader-pinwheel.js';
// Loader aliases
export { default as LoaderIcon } from '../icons/loader.js';
// LocateFixed aliases
export { default as LocateFixedIcon } from '../icons/locate-fixed.js';
// LocateOff aliases
export { default as LocateOffIcon } from '../icons/locate-off.js';
// Locate aliases
export { default as LocateIcon } from '../icons/locate.js';
// LockKeyhole aliases
export { default as LockKeyholeIcon } from '../icons/lock-keyhole.js';
// LogIn aliases
export { default as LogInIcon } from '../icons/log-in.js';
// Lock aliases
export { default as LockIcon } from '../icons/lock.js';
// LogOut aliases
export { default as LogOutIcon } from '../icons/log-out.js';
// Logs aliases
export { default as LogsIcon } from '../icons/logs.js';
// Lollipop aliases
export { default as LollipopIcon } from '../icons/lollipop.js';
// Luggage aliases
export { default as LuggageIcon } from '../icons/luggage.js';
// Magnet aliases
export { default as MagnetIcon } from '../icons/magnet.js';
// MailCheck aliases
export { default as MailCheckIcon } from '../icons/mail-check.js';
// MailMinus aliases
export { default as MailMinusIcon } from '../icons/mail-minus.js';
// MailOpen aliases
export { default as MailOpenIcon } from '../icons/mail-open.js';
// MailPlus aliases
export { default as MailPlusIcon } from '../icons/mail-plus.js';
// MailSearch aliases
export { default as MailSearchIcon } from '../icons/mail-search.js';
// MailWarning aliases
export { default as MailWarningIcon } from '../icons/mail-warning.js';
// Mail aliases
export { default as MailIcon } from '../icons/mail.js';
// MailX aliases
export { default as MailXIcon } from '../icons/mail-x.js';
// Mailbox aliases
export { default as MailboxIcon } from '../icons/mailbox.js';
// Mails aliases
export { default as MailsIcon } from '../icons/mails.js';
// MapMinus aliases
export { default as MapMinusIcon } from '../icons/map-minus.js';
// MapPinCheckInside aliases
export { default as MapPinCheckInsideIcon } from '../icons/map-pin-check-inside.js';
// MapPinCheck aliases
export { default as MapPinCheckIcon } from '../icons/map-pin-check.js';
// MapPinHouse aliases
export { default as MapPinHouseIcon } from '../icons/map-pin-house.js';
// MapPinMinusInside aliases
export { default as MapPinMinusInsideIcon } from '../icons/map-pin-minus-inside.js';
// MapPinMinus aliases
export { default as MapPinMinusIcon } from '../icons/map-pin-minus.js';
// MapPinOff aliases
export { default as MapPinOffIcon } from '../icons/map-pin-off.js';
// MapPinPlusInside aliases
export { default as MapPinPlusInsideIcon } from '../icons/map-pin-plus-inside.js';
// MapPinPlus aliases
export { default as MapPinPlusIcon } from '../icons/map-pin-plus.js';
// MapPinXInside aliases
export { default as MapPinXInsideIcon } from '../icons/map-pin-x-inside.js';
// MapPinX aliases
export { default as MapPinXIcon } from '../icons/map-pin-x.js';
// MapPin aliases
export { default as MapPinIcon } from '../icons/map-pin.js';
// MapPinned aliases
export { default as MapPinnedIcon } from '../icons/map-pinned.js';
// MapPlus aliases
export { default as MapPlusIcon } from '../icons/map-plus.js';
// Map aliases
export { default as MapIcon } from '../icons/map.js';
// MarsStroke aliases
export { default as MarsStrokeIcon } from '../icons/mars-stroke.js';
// Mars aliases
export { default as MarsIcon } from '../icons/mars.js';
// Maximize2 aliases
export { default as Maximize2Icon } from '../icons/maximize-2.js';
// Martini aliases
export { default as MartiniIcon } from '../icons/martini.js';
// Maximize aliases
export { default as MaximizeIcon } from '../icons/maximize.js';
// Medal aliases
export { default as MedalIcon } from '../icons/medal.js';
// Megaphone aliases
export { default as MegaphoneIcon } from '../icons/megaphone.js';
// MegaphoneOff aliases
export { default as MegaphoneOffIcon } from '../icons/megaphone-off.js';
// MemoryStick aliases
export { default as MemoryStickIcon } from '../icons/memory-stick.js';
// Menu aliases
export { default as MenuIcon } from '../icons/menu.js';
// Meh aliases
export { default as MehIcon } from '../icons/meh.js';
// MessageCircleCode aliases
export { default as MessageCircleCodeIcon } from '../icons/message-circle-code.js';
// MessageCircleDashed aliases
export { default as MessageCircleDashedIcon } from '../icons/message-circle-dashed.js';
// Merge aliases
export { default as MergeIcon } from '../icons/merge.js';
// MessageCircleHeart aliases
export { default as MessageCircleHeartIcon } from '../icons/message-circle-heart.js';
// MessageCircleMore aliases
export { default as MessageCircleMoreIcon } from '../icons/message-circle-more.js';
// MessageCircleReply aliases
export { default as MessageCircleReplyIcon } from '../icons/message-circle-reply.js';
// MessageCircleOff aliases
export { default as MessageCircleOffIcon } from '../icons/message-circle-off.js';
// MessageCircleWarning aliases
export { default as MessageCircleWarningIcon } from '../icons/message-circle-warning.js';
// MessageCirclePlus aliases
export { default as MessageCirclePlusIcon } from '../icons/message-circle-plus.js';
// MessageCircleX aliases
export { default as MessageCircleXIcon } from '../icons/message-circle-x.js';
// MessageCircle aliases
export { default as MessageCircleIcon } from '../icons/message-circle.js';
// MessageSquareCode aliases
export { default as MessageSquareCodeIcon } from '../icons/message-square-code.js';
// MessageSquareDashed aliases
export { default as MessageSquareDashedIcon } from '../icons/message-square-dashed.js';
// MessageSquareDiff aliases
export { default as MessageSquareDiffIcon } from '../icons/message-square-diff.js';
// MessageSquareDot aliases
export { default as MessageSquareDotIcon } from '../icons/message-square-dot.js';
// MessageSquareLock aliases
export { default as MessageSquareLockIcon } from '../icons/message-square-lock.js';
// MessageSquareHeart aliases
export { default as MessageSquareHeartIcon } from '../icons/message-square-heart.js';
// MessageSquareMore aliases
export { default as MessageSquareMoreIcon } from '../icons/message-square-more.js';
// MessageSquareOff aliases
export { default as MessageSquareOffIcon } from '../icons/message-square-off.js';
// MessageSquarePlus aliases
export { default as MessageSquarePlusIcon } from '../icons/message-square-plus.js';
// MessageSquareQuote aliases
export { default as MessageSquareQuoteIcon } from '../icons/message-square-quote.js';
// MessageSquareReply aliases
export { default as MessageSquareReplyIcon } from '../icons/message-square-reply.js';
// MessageSquareText aliases
export { default as MessageSquareTextIcon } from '../icons/message-square-text.js';
// MessageSquareShare aliases
export { default as MessageSquareShareIcon } from '../icons/message-square-share.js';
// MessageSquareWarning aliases
export { default as MessageSquareWarningIcon } from '../icons/message-square-warning.js';
// MessageSquareX aliases
export { default as MessageSquareXIcon } from '../icons/message-square-x.js';
// MessageSquare aliases
export { default as MessageSquareIcon } from '../icons/message-square.js';
// MessagesSquare aliases
export { default as MessagesSquareIcon } from '../icons/messages-square.js';
// MicOff aliases
export { default as MicOffIcon } from '../icons/mic-off.js';
// Mic aliases
export { default as MicIcon } from '../icons/mic.js';
// Microchip aliases
export { default as MicrochipIcon } from '../icons/microchip.js';
// Microscope aliases
export { default as MicroscopeIcon } from '../icons/microscope.js';
// Microwave aliases
export { default as MicrowaveIcon } from '../icons/microwave.js';
// Milestone aliases
export { default as MilestoneIcon } from '../icons/milestone.js';
// MilkOff aliases
export { default as MilkOffIcon } from '../icons/milk-off.js';
// Minimize2 aliases
export { default as Minimize2Icon } from '../icons/minimize-2.js';
// Milk aliases
export { default as MilkIcon } from '../icons/milk.js';
// Minimize aliases
export { default as MinimizeIcon } from '../icons/minimize.js';
// Minus aliases
export { default as MinusIcon } from '../icons/minus.js';
// MonitorCheck aliases
export { default as MonitorCheckIcon } from '../icons/monitor-check.js';
// MonitorCog aliases
export { default as MonitorCogIcon } from '../icons/monitor-cog.js';
// MonitorDot aliases
export { default as MonitorDotIcon } from '../icons/monitor-dot.js';
// MonitorOff aliases
export { default as MonitorOffIcon } from '../icons/monitor-off.js';
// MonitorPlay aliases
export { default as MonitorPlayIcon } from '../icons/monitor-play.js';
// MonitorDown aliases
export { default as MonitorDownIcon } from '../icons/monitor-down.js';
// MonitorSmartphone aliases
export { default as MonitorSmartphoneIcon } from '../icons/monitor-smartphone.js';
// MonitorSpeaker aliases
export { default as MonitorSpeakerIcon } from '../icons/monitor-speaker.js';
// MonitorStop aliases
export { default as MonitorStopIcon } from '../icons/monitor-stop.js';
// MonitorPause aliases
export { default as MonitorPauseIcon } from '../icons/monitor-pause.js';
// MonitorUp aliases
export { default as MonitorUpIcon } from '../icons/monitor-up.js';
// MonitorX aliases
export { default as MonitorXIcon } from '../icons/monitor-x.js';
// Monitor aliases
export { default as MonitorIcon } from '../icons/monitor.js';
// Moon aliases
export { default as MoonIcon } from '../icons/moon.js';
// MoonStar aliases
export { default as MoonStarIcon } from '../icons/moon-star.js';
// MountainSnow aliases
export { default as MountainSnowIcon } from '../icons/mountain-snow.js';
// Mountain aliases
export { default as MountainIcon } from '../icons/mountain.js';
// MouseOff aliases
export { default as MouseOffIcon } from '../icons/mouse-off.js';
// MousePointerBan aliases
export { default as MousePointerBanIcon } from '../icons/mouse-pointer-ban.js';
// MousePointer2 aliases
export { default as MousePointer2Icon } from '../icons/mouse-pointer-2.js';
// MousePointerClick aliases
export { default as MousePointerClickIcon } from '../icons/mouse-pointer-click.js';
// MousePointer aliases
export { default as MousePointerIcon } from '../icons/mouse-pointer.js';
// Mouse aliases
export { default as MouseIcon } from '../icons/mouse.js';
// MoveDiagonal aliases
export { default as MoveDiagonalIcon } from '../icons/move-diagonal.js';
// MoveDiagonal2 aliases
export { default as MoveDiagonal2Icon } from '../icons/move-diagonal-2.js';
// MoveDownLeft aliases
export { default as MoveDownLeftIcon } from '../icons/move-down-left.js';
// MoveDownRight aliases
export { default as MoveDownRightIcon } from '../icons/move-down-right.js';
// MoveHorizontal aliases
export { default as MoveHorizontalIcon } from '../icons/move-horizontal.js';
// MoveLeft aliases
export { default as MoveLeftIcon } from '../icons/move-left.js';
// MoveRight aliases
export { default as MoveRightIcon } from '../icons/move-right.js';
// MoveDown aliases
export { default as MoveDownIcon } from '../icons/move-down.js';
// MoveUpLeft aliases
export { default as MoveUpLeftIcon } from '../icons/move-up-left.js';
// MoveUpRight aliases
export { default as MoveUpRightIcon } from '../icons/move-up-right.js';
// MoveUp aliases
export { default as MoveUpIcon } from '../icons/move-up.js';
// MoveVertical aliases
export { default as MoveVerticalIcon } from '../icons/move-vertical.js';
// Move aliases
export { default as MoveIcon } from '../icons/move.js';
// Music2 aliases
export { default as Music2Icon } from '../icons/music-2.js';
// Music4 aliases
export { default as Music4Icon } from '../icons/music-4.js';
// Music3 aliases
export { default as Music3Icon } from '../icons/music-3.js';
// Music aliases
export { default as MusicIcon } from '../icons/music.js';
// Navigation2Off aliases
export { default as Navigation2OffIcon } from '../icons/navigation-2-off.js';
// NavigationOff aliases
export { default as NavigationOffIcon } from '../icons/navigation-off.js';
// Navigation2 aliases
export { default as Navigation2Icon } from '../icons/navigation-2.js';
// Navigation aliases
export { default as NavigationIcon } from '../icons/navigation.js';
// Network aliases
export { default as NetworkIcon } from '../icons/network.js';
// Newspaper aliases
export { default as NewspaperIcon } from '../icons/newspaper.js';
// NonBinary aliases
export { default as NonBinaryIcon } from '../icons/non-binary.js';
// Nfc aliases
export { default as NfcIcon } from '../icons/nfc.js';
// NotebookPen aliases
export { default as NotebookPenIcon } from '../icons/notebook-pen.js';
// NotebookTabs aliases
export { default as NotebookTabsIcon } from '../icons/notebook-tabs.js';
// NotebookText aliases
export { default as NotebookTextIcon } from '../icons/notebook-text.js';
// Notebook aliases
export { default as NotebookIcon } from '../icons/notebook.js';
// NotepadText aliases
export { default as NotepadTextIcon } from '../icons/notepad-text.js';
// Nut aliases
export { default as NutIcon } from '../icons/nut.js';
// NotepadTextDashed aliases
export { default as NotepadTextDashedIcon } from '../icons/notepad-text-dashed.js';
// NutOff aliases
export { default as NutOffIcon } from '../icons/nut-off.js';
// OctagonMinus aliases
export { default as OctagonMinusIcon } from '../icons/octagon-minus.js';
// Octagon aliases
export { default as OctagonIcon } from '../icons/octagon.js';
// Omega aliases
export { default as OmegaIcon } from '../icons/omega.js';
// Option aliases
export { default as OptionIcon } from '../icons/option.js';
// Orbit aliases
export { default as OrbitIcon } from '../icons/orbit.js';
// Origami aliases
export { default as OrigamiIcon } from '../icons/origami.js';
// Package2 aliases
export { default as Package2Icon } from '../icons/package-2.js';
// PackageMinus aliases
export { default as PackageMinusIcon } from '../icons/package-minus.js';
// PackageOpen aliases
export { default as PackageOpenIcon } from '../icons/package-open.js';
// PackageCheck aliases
export { default as PackageCheckIcon } from '../icons/package-check.js';
// PackagePlus aliases
export { default as PackagePlusIcon } from '../icons/package-plus.js';
// PackageSearch aliases
export { default as PackageSearchIcon } from '../icons/package-search.js';
// Package aliases
export { default as PackageIcon } from '../icons/package.js';
// PackageX aliases
export { default as PackageXIcon } from '../icons/package-x.js';
// PaintBucket aliases
export { default as PaintBucketIcon } from '../icons/paint-bucket.js';
// PaintRoller aliases
export { default as PaintRollerIcon } from '../icons/paint-roller.js';
// Paintbrush aliases
export { default as PaintbrushIcon } from '../icons/paintbrush.js';
// Palette aliases
export { default as PaletteIcon } from '../icons/palette.js';
// Panda aliases
export { default as PandaIcon } from '../icons/panda.js';
// PanelBottomClose aliases
export { default as PanelBottomCloseIcon } from '../icons/panel-bottom-close.js';
// PanelBottomOpen aliases
export { default as PanelBottomOpenIcon } from '../icons/panel-bottom-open.js';
// PanelBottom aliases
export { default as PanelBottomIcon } from '../icons/panel-bottom.js';
// PanelRightClose aliases
export { default as PanelRightCloseIcon } from '../icons/panel-right-close.js';
// PanelRightOpen aliases
export { default as PanelRightOpenIcon } from '../icons/panel-right-open.js';
// PanelRight aliases
export { default as PanelRightIcon } from '../icons/panel-right.js';
// PanelTopClose aliases
export { default as PanelTopCloseIcon } from '../icons/panel-top-close.js';
// PanelTopOpen aliases
export { default as PanelTopOpenIcon } from '../icons/panel-top-open.js';
// PanelTop aliases
export { default as PanelTopIcon } from '../icons/panel-top.js';
// PanelsRightBottom aliases
export { default as PanelsRightBottomIcon } from '../icons/panels-right-bottom.js';
// PanelsLeftBottom aliases
export { default as PanelsLeftBottomIcon } from '../icons/panels-left-bottom.js';
// Paperclip aliases
export { default as PaperclipIcon } from '../icons/paperclip.js';
// Parentheses aliases
export { default as ParenthesesIcon } from '../icons/parentheses.js';
// ParkingMeter aliases
export { default as ParkingMeterIcon } from '../icons/parking-meter.js';
// PartyPopper aliases
export { default as PartyPopperIcon } from '../icons/party-popper.js';
// Pause aliases
export { default as PauseIcon } from '../icons/pause.js';
// PawPrint aliases
export { default as PawPrintIcon } from '../icons/paw-print.js';
// PcCase aliases
export { default as PcCaseIcon } from '../icons/pc-case.js';
// PenOff aliases
export { default as PenOffIcon } from '../icons/pen-off.js';
// PenTool aliases
export { default as PenToolIcon } from '../icons/pen-tool.js';
// PencilLine aliases
export { default as PencilLineIcon } from '../icons/pencil-line.js';
// PencilOff aliases
export { default as PencilOffIcon } from '../icons/pencil-off.js';
// PencilRuler aliases
export { default as PencilRulerIcon } from '../icons/pencil-ruler.js';
// Pencil aliases
export { default as PencilIcon } from '../icons/pencil.js';
// Percent aliases
export { default as PercentIcon } from '../icons/percent.js';
// Pentagon aliases
export { default as PentagonIcon } from '../icons/pentagon.js';
// PersonStanding aliases
export { default as PersonStandingIcon } from '../icons/person-standing.js';
// PhilippinePeso aliases
export { default as PhilippinePesoIcon } from '../icons/philippine-peso.js';
// PhoneCall aliases
export { default as PhoneCallIcon } from '../icons/phone-call.js';
// PhoneForwarded aliases
export { default as PhoneForwardedIcon } from '../icons/phone-forwarded.js';
// PhoneIncoming aliases
export { default as PhoneIncomingIcon } from '../icons/phone-incoming.js';
// PhoneMissed aliases
export { default as PhoneMissedIcon } from '../icons/phone-missed.js';
// PhoneOff aliases
export { default as PhoneOffIcon } from '../icons/phone-off.js';
// PhoneOutgoing aliases
export { default as PhoneOutgoingIcon } from '../icons/phone-outgoing.js';
// Phone aliases
export { default as PhoneIcon } from '../icons/phone.js';
// Pi aliases
export { default as PiIcon } from '../icons/pi.js';
// Piano aliases
export { default as PianoIcon } from '../icons/piano.js';
// Pickaxe aliases
export { default as PickaxeIcon } from '../icons/pickaxe.js';
// PictureInPicture aliases
export { default as PictureInPictureIcon } from '../icons/picture-in-picture.js';
// PictureInPicture2 aliases
export { default as PictureInPicture2Icon } from '../icons/picture-in-picture-2.js';
// PiggyBank aliases
export { default as PiggyBankIcon } from '../icons/piggy-bank.js';
// PilcrowLeft aliases
export { default as PilcrowLeftIcon } from '../icons/pilcrow-left.js';
// PilcrowRight aliases
export { default as PilcrowRightIcon } from '../icons/pilcrow-right.js';
// Pilcrow aliases
export { default as PilcrowIcon } from '../icons/pilcrow.js';
// PillBottle aliases
export { default as PillBottleIcon } from '../icons/pill-bottle.js';
// PinOff aliases
export { default as PinOffIcon } from '../icons/pin-off.js';
// Pill aliases
export { default as PillIcon } from '../icons/pill.js';
// Pipette aliases
export { default as PipetteIcon } from '../icons/pipette.js';
// Pin aliases
export { default as PinIcon } from '../icons/pin.js';
// Pizza aliases
export { default as PizzaIcon } from '../icons/pizza.js';
// PlaneLanding aliases
export { default as PlaneLandingIcon } from '../icons/plane-landing.js';
// PlaneTakeoff aliases
export { default as PlaneTakeoffIcon } from '../icons/plane-takeoff.js';
// Play aliases
export { default as PlayIcon } from '../icons/play.js';
// Plane aliases
export { default as PlaneIcon } from '../icons/plane.js';
// Plug2 aliases
export { default as Plug2Icon } from '../icons/plug-2.js';
// Plus aliases
export { default as PlusIcon } from '../icons/plus.js';
// Plug aliases
export { default as PlugIcon } from '../icons/plug.js';
// PocketKnife aliases
export { default as PocketKnifeIcon } from '../icons/pocket-knife.js';
// Pocket aliases
export { default as PocketIcon } from '../icons/pocket.js';
// Podcast aliases
export { default as PodcastIcon } from '../icons/podcast.js';
// PointerOff aliases
export { default as PointerOffIcon } from '../icons/pointer-off.js';
// Pointer aliases
export { default as PointerIcon } from '../icons/pointer.js';
// Popcorn aliases
export { default as PopcornIcon } from '../icons/popcorn.js';
// Popsicle aliases
export { default as PopsicleIcon } from '../icons/popsicle.js';
// PoundSterling aliases
export { default as PoundSterlingIcon } from '../icons/pound-sterling.js';
// PowerOff aliases
export { default as PowerOffIcon } from '../icons/power-off.js';
// Power aliases
export { default as PowerIcon } from '../icons/power.js';
// Presentation aliases
export { default as PresentationIcon } from '../icons/presentation.js';
// PrinterCheck aliases
export { default as PrinterCheckIcon } from '../icons/printer-check.js';
// Printer aliases
export { default as PrinterIcon } from '../icons/printer.js';
// Projector aliases
export { default as ProjectorIcon } from '../icons/projector.js';
// Proportions aliases
export { default as ProportionsIcon } from '../icons/proportions.js';
// Puzzle aliases
export { default as PuzzleIcon } from '../icons/puzzle.js';
// Pyramid aliases
export { default as PyramidIcon } from '../icons/pyramid.js';
// QrCode aliases
export { default as QrCodeIcon } from '../icons/qr-code.js';
// Quote aliases
export { default as QuoteIcon } from '../icons/quote.js';
// Rabbit aliases
export { default as RabbitIcon } from '../icons/rabbit.js';
// Radiation aliases
export { default as RadiationIcon } from '../icons/radiation.js';
// Radar aliases
export { default as RadarIcon } from '../icons/radar.js';
// RadioReceiver aliases
export { default as RadioReceiverIcon } from '../icons/radio-receiver.js';
// Radical aliases
export { default as RadicalIcon } from '../icons/radical.js';
// RadioTower aliases
export { default as RadioTowerIcon } from '../icons/radio-tower.js';
// Radius aliases
export { default as RadiusIcon } from '../icons/radius.js';
// Radio aliases
export { default as RadioIcon } from '../icons/radio.js';
// RailSymbol aliases
export { default as RailSymbolIcon } from '../icons/rail-symbol.js';
// Rainbow aliases
export { default as RainbowIcon } from '../icons/rainbow.js';
// Rat aliases
export { default as RatIcon } from '../icons/rat.js';
// Ratio aliases
export { default as RatioIcon } from '../icons/ratio.js';
// ReceiptCent aliases
export { default as ReceiptCentIcon } from '../icons/receipt-cent.js';
// ReceiptEuro aliases
export { default as ReceiptEuroIcon } from '../icons/receipt-euro.js';
// ReceiptJapaneseYen aliases
export { default as ReceiptJapaneseYenIcon } from '../icons/receipt-japanese-yen.js';
// ReceiptIndianRupee aliases
export { default as ReceiptIndianRupeeIcon } from '../icons/receipt-indian-rupee.js';
// ReceiptPoundSterling aliases
export { default as ReceiptPoundSterlingIcon } from '../icons/receipt-pound-sterling.js';
// ReceiptRussianRuble aliases
export { default as ReceiptRussianRubleIcon } from '../icons/receipt-russian-ruble.js';
// ReceiptSwissFranc aliases
export { default as ReceiptSwissFrancIcon } from '../icons/receipt-swiss-franc.js';
// ReceiptText aliases
export { default as ReceiptTextIcon } from '../icons/receipt-text.js';
// ReceiptTurkishLira aliases
export { default as ReceiptTurkishLiraIcon } from '../icons/receipt-turkish-lira.js';
// Receipt aliases
export { default as ReceiptIcon } from '../icons/receipt.js';
// RectangleCircle aliases
export { default as RectangleCircleIcon } from '../icons/rectangle-circle.js';
// RectangleGoggles aliases
export { default as RectangleGogglesIcon } from '../icons/rectangle-goggles.js';
// RectangleHorizontal aliases
export { default as RectangleHorizontalIcon } from '../icons/rectangle-horizontal.js';
// Recycle aliases
export { default as RecycleIcon } from '../icons/recycle.js';
// RectangleVertical aliases
export { default as RectangleVerticalIcon } from '../icons/rectangle-vertical.js';
// Redo2 aliases
export { default as Redo2Icon } from '../icons/redo-2.js';
// RedoDot aliases
export { default as RedoDotIcon } from '../icons/redo-dot.js';
// Redo aliases
export { default as RedoIcon } from '../icons/redo.js';
// RefreshCcwDot aliases
export { default as RefreshCcwDotIcon } from '../icons/refresh-ccw-dot.js';
// RefreshCcw aliases
export { default as RefreshCcwIcon } from '../icons/refresh-ccw.js';
// RefreshCw aliases
export { default as RefreshCwIcon } from '../icons/refresh-cw.js';
// Refrigerator aliases
export { default as RefrigeratorIcon } from '../icons/refrigerator.js';
// RefreshCwOff aliases
export { default as RefreshCwOffIcon } from '../icons/refresh-cw-off.js';
// Regex aliases
export { default as RegexIcon } from '../icons/regex.js';
// RemoveFormatting aliases
export { default as RemoveFormattingIcon } from '../icons/remove-formatting.js';
// Repeat2 aliases
export { default as Repeat2Icon } from '../icons/repeat-2.js';
// Repeat1 aliases
export { default as Repeat1Icon } from '../icons/repeat-1.js';
// Repeat aliases
export { default as RepeatIcon } from '../icons/repeat.js';
// ReplaceAll aliases
export { default as ReplaceAllIcon } from '../icons/replace-all.js';
// Replace aliases
export { default as ReplaceIcon } from '../icons/replace.js';
// ReplyAll aliases
export { default as ReplyAllIcon } from '../icons/reply-all.js';
// Reply aliases
export { default as ReplyIcon } from '../icons/reply.js';
// Ribbon aliases
export { default as RibbonIcon } from '../icons/ribbon.js';
// Rocket aliases
export { default as RocketIcon } from '../icons/rocket.js';
// Rewind aliases
export { default as RewindIcon } from '../icons/rewind.js';
// RockingChair aliases
export { default as RockingChairIcon } from '../icons/rocking-chair.js';
// RotateCcwKey aliases
export { default as RotateCcwKeyIcon } from '../icons/rotate-ccw-key.js';
// RollerCoaster aliases
export { default as RollerCoasterIcon } from '../icons/roller-coaster.js';
// RotateCcwSquare aliases
export { default as RotateCcwSquareIcon } from '../icons/rotate-ccw-square.js';
// RotateCcw aliases
export { default as RotateCcwIcon } from '../icons/rotate-ccw.js';
// RotateCwSquare aliases
export { default as RotateCwSquareIcon } from '../icons/rotate-cw-square.js';
// RotateCw aliases
export { default as RotateCwIcon } from '../icons/rotate-cw.js';
// RouteOff aliases
export { default as RouteOffIcon } from '../icons/route-off.js';
// Route aliases
export { default as RouteIcon } from '../icons/route.js';
// Router aliases
export { default as RouterIcon } from '../icons/router.js';
// Rows4 aliases
export { default as Rows4Icon } from '../icons/rows-4.js';
// Rss aliases
export { default as RssIcon } from '../icons/rss.js';
// RulerDimensionLine aliases
export { default as RulerDimensionLineIcon } from '../icons/ruler-dimension-line.js';
// Ruler aliases
export { default as RulerIcon } from '../icons/ruler.js';
// RussianRuble aliases
export { default as RussianRubleIcon } from '../icons/russian-ruble.js';
// Salad aliases
export { default as SaladIcon } from '../icons/salad.js';
// Sandwich aliases
export { default as SandwichIcon } from '../icons/sandwich.js';
// SatelliteDish aliases
export { default as SatelliteDishIcon } from '../icons/satellite-dish.js';
// Sailboat aliases
export { default as SailboatIcon } from '../icons/sailboat.js';
// Satellite aliases
export { default as SatelliteIcon } from '../icons/satellite.js';
// SaudiRiyal aliases
export { default as SaudiRiyalIcon } from '../icons/saudi-riyal.js';
// SaveAll aliases
export { default as SaveAllIcon } from '../icons/save-all.js';
// SaveOff aliases
export { default as SaveOffIcon } from '../icons/save-off.js';
// Save aliases
export { default as SaveIcon } from '../icons/save.js';
// Scale aliases
export { default as ScaleIcon } from '../icons/scale.js';
// Scaling aliases
export { default as ScalingIcon } from '../icons/scaling.js';
// ScanBarcode aliases
export { default as ScanBarcodeIcon } from '../icons/scan-barcode.js';
// ScanFace aliases
export { default as ScanFaceIcon } from '../icons/scan-face.js';
// ScanEye aliases
export { default as ScanEyeIcon } from '../icons/scan-eye.js';
// ScanLine aliases
export { default as ScanLineIcon } from '../icons/scan-line.js';
// ScanHeart aliases
export { default as ScanHeartIcon } from '../icons/scan-heart.js';
// ScanQrCode aliases
export { default as ScanQrCodeIcon } from '../icons/scan-qr-code.js';
// ScanSearch aliases
export { default as ScanSearchIcon } from '../icons/scan-search.js';
// Scan aliases
export { default as ScanIcon } from '../icons/scan.js';
// ScanText aliases
export { default as ScanTextIcon } from '../icons/scan-text.js';
// School aliases
export { default as SchoolIcon } from '../icons/school.js';
// ScissorsLineDashed aliases
export { default as ScissorsLineDashedIcon } from '../icons/scissors-line-dashed.js';
// Scissors aliases
export { default as ScissorsIcon } from '../icons/scissors.js';
// ScreenShareOff aliases
export { default as ScreenShareOffIcon } from '../icons/screen-share-off.js';
// ScreenShare aliases
export { default as ScreenShareIcon } from '../icons/screen-share.js';
// Scroll aliases
export { default as ScrollIcon } from '../icons/scroll.js';
// ScrollText aliases
export { default as ScrollTextIcon } from '../icons/scroll-text.js';
// SearchCheck aliases
export { default as SearchCheckIcon } from '../icons/search-check.js';
// SearchCode aliases
export { default as SearchCodeIcon } from '../icons/search-code.js';
// SearchSlash aliases
export { default as SearchSlashIcon } from '../icons/search-slash.js';
// SearchX aliases
export { default as SearchXIcon } from '../icons/search-x.js';
// Search aliases
export { default as SearchIcon } from '../icons/search.js';
// SendToBack aliases
export { default as SendToBackIcon } from '../icons/send-to-back.js';
// Section aliases
export { default as SectionIcon } from '../icons/section.js';
// Send aliases
export { default as SendIcon } from '../icons/send.js';
// SeparatorHorizontal aliases
export { default as SeparatorHorizontalIcon } from '../icons/separator-horizontal.js';
// SeparatorVertical aliases
export { default as SeparatorVerticalIcon } from '../icons/separator-vertical.js';
// ServerCrash aliases
export { default as ServerCrashIcon } from '../icons/server-crash.js';
// ServerOff aliases
export { default as ServerOffIcon } from '../icons/server-off.js';
// ServerCog aliases
export { default as ServerCogIcon } from '../icons/server-cog.js';
// Server aliases
export { default as ServerIcon } from '../icons/server.js';
// Settings2 aliases
export { default as Settings2Icon } from '../icons/settings-2.js';
// Settings aliases
export { default as SettingsIcon } from '../icons/settings.js';
// Shapes aliases
export { default as ShapesIcon } from '../icons/shapes.js';
// Share2 aliases
export { default as Share2Icon } from '../icons/share-2.js';
// Share aliases
export { default as ShareIcon } from '../icons/share.js';
// Sheet aliases
export { default as SheetIcon } from '../icons/sheet.js';
// Shell aliases
export { default as ShellIcon } from '../icons/shell.js';
// ShieldAlert aliases
export { default as ShieldAlertIcon } from '../icons/shield-alert.js';
// ShieldBan aliases
export { default as ShieldBanIcon } from '../icons/shield-ban.js';
// ShieldCheck aliases
export { default as ShieldCheckIcon } from '../icons/shield-check.js';
// ShieldEllipsis aliases
export { default as ShieldEllipsisIcon } from '../icons/shield-ellipsis.js';
// ShieldHalf aliases
export { default as ShieldHalfIcon } from '../icons/shield-half.js';
// ShieldMinus aliases
export { default as ShieldMinusIcon } from '../icons/shield-minus.js';
// ShieldOff aliases
export { default as ShieldOffIcon } from '../icons/shield-off.js';
// ShieldPlus aliases
export { default as ShieldPlusIcon } from '../icons/shield-plus.js';
// ShieldUser aliases
export { default as ShieldUserIcon } from '../icons/shield-user.js';
// Shield aliases
export { default as ShieldIcon } from '../icons/shield.js';
// ShipWheel aliases
export { default as ShipWheelIcon } from '../icons/ship-wheel.js';
// Ship aliases
export { default as ShipIcon } from '../icons/ship.js';
// Shirt aliases
export { default as ShirtIcon } from '../icons/shirt.js';
// ShoppingBag aliases
export { default as ShoppingBagIcon } from '../icons/shopping-bag.js';
// ShoppingBasket aliases
export { default as ShoppingBasketIcon } from '../icons/shopping-basket.js';
// ShoppingCart aliases
export { default as ShoppingCartIcon } from '../icons/shopping-cart.js';
// Shovel aliases
export { default as ShovelIcon } from '../icons/shovel.js';
// ShowerHead aliases
export { default as ShowerHeadIcon } from '../icons/shower-head.js';
// Shredder aliases
export { default as ShredderIcon } from '../icons/shredder.js';
// Shrimp aliases
export { default as ShrimpIcon } from '../icons/shrimp.js';
// Shrink aliases
export { default as ShrinkIcon } from '../icons/shrink.js';
// Shrub aliases
export { default as ShrubIcon } from '../icons/shrub.js';
// Sigma aliases
export { default as SigmaIcon } from '../icons/sigma.js';
// SignalHigh aliases
export { default as SignalHighIcon } from '../icons/signal-high.js';
// Shuffle aliases
export { default as ShuffleIcon } from '../icons/shuffle.js';
// SignalLow aliases
export { default as SignalLowIcon } from '../icons/signal-low.js';
// SignalMedium aliases
export { default as SignalMediumIcon } from '../icons/signal-medium.js';
// SignalZero aliases
export { default as SignalZeroIcon } from '../icons/signal-zero.js';
// Signal aliases
export { default as SignalIcon } from '../icons/signal.js';
// Signature aliases
export { default as SignatureIcon } from '../icons/signature.js';
// Signpost aliases
export { default as SignpostIcon } from '../icons/signpost.js';
// SignpostBig aliases
export { default as SignpostBigIcon } from '../icons/signpost-big.js';
// Siren aliases
export { default as SirenIcon } from '../icons/siren.js';
// SkipBack aliases
export { default as SkipBackIcon } from '../icons/skip-back.js';
// SkipForward aliases
export { default as SkipForwardIcon } from '../icons/skip-forward.js';
// Slack aliases
export { default as SlackIcon } from '../icons/slack.js';
// Skull aliases
export { default as SkullIcon } from '../icons/skull.js';
// Slash aliases
export { default as SlashIcon } from '../icons/slash.js';
// Slice aliases
export { default as SliceIcon } from '../icons/slice.js';
// SlidersHorizontal aliases
export { default as SlidersHorizontalIcon } from '../icons/sliders-horizontal.js';
// SmartphoneCharging aliases
export { default as SmartphoneChargingIcon } from '../icons/smartphone-charging.js';
// SmartphoneNfc aliases
export { default as SmartphoneNfcIcon } from '../icons/smartphone-nfc.js';
// Smartphone aliases
export { default as SmartphoneIcon } from '../icons/smartphone.js';
// SmilePlus aliases
export { default as SmilePlusIcon } from '../icons/smile-plus.js';
// Smile aliases
export { default as SmileIcon } from '../icons/smile.js';
// Snail aliases
export { default as SnailIcon } from '../icons/snail.js';
// SoapDispenserDroplet aliases
export { default as SoapDispenserDropletIcon } from '../icons/soap-dispenser-droplet.js';
// Snowflake aliases
export { default as SnowflakeIcon } from '../icons/snowflake.js';
// Soup aliases
export { default as SoupIcon } from '../icons/soup.js';
// Sofa aliases
export { default as SofaIcon } from '../icons/sofa.js';
// Space aliases
export { default as SpaceIcon } from '../icons/space.js';
// Spade aliases
export { default as SpadeIcon } from '../icons/spade.js';
// Speaker aliases
export { default as SpeakerIcon } from '../icons/speaker.js';
// Sparkle aliases
export { default as SparkleIcon } from '../icons/sparkle.js';
// Speech aliases
export { default as SpeechIcon } from '../icons/speech.js';
// SpellCheck2 aliases
export { default as SpellCheck2Icon } from '../icons/spell-check-2.js';
// SpellCheck aliases
export { default as SpellCheckIcon } from '../icons/spell-check.js';
// SplinePointer aliases
export { default as SplinePointerIcon } from '../icons/spline-pointer.js';
// Spline aliases
export { default as SplineIcon } from '../icons/spline.js';
// Split aliases
export { default as SplitIcon } from '../icons/split.js';
// Spotlight aliases
export { default as SpotlightIcon } from '../icons/spotlight.js';
// Spool aliases
export { default as SpoolIcon } from '../icons/spool.js';
// SprayCan aliases
export { default as SprayCanIcon } from '../icons/spray-can.js';
// Sprout aliases
export { default as SproutIcon } from '../icons/sprout.js';
// SquareDashedBottomCode aliases
export { default as SquareDashedBottomCodeIcon } from '../icons/square-dashed-bottom-code.js';
// SquareDashedBottom aliases
export { default as SquareDashedBottomIcon } from '../icons/square-dashed-bottom.js';
// SquareDashedTopSolid aliases
export { default as SquareDashedTopSolidIcon } from '../icons/square-dashed-top-solid.js';
// SquarePause aliases
export { default as SquarePauseIcon } from '../icons/square-pause.js';
// SquareRadical aliases
export { default as SquareRadicalIcon } from '../icons/square-radical.js';
// SquareRoundCorner aliases
export { default as SquareRoundCornerIcon } from '../icons/square-round-corner.js';
// SquareSquare aliases
export { default as SquareSquareIcon } from '../icons/square-square.js';
// SquareStack aliases
export { default as SquareStackIcon } from '../icons/square-stack.js';
// SquareStop aliases
export { default as SquareStopIcon } from '../icons/square-stop.js';
// Square aliases
export { default as SquareIcon } from '../icons/square.js';
// SquaresExclude aliases
export { default as SquaresExcludeIcon } from '../icons/squares-exclude.js';
// SquaresIntersect aliases
export { default as SquaresIntersectIcon } from '../icons/squares-intersect.js';
// SquaresSubtract aliases
export { default as SquaresSubtractIcon } from '../icons/squares-subtract.js';
// SquaresUnite aliases
export { default as SquaresUniteIcon } from '../icons/squares-unite.js';
// SquircleDashed aliases
export { default as SquircleDashedIcon } from '../icons/squircle-dashed.js';
// Squirrel aliases
export { default as SquirrelIcon } from '../icons/squirrel.js';
// Squircle aliases
export { default as SquircleIcon } from '../icons/squircle.js';
// Stamp aliases
export { default as StampIcon } from '../icons/stamp.js';
// StarHalf aliases
export { default as StarHalfIcon } from '../icons/star-half.js';
// StarOff aliases
export { default as StarOffIcon } from '../icons/star-off.js';
// Star aliases
export { default as StarIcon } from '../icons/star.js';
// StepBack aliases
export { default as StepBackIcon } from '../icons/step-back.js';
// StepForward aliases
export { default as StepForwardIcon } from '../icons/step-forward.js';
// Stethoscope aliases
export { default as StethoscopeIcon } from '../icons/stethoscope.js';
// StickyNote aliases
export { default as StickyNoteIcon } from '../icons/sticky-note.js';
// Store aliases
export { default as StoreIcon } from '../icons/store.js';
// StretchVertical aliases
export { default as StretchVerticalIcon } from '../icons/stretch-vertical.js';
// Sticker aliases
export { default as StickerIcon } from '../icons/sticker.js';
// Strikethrough aliases
export { default as StrikethroughIcon } from '../icons/strikethrough.js';
// StretchHorizontal aliases
export { default as StretchHorizontalIcon } from '../icons/stretch-horizontal.js';
// SunDim aliases
export { default as SunDimIcon } from '../icons/sun-dim.js';
// Subscript aliases
export { default as SubscriptIcon } from '../icons/subscript.js';
// SunMedium aliases
export { default as SunMediumIcon } from '../icons/sun-medium.js';
// SunMoon aliases
export { default as SunMoonIcon } from '../icons/sun-moon.js';
// Sun aliases
export { default as SunIcon } from '../icons/sun.js';
// SunSnow aliases
export { default as SunSnowIcon } from '../icons/sun-snow.js';
// Sunset aliases
export { default as SunsetIcon } from '../icons/sunset.js';
// Sunrise aliases
export { default as SunriseIcon } from '../icons/sunrise.js';
// Superscript aliases
export { default as SuperscriptIcon } from '../icons/superscript.js';
// SwissFranc aliases
export { default as SwissFrancIcon } from '../icons/swiss-franc.js';
// SwatchBook aliases
export { default as SwatchBookIcon } from '../icons/swatch-book.js';
// SwitchCamera aliases
export { default as SwitchCameraIcon } from '../icons/switch-camera.js';
// Sword aliases
export { default as SwordIcon } from '../icons/sword.js';
// Swords aliases
export { default as SwordsIcon } from '../icons/swords.js';
// Table2 aliases
export { default as Table2Icon } from '../icons/table-2.js';
// TableCellsMerge aliases
export { default as TableCellsMergeIcon } from '../icons/table-cells-merge.js';
// Syringe aliases
export { default as SyringeIcon } from '../icons/syringe.js';
// TableCellsSplit aliases
export { default as TableCellsSplitIcon } from '../icons/table-cells-split.js';
// TableColumnsSplit aliases
export { default as TableColumnsSplitIcon } from '../icons/table-columns-split.js';
// TableProperties aliases
export { default as TablePropertiesIcon } from '../icons/table-properties.js';
// TableRowsSplit aliases
export { default as TableRowsSplitIcon } from '../icons/table-rows-split.js';
// TableOfContents aliases
export { default as TableOfContentsIcon } from '../icons/table-of-contents.js';
// TabletSmartphone aliases
export { default as TabletSmartphoneIcon } from '../icons/tablet-smartphone.js';
// Table aliases
export { default as TableIcon } from '../icons/table.js';
// Tablets aliases
export { default as TabletsIcon } from '../icons/tablets.js';
// Tablet aliases
export { default as TabletIcon } from '../icons/tablet.js';
// Tag aliases
export { default as TagIcon } from '../icons/tag.js';
// Tags aliases
export { default as TagsIcon } from '../icons/tags.js';
// Tally2 aliases
export { default as Tally2Icon } from '../icons/tally-2.js';
// Tally1 aliases
export { default as Tally1Icon } from '../icons/tally-1.js';
// Tally3 aliases
export { default as Tally3Icon } from '../icons/tally-3.js';
// Tangent aliases
export { default as TangentIcon } from '../icons/tangent.js';
// Tally4 aliases
export { default as Tally4Icon } from '../icons/tally-4.js';
// Tally5 aliases
export { default as Tally5Icon } from '../icons/tally-5.js';
// Telescope aliases
export { default as TelescopeIcon } from '../icons/telescope.js';
// Target aliases
export { default as TargetIcon } from '../icons/target.js';
// TentTree aliases
export { default as TentTreeIcon } from '../icons/tent-tree.js';
// Tent aliases
export { default as TentIcon } from '../icons/tent.js';
// TestTube aliases
export { default as TestTubeIcon } from '../icons/test-tube.js';
// Terminal aliases
export { default as TerminalIcon } from '../icons/terminal.js';
// TestTubes aliases
export { default as TestTubesIcon } from '../icons/test-tubes.js';
// TextCursor aliases
export { default as TextCursorIcon } from '../icons/text-cursor.js';
// TextCursorInput aliases
export { default as TextCursorInputIcon } from '../icons/text-cursor-input.js';
// TextQuote aliases
export { default as TextQuoteIcon } from '../icons/text-quote.js';
// TextSearch aliases
export { default as TextSearchIcon } from '../icons/text-search.js';
// Text aliases
export { default as TextIcon } from '../icons/text.js';
// Theater aliases
export { default as TheaterIcon } from '../icons/theater.js';
// ThermometerSnowflake aliases
export { default as ThermometerSnowflakeIcon } from '../icons/thermometer-snowflake.js';
// ThermometerSun aliases
export { default as ThermometerSunIcon } from '../icons/thermometer-sun.js';
// Thermometer aliases
export { default as ThermometerIcon } from '../icons/thermometer.js';
// ThumbsDown aliases
export { default as ThumbsDownIcon } from '../icons/thumbs-down.js';
// TicketCheck aliases
export { default as TicketCheckIcon } from '../icons/ticket-check.js';
// ThumbsUp aliases
export { default as ThumbsUpIcon } from '../icons/thumbs-up.js';
// TicketMinus aliases
export { default as TicketMinusIcon } from '../icons/ticket-minus.js';
// TicketPercent aliases
export { default as TicketPercentIcon } from '../icons/ticket-percent.js';
// TicketPlus aliases
export { default as TicketPlusIcon } from '../icons/ticket-plus.js';
// TicketSlash aliases
export { default as TicketSlashIcon } from '../icons/ticket-slash.js';
// TicketX aliases
export { default as TicketXIcon } from '../icons/ticket-x.js';
// TicketsPlane aliases
export { default as TicketsPlaneIcon } from '../icons/tickets-plane.js';
// Ticket aliases
export { default as TicketIcon } from '../icons/ticket.js';
// Tickets aliases
export { default as TicketsIcon } from '../icons/tickets.js';
// TimerOff aliases
export { default as TimerOffIcon } from '../icons/timer-off.js';
// TimerReset aliases
export { default as TimerResetIcon } from '../icons/timer-reset.js';
// ToggleLeft aliases
export { default as ToggleLeftIcon } from '../icons/toggle-left.js';
// Timer aliases
export { default as TimerIcon } from '../icons/timer.js';
// ToggleRight aliases
export { default as ToggleRightIcon } from '../icons/toggle-right.js';
// ToolCase aliases
export { default as ToolCaseIcon } from '../icons/tool-case.js';
// Toilet aliases
export { default as ToiletIcon } from '../icons/toilet.js';
// Tornado aliases
export { default as TornadoIcon } from '../icons/tornado.js';
// Torus aliases
export { default as TorusIcon } from '../icons/torus.js';
// TouchpadOff aliases
export { default as TouchpadOffIcon } from '../icons/touchpad-off.js';
// TowerControl aliases
export { default as TowerControlIcon } from '../icons/tower-control.js';
// Touchpad aliases
export { default as TouchpadIcon } from '../icons/touchpad.js';
// ToyBrick aliases
export { default as ToyBrickIcon } from '../icons/toy-brick.js';
// TrafficCone aliases
export { default as TrafficConeIcon } from '../icons/traffic-cone.js';
// Tractor aliases
export { default as TractorIcon } from '../icons/tractor.js';
// TrainFrontTunnel aliases
export { default as TrainFrontTunnelIcon } from '../icons/train-front-tunnel.js';
// TrainFront aliases
export { default as TrainFrontIcon } from '../icons/train-front.js';
// TrainTrack aliases
export { default as TrainTrackIcon } from '../icons/train-track.js';
// Transgender aliases
export { default as TransgenderIcon } from '../icons/transgender.js';
// Trash2 aliases
export { default as Trash2Icon } from '../icons/trash-2.js';
// Trash aliases
export { default as TrashIcon } from '../icons/trash.js';
// TreeDeciduous aliases
export { default as TreeDeciduousIcon } from '../icons/tree-deciduous.js';
// TreePine aliases
export { default as TreePineIcon } from '../icons/tree-pine.js';
// Trees aliases
export { default as TreesIcon } from '../icons/trees.js';
// Trello aliases
export { default as TrelloIcon } from '../icons/trello.js';
// TrendingDown aliases
export { default as TrendingDownIcon } from '../icons/trending-down.js';
// TrendingUp aliases
export { default as TrendingUpIcon } from '../icons/trending-up.js';
// TrendingUpDown aliases
export { default as TrendingUpDownIcon } from '../icons/trending-up-down.js';
// TriangleDashed aliases
export { default as TriangleDashedIcon } from '../icons/triangle-dashed.js';
// TriangleRight aliases
export { default as TriangleRightIcon } from '../icons/triangle-right.js';
// Triangle aliases
export { default as TriangleIcon } from '../icons/triangle.js';
// Trophy aliases
export { default as TrophyIcon } from '../icons/trophy.js';
// TruckElectric aliases
export { default as TruckElectricIcon } from '../icons/truck-electric.js';
// Truck aliases
export { default as TruckIcon } from '../icons/truck.js';
// TurkishLira aliases
export { default as TurkishLiraIcon } from '../icons/turkish-lira.js';
// Turtle aliases
export { default as TurtleIcon } from '../icons/turtle.js';
// TvMinimalPlay aliases
export { default as TvMinimalPlayIcon } from '../icons/tv-minimal-play.js';
// Twitch aliases
export { default as TwitchIcon } from '../icons/twitch.js';
// Tv aliases
export { default as TvIcon } from '../icons/tv.js';
// Twitter aliases
export { default as TwitterIcon } from '../icons/twitter.js';
// Type aliases
export { default as TypeIcon } from '../icons/type.js';
// TypeOutline aliases
export { default as TypeOutlineIcon } from '../icons/type-outline.js';
// UmbrellaOff aliases
export { default as UmbrellaOffIcon } from '../icons/umbrella-off.js';
// Underline aliases
export { default as UnderlineIcon } from '../icons/underline.js';
// Umbrella aliases
export { default as UmbrellaIcon } from '../icons/umbrella.js';
// Undo2 aliases
export { default as Undo2Icon } from '../icons/undo-2.js';
// UndoDot aliases
export { default as UndoDotIcon } from '../icons/undo-dot.js';
// Undo aliases
export { default as UndoIcon } from '../icons/undo.js';
// UnfoldHorizontal aliases
export { default as UnfoldHorizontalIcon } from '../icons/unfold-horizontal.js';
// UnfoldVertical aliases
export { default as UnfoldVerticalIcon } from '../icons/unfold-vertical.js';
// Ungroup aliases
export { default as UngroupIcon } from '../icons/ungroup.js';
// Unlink2 aliases
export { default as Unlink2Icon } from '../icons/unlink-2.js';
// Unlink aliases
export { default as UnlinkIcon } from '../icons/unlink.js';
// Unplug aliases
export { default as UnplugIcon } from '../icons/unplug.js';
// Upload aliases
export { default as UploadIcon } from '../icons/upload.js';
// Usb aliases
export { default as UsbIcon } from '../icons/usb.js';
// UserCog aliases
export { default as UserCogIcon } from '../icons/user-cog.js';
// UserCheck aliases
export { default as UserCheckIcon } from '../icons/user-check.js';
// UserLock aliases
export { default as UserLockIcon } from '../icons/user-lock.js';
// UserMinus aliases
export { default as UserMinusIcon } from '../icons/user-minus.js';
// UserPlus aliases
export { default as UserPlusIcon } from '../icons/user-plus.js';
// UserPen aliases
export { default as UserPenIcon } from '../icons/user-pen.js';
// UserRoundPen aliases
export { default as UserRoundPenIcon } from '../icons/user-round-pen.js';
// UserRoundSearch aliases
export { default as UserRoundSearchIcon } from '../icons/user-round-search.js';
// UserSearch aliases
export { default as UserSearchIcon } from '../icons/user-search.js';
// UserStar aliases
export { default as UserStarIcon } from '../icons/user-star.js';
// UserX aliases
export { default as UserXIcon } from '../icons/user-x.js';
// User aliases
export { default as UserIcon } from '../icons/user.js';
// Users aliases
export { default as UsersIcon } from '../icons/users.js';
// UtilityPole aliases
export { default as UtilityPoleIcon } from '../icons/utility-pole.js';
// Variable aliases
export { default as VariableIcon } from '../icons/variable.js';
// Vault aliases
export { default as VaultIcon } from '../icons/vault.js';
// VectorSquare aliases
export { default as VectorSquareIcon } from '../icons/vector-square.js';
// Vegan aliases
export { default as VeganIcon } from '../icons/vegan.js';
// VenetianMask aliases
export { default as VenetianMaskIcon } from '../icons/venetian-mask.js';
// VenusAndMars aliases
export { default as VenusAndMarsIcon } from '../icons/venus-and-mars.js';
// Venus aliases
export { default as VenusIcon } from '../icons/venus.js';
// VibrateOff aliases
export { default as VibrateOffIcon } from '../icons/vibrate-off.js';
// Vibrate aliases
export { default as VibrateIcon } from '../icons/vibrate.js';
// VideoOff aliases
export { default as VideoOffIcon } from '../icons/video-off.js';
// Video aliases
export { default as VideoIcon } from '../icons/video.js';
// Videotape aliases
export { default as VideotapeIcon } from '../icons/videotape.js';
// View aliases
export { default as ViewIcon } from '../icons/view.js';
// Voicemail aliases
export { default as VoicemailIcon } from '../icons/voicemail.js';
// Volleyball aliases
export { default as VolleyballIcon } from '../icons/volleyball.js';
// Volume1 aliases
export { default as Volume1Icon } from '../icons/volume-1.js';
// Volume2 aliases
export { default as Volume2Icon } from '../icons/volume-2.js';
// VolumeOff aliases
export { default as VolumeOffIcon } from '../icons/volume-off.js';
// VolumeX aliases
export { default as VolumeXIcon } from '../icons/volume-x.js';
// Volume aliases
export { default as VolumeIcon } from '../icons/volume.js';
// Vote aliases
export { default as VoteIcon } from '../icons/vote.js';
// WalletCards aliases
export { default as WalletCardsIcon } from '../icons/wallet-cards.js';
// Wallet aliases
export { default as WalletIcon } from '../icons/wallet.js';
// Wallpaper aliases
export { default as WallpaperIcon } from '../icons/wallpaper.js';
// Wand aliases
export { default as WandIcon } from '../icons/wand.js';
// Warehouse aliases
export { default as WarehouseIcon } from '../icons/warehouse.js';
// WashingMachine aliases
export { default as WashingMachineIcon } from '../icons/washing-machine.js';
// Watch aliases
export { default as WatchIcon } from '../icons/watch.js';
// WavesLadder aliases
export { default as WavesLadderIcon } from '../icons/waves-ladder.js';
// Waves aliases
export { default as WavesIcon } from '../icons/waves.js';
// Waypoints aliases
export { default as WaypointsIcon } from '../icons/waypoints.js';
// Webcam aliases
export { default as WebcamIcon } from '../icons/webcam.js';
// Webhook aliases
export { default as WebhookIcon } from '../icons/webhook.js';
// WebhookOff aliases
export { default as WebhookOffIcon } from '../icons/webhook-off.js';
// Weight aliases
export { default as WeightIcon } from '../icons/weight.js';
// Wheat aliases
export { default as WheatIcon } from '../icons/wheat.js';
// WholeWord aliases
export { default as WholeWordIcon } from '../icons/whole-word.js';
// WheatOff aliases
export { default as WheatOffIcon } from '../icons/wheat-off.js';
// WifiCog aliases
export { default as WifiCogIcon } from '../icons/wifi-cog.js';
// WifiHigh aliases
export { default as WifiHighIcon } from '../icons/wifi-high.js';
// WifiOff aliases
export { default as WifiOffIcon } from '../icons/wifi-off.js';
// WifiLow aliases
export { default as WifiLowIcon } from '../icons/wifi-low.js';
// WifiSync aliases
export { default as WifiSyncIcon } from '../icons/wifi-sync.js';
// WifiPen aliases
export { default as WifiPenIcon } from '../icons/wifi-pen.js';
// WifiZero aliases
export { default as WifiZeroIcon } from '../icons/wifi-zero.js';
// Wifi aliases
export { default as WifiIcon } from '../icons/wifi.js';
// WindArrowDown aliases
export { default as WindArrowDownIcon } from '../icons/wind-arrow-down.js';
// Wind aliases
export { default as WindIcon } from '../icons/wind.js';
// Wine aliases
export { default as WineIcon } from '../icons/wine.js';
// Workflow aliases
export { default as WorkflowIcon } from '../icons/workflow.js';
// WineOff aliases
export { default as WineOffIcon } from '../icons/wine-off.js';
// Worm aliases
export { default as WormIcon } from '../icons/worm.js';
// Wrench aliases
export { default as WrenchIcon } from '../icons/wrench.js';
// WrapText aliases
export { default as WrapTextIcon } from '../icons/wrap-text.js';
// X aliases
export { default as XIcon } from '../icons/x.js';
// Youtube aliases
export { default as YoutubeIcon } from '../icons/youtube.js';
// Zap aliases
export { default as ZapIcon } from '../icons/zap.js';
// ZapOff aliases
export { default as ZapOffIcon } from '../icons/zap-off.js';
// ZoomIn aliases
export { default as ZoomInIcon } from '../icons/zoom-in.js';
// ZoomOut aliases
export { default as ZoomOutIcon } from '../icons/zoom-out.js';
// ArrowDown10 aliases
export { default as ArrowDown10Icon } from '../icons/arrow-down-1-0.js';
// ArrowDown01 aliases
export { default as ArrowDown01Icon } from '../icons/arrow-down-0-1.js';
// ArrowUp01 aliases
export { default as ArrowUp01Icon } from '../icons/arrow-up-0-1.js';
// ArrowUp10 aliases
export { default as ArrowUp10Icon } from '../icons/arrow-up-1-0.js';
// AlarmClockCheck aliases
export { default as AlarmClockCheckIcon } from '../icons/alarm-clock-check.js';
export { default as AlarmCheckIcon } from '../icons/alarm-check.js';
// AlarmClockMinus aliases
export { default as AlarmClockMinusIcon } from '../icons/alarm-clock-minus.js';
export { default as AlarmMinusIcon } from '../icons/alarm-minus.js';
// AlarmClockPlus aliases
export { default as AlarmClockPlusIcon } from '../icons/alarm-clock-plus.js';
export { default as AlarmPlusIcon } from '../icons/alarm-plus.js';
// ArrowDownAZ aliases
export { default as ArrowDownAZIcon } from '../icons/arrow-down-a-z.js';
export { default as ArrowDownAzIcon } from '../icons/arrow-down-az.js';
// ArrowDownZA aliases
export { default as ArrowDownZAIcon } from '../icons/arrow-down-z-a.js';
export { default as ArrowDownZaIcon } from '../icons/arrow-down-za.js';
// ArrowDownWideNarrow aliases
export { default as ArrowDownWideNarrowIcon } from '../icons/arrow-down-wide-narrow.js';
export { default as SortDescIcon } from '../icons/sort-desc.js';
// ArrowUpAZ aliases
export { default as ArrowUpAZIcon } from '../icons/arrow-up-a-z.js';
export { default as ArrowUpAzIcon } from '../icons/arrow-up-az.js';
// ArrowUpNarrowWide aliases
export { default as ArrowUpNarrowWideIcon } from '../icons/arrow-up-narrow-wide.js';
export { default as SortAscIcon } from '../icons/sort-asc.js';
// ArrowUpZA aliases
export { default as ArrowUpZAIcon } from '../icons/arrow-up-z-a.js';
export { default as ArrowUpZaIcon } from '../icons/arrow-up-za.js';
// Axis3d aliases
export { default as Axis3dIcon } from '../icons/axis-3d.js';
export { default as Axis3DIcon } from '../icons/axis-3-d.js';
// BadgeCheck aliases
export { default as BadgeCheckIcon } from '../icons/badge-check.js';
export { default as VerifiedIcon } from '../icons/verified.js';
// BadgeQuestionMark aliases
export { default as BadgeQuestionMarkIcon } from '../icons/badge-question-mark.js';
export { default as BadgeHelpIcon } from '../icons/badge-help.js';
// BetweenHorizontalStart aliases
export { default as BetweenHorizontalStartIcon } from '../icons/between-horizontal-start.js';
export { 
/** @deprecated Renamed because of typo, use {@link BetweenHorizontalStart} instead. This alias will be removed in v1.0 */
default as BetweenHorizonalStartIcon } from '../icons/between-horizonal-start.js';
// BetweenHorizontalEnd aliases
export { default as BetweenHorizontalEndIcon } from '../icons/between-horizontal-end.js';
export { 
/** @deprecated Renamed because of typo, use {@link BetweenHorizontalEnd} instead. This alias will be removed in v1.0 */
default as BetweenHorizonalEndIcon } from '../icons/between-horizonal-end.js';
// BookDashed aliases
export { default as BookDashedIcon } from '../icons/book-dashed.js';
export { default as BookTemplateIcon } from '../icons/book-template.js';
// Braces aliases
export { default as BracesIcon } from '../icons/braces.js';
export { default as CurlyBracesIcon } from '../icons/curly-braces.js';
// Captions aliases
export { default as CaptionsIcon } from '../icons/captions.js';
export { default as SubtitlesIcon } from '../icons/subtitles.js';
// ChartArea aliases
export { default as ChartAreaIcon } from '../icons/chart-area.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartArea} instead. This alias will be removed in v1.0 */
default as AreaChartIcon } from '../icons/area-chart.js';
// ChartBarBig aliases
export { default as ChartBarBigIcon } from '../icons/chart-bar-big.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartBarBig} instead. This alias will be removed in v1.0 */
default as BarChartHorizontalBigIcon } from '../icons/bar-chart-horizontal-big.js';
// ChartBar aliases
export { default as ChartBarIcon } from '../icons/chart-bar.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartBar} instead. This alias will be removed in v1.0 */
default as BarChartHorizontalIcon } from '../icons/bar-chart-horizontal.js';
// ChartCandlestick aliases
export { default as ChartCandlestickIcon } from '../icons/chart-candlestick.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartCandlestick} instead. This alias will be removed in v1.0 */
default as CandlestickChartIcon } from '../icons/candlestick-chart.js';
// ChartColumnBig aliases
export { default as ChartColumnBigIcon } from '../icons/chart-column-big.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartColumnBig} instead. This alias will be removed in v1.0 */
default as BarChartBigIcon } from '../icons/bar-chart-big.js';
// ChartColumnIncreasing aliases
export { default as ChartColumnIncreasingIcon } from '../icons/chart-column-increasing.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartColumnIncreasing} instead. This alias will be removed in v1.0 */
default as BarChart4Icon } from '../icons/bar-chart-4.js';
// ChartColumn aliases
export { default as ChartColumnIcon } from '../icons/chart-column.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartColumn} instead. This alias will be removed in v1.0 */
default as BarChart3Icon } from '../icons/bar-chart-3.js';
// ChartLine aliases
export { default as ChartLineIcon } from '../icons/chart-line.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartLine} instead. This alias will be removed in v1.0 */
default as LineChartIcon } from '../icons/line-chart.js';
// ChartNoAxesColumnIncreasing aliases
export { default as ChartNoAxesColumnIncreasingIcon } from '../icons/chart-no-axes-column-increasing.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartNoAxesColumnIncreasing} instead. This alias will be removed in v1.0 */
default as BarChartIcon } from '../icons/bar-chart.js';
// ChartNoAxesColumn aliases
export { default as ChartNoAxesColumnIcon } from '../icons/chart-no-axes-column.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartNoAxesColumn} instead. This alias will be removed in v1.0 */
default as BarChart2Icon } from '../icons/bar-chart-2.js';
// ChartNoAxesGantt aliases
export { default as ChartNoAxesGanttIcon } from '../icons/chart-no-axes-gantt.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartNoAxesGantt} instead. This alias will be removed in v1.0 */
default as GanttChartIcon } from '../icons/gantt-chart.js';
// ChartScatter aliases
export { default as ChartScatterIcon } from '../icons/chart-scatter.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartScatter} instead. This alias will be removed in v1.0 */
default as ScatterChartIcon } from '../icons/scatter-chart.js';
// ChartPie aliases
export { default as ChartPieIcon } from '../icons/chart-pie.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ChartPie} instead. This alias will be removed in v1.0 */
default as PieChartIcon } from '../icons/pie-chart.js';
// CircleAlert aliases
export { default as CircleAlertIcon } from '../icons/circle-alert.js';
export { default as AlertCircleIcon } from '../icons/alert-circle.js';
// CircleArrowDown aliases
export { default as CircleArrowDownIcon } from '../icons/circle-arrow-down.js';
export { default as ArrowDownCircleIcon } from '../icons/arrow-down-circle.js';
// CircleArrowLeft aliases
export { default as CircleArrowLeftIcon } from '../icons/circle-arrow-left.js';
export { default as ArrowLeftCircleIcon } from '../icons/arrow-left-circle.js';
// CircleArrowOutDownLeft aliases
export { default as CircleArrowOutDownLeftIcon } from '../icons/circle-arrow-out-down-left.js';
export { default as ArrowDownLeftFromCircleIcon } from '../icons/arrow-down-left-from-circle.js';
// CircleArrowOutDownRight aliases
export { default as CircleArrowOutDownRightIcon } from '../icons/circle-arrow-out-down-right.js';
export { default as ArrowDownRightFromCircleIcon } from '../icons/arrow-down-right-from-circle.js';
// CircleArrowOutUpLeft aliases
export { default as CircleArrowOutUpLeftIcon } from '../icons/circle-arrow-out-up-left.js';
export { default as ArrowUpLeftFromCircleIcon } from '../icons/arrow-up-left-from-circle.js';
// CircleArrowOutUpRight aliases
export { default as CircleArrowOutUpRightIcon } from '../icons/circle-arrow-out-up-right.js';
export { default as ArrowUpRightFromCircleIcon } from '../icons/arrow-up-right-from-circle.js';
// CircleArrowRight aliases
export { default as CircleArrowRightIcon } from '../icons/circle-arrow-right.js';
export { default as ArrowRightCircleIcon } from '../icons/arrow-right-circle.js';
// CircleArrowUp aliases
export { default as CircleArrowUpIcon } from '../icons/circle-arrow-up.js';
export { default as ArrowUpCircleIcon } from '../icons/arrow-up-circle.js';
// CircleCheckBig aliases
export { default as CircleCheckBigIcon } from '../icons/circle-check-big.js';
export { default as CheckCircleIcon } from '../icons/check-circle.js';
// CircleCheck aliases
export { default as CircleCheckIcon } from '../icons/circle-check.js';
export { default as CheckCircle2Icon } from '../icons/check-circle-2.js';
// CircleChevronDown aliases
export { default as CircleChevronDownIcon } from '../icons/circle-chevron-down.js';
export { default as ChevronDownCircleIcon } from '../icons/chevron-down-circle.js';
// CircleChevronLeft aliases
export { default as CircleChevronLeftIcon } from '../icons/circle-chevron-left.js';
export { default as ChevronLeftCircleIcon } from '../icons/chevron-left-circle.js';
// CircleChevronRight aliases
export { default as CircleChevronRightIcon } from '../icons/circle-chevron-right.js';
export { default as ChevronRightCircleIcon } from '../icons/chevron-right-circle.js';
// CircleChevronUp aliases
export { default as CircleChevronUpIcon } from '../icons/circle-chevron-up.js';
export { default as ChevronUpCircleIcon } from '../icons/chevron-up-circle.js';
// CircleDivide aliases
export { default as CircleDivideIcon } from '../icons/circle-divide.js';
export { default as DivideCircleIcon } from '../icons/divide-circle.js';
// CircleGauge aliases
export { default as CircleGaugeIcon } from '../icons/circle-gauge.js';
export { default as GaugeCircleIcon } from '../icons/gauge-circle.js';
// CircleMinus aliases
export { default as CircleMinusIcon } from '../icons/circle-minus.js';
export { default as MinusCircleIcon } from '../icons/minus-circle.js';
// CircleParking aliases
export { default as CircleParkingIcon } from '../icons/circle-parking.js';
export { default as ParkingCircleIcon } from '../icons/parking-circle.js';
// CircleParkingOff aliases
export { default as CircleParkingOffIcon } from '../icons/circle-parking-off.js';
export { default as ParkingCircleOffIcon } from '../icons/parking-circle-off.js';
// CirclePause aliases
export { default as CirclePauseIcon } from '../icons/circle-pause.js';
export { default as PauseCircleIcon } from '../icons/pause-circle.js';
// CirclePlay aliases
export { default as CirclePlayIcon } from '../icons/circle-play.js';
export { default as PlayCircleIcon } from '../icons/play-circle.js';
// CirclePercent aliases
export { default as CirclePercentIcon } from '../icons/circle-percent.js';
export { default as PercentCircleIcon } from '../icons/percent-circle.js';
// CirclePower aliases
export { default as CirclePowerIcon } from '../icons/circle-power.js';
export { default as PowerCircleIcon } from '../icons/power-circle.js';
// CirclePlus aliases
export { default as CirclePlusIcon } from '../icons/circle-plus.js';
export { default as PlusCircleIcon } from '../icons/plus-circle.js';
// CircleQuestionMark aliases
export { default as CircleQuestionMarkIcon } from '../icons/circle-question-mark.js';
export { default as HelpCircleIcon } from '../icons/help-circle.js';
export { default as CircleHelpIcon } from '../icons/circle-help.js';
// CircleSlash2 aliases
export { default as CircleSlash2Icon } from '../icons/circle-slash-2.js';
export { default as CircleSlashedIcon } from '../icons/circle-slashed.js';
// CircleStop aliases
export { default as CircleStopIcon } from '../icons/circle-stop.js';
export { default as StopCircleIcon } from '../icons/stop-circle.js';
// CircleUserRound aliases
export { default as CircleUserRoundIcon } from '../icons/circle-user-round.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link CircleUserRound} instead. This alias will be removed in v1.0 */
default as UserCircle2Icon } from '../icons/user-circle-2.js';
// CircleUser aliases
export { default as CircleUserIcon } from '../icons/circle-user.js';
export { default as UserCircleIcon } from '../icons/user-circle.js';
// CircleX aliases
export { default as CircleXIcon } from '../icons/circle-x.js';
export { default as XCircleIcon } from '../icons/x-circle.js';
// ClipboardPen aliases
export { default as ClipboardPenIcon } from '../icons/clipboard-pen.js';
export { default as ClipboardEditIcon } from '../icons/clipboard-edit.js';
// ClipboardPenLine aliases
export { default as ClipboardPenLineIcon } from '../icons/clipboard-pen-line.js';
export { default as ClipboardSignatureIcon } from '../icons/clipboard-signature.js';
// CloudDownload aliases
export { default as CloudDownloadIcon } from '../icons/cloud-download.js';
export { default as DownloadCloudIcon } from '../icons/download-cloud.js';
// CloudUpload aliases
export { default as CloudUploadIcon } from '../icons/cloud-upload.js';
export { default as UploadCloudIcon } from '../icons/upload-cloud.js';
// CodeXml aliases
export { default as CodeXmlIcon } from '../icons/code-xml.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link CodeXml} instead. This alias will be removed in v1.0 */
default as Code2Icon } from '../icons/code-2.js';
// Columns2 aliases
export { default as Columns2Icon } from '../icons/columns-2.js';
export { default as ColumnsIcon } from '../icons/columns.js';
// Columns3 aliases
export { default as Columns3Icon } from '../icons/columns-3.js';
export { default as PanelsLeftRightIcon } from '../icons/panels-left-right.js';
// Columns3Cog aliases
export { default as Columns3CogIcon } from '../icons/columns-3-cog.js';
export { default as ColumnsSettingsIcon } from '../icons/columns-settings.js';
export { default as TableConfigIcon } from '../icons/table-config.js';
// ContactRound aliases
export { default as ContactRoundIcon } from '../icons/contact-round.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link ContactRound} instead. This alias will be removed in v1.0 */
default as Contact2Icon } from '../icons/contact-2.js';
// DiamondPercent aliases
export { default as DiamondPercentIcon } from '../icons/diamond-percent.js';
export { default as PercentDiamondIcon } from '../icons/percent-diamond.js';
// Earth aliases
export { default as EarthIcon } from '../icons/earth.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link Earth} instead. This alias will be removed in v1.0 */
default as Globe2Icon } from '../icons/globe-2.js';
// EllipsisVertical aliases
export { default as EllipsisVerticalIcon } from '../icons/ellipsis-vertical.js';
export { default as MoreVerticalIcon } from '../icons/more-vertical.js';
// Ellipsis aliases
export { default as EllipsisIcon } from '../icons/ellipsis.js';
export { default as MoreHorizontalIcon } from '../icons/more-horizontal.js';
// FileChartColumnIncreasing aliases
export { default as FileChartColumnIncreasingIcon } from '../icons/file-chart-column-increasing.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileChartColumnIncreasing} instead. This alias will be removed in v1.0 */
default as FileBarChartIcon } from '../icons/file-bar-chart.js';
// FileAxis3d aliases
export { default as FileAxis3dIcon } from '../icons/file-axis-3d.js';
export { default as FileAxis3DIcon } from '../icons/file-axis-3-d.js';
// FileChartColumn aliases
export { default as FileChartColumnIcon } from '../icons/file-chart-column.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileChartColumn} instead. This alias will be removed in v1.0 */
default as FileBarChart2Icon } from '../icons/file-bar-chart-2.js';
// FileChartLine aliases
export { default as FileChartLineIcon } from '../icons/file-chart-line.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileChartLine} instead. This alias will be removed in v1.0 */
default as FileLineChartIcon } from '../icons/file-line-chart.js';
// FileCog aliases
export { default as FileCogIcon } from '../icons/file-cog.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileCog} instead. This alias will be removed in v1.0 */
default as FileCog2Icon } from '../icons/file-cog-2.js';
// FilePenLine aliases
export { default as FilePenLineIcon } from '../icons/file-pen-line.js';
export { default as FileSignatureIcon } from '../icons/file-signature.js';
// FileChartPie aliases
export { default as FileChartPieIcon } from '../icons/file-chart-pie.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FileChartPie} instead. This alias will be removed in v1.0 */
default as FilePieChartIcon } from '../icons/file-pie-chart.js';
// FilePlay aliases
export { default as FilePlayIcon } from '../icons/file-play.js';
export { default as FileVideoIcon } from '../icons/file-video.js';
// FilePen aliases
export { default as FilePenIcon } from '../icons/file-pen.js';
export { default as FileEditIcon } from '../icons/file-edit.js';
// FileQuestionMark aliases
export { default as FileQuestionMarkIcon } from '../icons/file-question-mark.js';
export { default as FileQuestionIcon } from '../icons/file-question.js';
// FileVideoCamera aliases
export { default as FileVideoCameraIcon } from '../icons/file-video-camera.js';
export { default as FileVideo2Icon } from '../icons/file-video-2.js';
// FolderCog aliases
export { default as FolderCogIcon } from '../icons/folder-cog.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FolderCog} instead. This alias will be removed in v1.0 */
default as FolderCog2Icon } from '../icons/folder-cog-2.js';
// FolderPen aliases
export { default as FolderPenIcon } from '../icons/folder-pen.js';
export { default as FolderEditIcon } from '../icons/folder-edit.js';
// FunnelX aliases
export { default as FunnelXIcon } from '../icons/funnel-x.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link FunnelX} instead. This alias will be removed in v1.0 */
default as FilterXIcon } from '../icons/filter-x.js';
// Funnel aliases
export { default as FunnelIcon } from '../icons/funnel.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link Funnel} instead. This alias will be removed in v1.0 */
default as FilterIcon } from '../icons/filter.js';
// GitCommitHorizontal aliases
export { default as GitCommitHorizontalIcon } from '../icons/git-commit-horizontal.js';
export { default as GitCommitIcon } from '../icons/git-commit.js';
// Grid2x2Check aliases
export { default as Grid2x2CheckIcon } from '../icons/grid-2x2-check.js';
export { default as Grid2X2CheckIcon } from '../icons/grid-2-x-2-check.js';
// Grid2x2X aliases
export { default as Grid2x2XIcon } from '../icons/grid-2x2-x.js';
export { default as Grid2X2XIcon } from '../icons/grid-2-x-2-x.js';
// Grid2x2Plus aliases
export { default as Grid2x2PlusIcon } from '../icons/grid-2x2-plus.js';
export { default as Grid2X2PlusIcon } from '../icons/grid-2-x-2-plus.js';
// Grid2x2 aliases
export { default as Grid2x2Icon } from '../icons/grid-2x2.js';
export { default as Grid2X2Icon } from '../icons/grid-2-x-2.js';
// Grid3x3 aliases
export { default as Grid3x3Icon } from '../icons/grid-3x3.js';
export { default as GridIcon } from '../icons/grid.js';
export { default as Grid3X3Icon } from '../icons/grid-3-x-3.js';
// HandHelping aliases
export { default as HandHelpingIcon } from '../icons/hand-helping.js';
export { default as HelpingHandIcon } from '../icons/helping-hand.js';
// House aliases
export { default as HouseIcon } from '../icons/house.js';
export { default as HomeIcon } from '../icons/home.js';
// IceCreamBowl aliases
export { default as IceCreamBowlIcon } from '../icons/ice-cream-bowl.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link IceCreamBowl} instead. This alias will be removed in v1.0 */
default as IceCream2Icon } from '../icons/ice-cream-2.js';
// IceCreamCone aliases
export { default as IceCreamConeIcon } from '../icons/ice-cream-cone.js';
export { default as IceCreamIcon } from '../icons/ice-cream.js';
// IndentDecrease aliases
export { default as IndentDecreaseIcon } from '../icons/indent-decrease.js';
export { default as OutdentIcon } from '../icons/outdent.js';
// IndentIncrease aliases
export { default as IndentIncreaseIcon } from '../icons/indent-increase.js';
export { default as IndentIcon } from '../icons/indent.js';
// LaptopMinimal aliases
export { default as LaptopMinimalIcon } from '../icons/laptop-minimal.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link LaptopMinimal} instead. This alias will be removed in v1.0 */
default as Laptop2Icon } from '../icons/laptop-2.js';
// Layers aliases
export { default as LayersIcon } from '../icons/layers.js';
export { default as Layers3Icon } from '../icons/layers-3.js';
// LoaderCircle aliases
export { default as LoaderCircleIcon } from '../icons/loader-circle.js';
export { default as Loader2Icon } from '../icons/loader-2.js';
// LockKeyholeOpen aliases
export { default as LockKeyholeOpenIcon } from '../icons/lock-keyhole-open.js';
export { default as UnlockKeyholeIcon } from '../icons/unlock-keyhole.js';
// LockOpen aliases
export { default as LockOpenIcon } from '../icons/lock-open.js';
export { default as UnlockIcon } from '../icons/unlock.js';
// MailQuestionMark aliases
export { default as MailQuestionMarkIcon } from '../icons/mail-question-mark.js';
export { default as MailQuestionIcon } from '../icons/mail-question.js';
// MapPinPen aliases
export { default as MapPinPenIcon } from '../icons/map-pin-pen.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link MapPinPen} instead. This alias will be removed in v1.0 */
default as LocationEditIcon } from '../icons/location-edit.js';
// MessageCircleQuestionMark aliases
export { default as MessageCircleQuestionMarkIcon } from '../icons/message-circle-question-mark.js';
export { default as MessageCircleQuestionIcon } from '../icons/message-circle-question.js';
// MicVocal aliases
export { default as MicVocalIcon } from '../icons/mic-vocal.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link MicVocal} instead. This alias will be removed in v1.0 */
default as Mic2Icon } from '../icons/mic-2.js';
// Move3d aliases
export { default as Move3dIcon } from '../icons/move-3d.js';
export { default as Move3DIcon } from '../icons/move-3-d.js';
// OctagonAlert aliases
export { default as OctagonAlertIcon } from '../icons/octagon-alert.js';
export { default as AlertOctagonIcon } from '../icons/alert-octagon.js';
// OctagonPause aliases
export { default as OctagonPauseIcon } from '../icons/octagon-pause.js';
export { default as PauseOctagonIcon } from '../icons/pause-octagon.js';
// OctagonX aliases
export { default as OctagonXIcon } from '../icons/octagon-x.js';
export { default as XOctagonIcon } from '../icons/x-octagon.js';
// PaintbrushVertical aliases
export { default as PaintbrushVerticalIcon } from '../icons/paintbrush-vertical.js';
export { default as Paintbrush2Icon } from '../icons/paintbrush-2.js';
// PanelBottomDashed aliases
export { default as PanelBottomDashedIcon } from '../icons/panel-bottom-dashed.js';
export { default as PanelBottomInactiveIcon } from '../icons/panel-bottom-inactive.js';
// PanelLeftClose aliases
export { default as PanelLeftCloseIcon } from '../icons/panel-left-close.js';
export { default as SidebarCloseIcon } from '../icons/sidebar-close.js';
// PanelLeftDashed aliases
export { default as PanelLeftDashedIcon } from '../icons/panel-left-dashed.js';
export { default as PanelLeftInactiveIcon } from '../icons/panel-left-inactive.js';
// PanelLeftOpen aliases
export { default as PanelLeftOpenIcon } from '../icons/panel-left-open.js';
export { default as SidebarOpenIcon } from '../icons/sidebar-open.js';
// PanelLeft aliases
export { default as PanelLeftIcon } from '../icons/panel-left.js';
export { default as SidebarIcon } from '../icons/sidebar.js';
// PanelRightDashed aliases
export { default as PanelRightDashedIcon } from '../icons/panel-right-dashed.js';
export { default as PanelRightInactiveIcon } from '../icons/panel-right-inactive.js';
// PanelTopDashed aliases
export { default as PanelTopDashedIcon } from '../icons/panel-top-dashed.js';
export { default as PanelTopInactiveIcon } from '../icons/panel-top-inactive.js';
// PanelsTopLeft aliases
export { default as PanelsTopLeftIcon } from '../icons/panels-top-left.js';
export { default as LayoutIcon } from '../icons/layout.js';
// PenLine aliases
export { default as PenLineIcon } from '../icons/pen-line.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link PenLine} instead. This alias will be removed in v1.0 */
default as Edit3Icon } from '../icons/edit-3.js';
// Pen aliases
export { default as PenIcon } from '../icons/pen.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link Pen} instead. This alias will be removed in v1.0 */
default as Edit2Icon } from '../icons/edit-2.js';
// PlugZap aliases
export { default as PlugZapIcon } from '../icons/plug-zap.js';
export { default as PlugZap2Icon } from '../icons/plug-zap-2.js';
// RectangleEllipsis aliases
export { default as RectangleEllipsisIcon } from '../icons/rectangle-ellipsis.js';
export { default as FormInputIcon } from '../icons/form-input.js';
// Rotate3d aliases
export { default as Rotate3dIcon } from '../icons/rotate-3d.js';
export { default as Rotate3DIcon } from '../icons/rotate-3-d.js';
// Rows2 aliases
export { default as Rows2Icon } from '../icons/rows-2.js';
export { default as RowsIcon } from '../icons/rows.js';
// Rows3 aliases
export { default as Rows3Icon } from '../icons/rows-3.js';
export { default as PanelsTopBottomIcon } from '../icons/panels-top-bottom.js';
// Scale3d aliases
export { default as Scale3dIcon } from '../icons/scale-3d.js';
export { default as Scale3DIcon } from '../icons/scale-3-d.js';
// SendHorizontal aliases
export { default as SendHorizontalIcon } from '../icons/send-horizontal.js';
export { 
/** @deprecated Renamed because of typo, use {@link SendHorizontal} instead. This alias will be removed in v1.0 */
default as SendHorizonalIcon } from '../icons/send-horizonal.js';
// ShieldQuestionMark aliases
export { default as ShieldQuestionMarkIcon } from '../icons/shield-question-mark.js';
export { default as ShieldQuestionIcon } from '../icons/shield-question.js';
// ShieldX aliases
export { default as ShieldXIcon } from '../icons/shield-x.js';
export { default as ShieldCloseIcon } from '../icons/shield-close.js';
// SlidersVertical aliases
export { default as SlidersVerticalIcon } from '../icons/sliders-vertical.js';
export { default as SlidersIcon } from '../icons/sliders.js';
// Sparkles aliases
export { default as SparklesIcon } from '../icons/sparkles.js';
export { default as StarsIcon } from '../icons/stars.js';
// SquareArrowDownLeft aliases
export { default as SquareArrowDownLeftIcon } from '../icons/square-arrow-down-left.js';
export { default as ArrowDownLeftSquareIcon } from '../icons/arrow-down-left-square.js';
// SquareActivity aliases
export { default as SquareActivityIcon } from '../icons/square-activity.js';
export { default as ActivitySquareIcon } from '../icons/activity-square.js';
// SquareArrowDownRight aliases
export { default as SquareArrowDownRightIcon } from '../icons/square-arrow-down-right.js';
export { default as ArrowDownRightSquareIcon } from '../icons/arrow-down-right-square.js';
// SquareArrowDown aliases
export { default as SquareArrowDownIcon } from '../icons/square-arrow-down.js';
export { default as ArrowDownSquareIcon } from '../icons/arrow-down-square.js';
// SquareArrowLeft aliases
export { default as SquareArrowLeftIcon } from '../icons/square-arrow-left.js';
export { default as ArrowLeftSquareIcon } from '../icons/arrow-left-square.js';
// SquareArrowOutDownLeft aliases
export { default as SquareArrowOutDownLeftIcon } from '../icons/square-arrow-out-down-left.js';
export { default as ArrowDownLeftFromSquareIcon } from '../icons/arrow-down-left-from-square.js';
// SquareArrowOutDownRight aliases
export { default as SquareArrowOutDownRightIcon } from '../icons/square-arrow-out-down-right.js';
export { default as ArrowDownRightFromSquareIcon } from '../icons/arrow-down-right-from-square.js';
// SquareArrowOutUpRight aliases
export { default as SquareArrowOutUpRightIcon } from '../icons/square-arrow-out-up-right.js';
export { default as ArrowUpRightFromSquareIcon } from '../icons/arrow-up-right-from-square.js';
// SquareArrowOutUpLeft aliases
export { default as SquareArrowOutUpLeftIcon } from '../icons/square-arrow-out-up-left.js';
export { default as ArrowUpLeftFromSquareIcon } from '../icons/arrow-up-left-from-square.js';
// SquareArrowRight aliases
export { default as SquareArrowRightIcon } from '../icons/square-arrow-right.js';
export { default as ArrowRightSquareIcon } from '../icons/arrow-right-square.js';
// SquareArrowUpLeft aliases
export { default as SquareArrowUpLeftIcon } from '../icons/square-arrow-up-left.js';
export { default as ArrowUpLeftSquareIcon } from '../icons/arrow-up-left-square.js';
// SquareArrowUpRight aliases
export { default as SquareArrowUpRightIcon } from '../icons/square-arrow-up-right.js';
export { default as ArrowUpRightSquareIcon } from '../icons/arrow-up-right-square.js';
// SquareArrowUp aliases
export { default as SquareArrowUpIcon } from '../icons/square-arrow-up.js';
export { default as ArrowUpSquareIcon } from '../icons/arrow-up-square.js';
// SquareAsterisk aliases
export { default as SquareAsteriskIcon } from '../icons/square-asterisk.js';
export { default as AsteriskSquareIcon } from '../icons/asterisk-square.js';
// SquareBottomDashedScissors aliases
export { default as SquareBottomDashedScissorsIcon } from '../icons/square-bottom-dashed-scissors.js';
export { default as ScissorsSquareDashedBottomIcon } from '../icons/scissors-square-dashed-bottom.js';
// SquareChartGantt aliases
export { default as SquareChartGanttIcon } from '../icons/square-chart-gantt.js';
export { default as GanttChartSquareIcon } from '../icons/gantt-chart-square.js';
export { default as SquareGanttChartIcon } from '../icons/square-gantt-chart.js';
// SquareCheck aliases
export { default as SquareCheckIcon } from '../icons/square-check.js';
export { default as CheckSquare2Icon } from '../icons/check-square-2.js';
// SquareCheckBig aliases
export { default as SquareCheckBigIcon } from '../icons/square-check-big.js';
export { default as CheckSquareIcon } from '../icons/check-square.js';
// SquareChevronDown aliases
export { default as SquareChevronDownIcon } from '../icons/square-chevron-down.js';
export { default as ChevronDownSquareIcon } from '../icons/chevron-down-square.js';
// SquareChevronLeft aliases
export { default as SquareChevronLeftIcon } from '../icons/square-chevron-left.js';
export { default as ChevronLeftSquareIcon } from '../icons/chevron-left-square.js';
// SquareChevronRight aliases
export { default as SquareChevronRightIcon } from '../icons/square-chevron-right.js';
export { default as ChevronRightSquareIcon } from '../icons/chevron-right-square.js';
// SquareChevronUp aliases
export { default as SquareChevronUpIcon } from '../icons/square-chevron-up.js';
export { default as ChevronUpSquareIcon } from '../icons/chevron-up-square.js';
// SquareCode aliases
export { default as SquareCodeIcon } from '../icons/square-code.js';
export { default as CodeSquareIcon } from '../icons/code-square.js';
// SquareDashedKanban aliases
export { default as SquareDashedKanbanIcon } from '../icons/square-dashed-kanban.js';
export { default as KanbanSquareDashedIcon } from '../icons/kanban-square-dashed.js';
// SquareDashed aliases
export { default as SquareDashedIcon } from '../icons/square-dashed.js';
export { default as BoxSelectIcon } from '../icons/box-select.js';
// SquareDashedMousePointer aliases
export { default as SquareDashedMousePointerIcon } from '../icons/square-dashed-mouse-pointer.js';
export { default as MousePointerSquareDashedIcon } from '../icons/mouse-pointer-square-dashed.js';
// SquareDot aliases
export { default as SquareDotIcon } from '../icons/square-dot.js';
export { default as DotSquareIcon } from '../icons/dot-square.js';
// SquareDivide aliases
export { default as SquareDivideIcon } from '../icons/square-divide.js';
export { default as DivideSquareIcon } from '../icons/divide-square.js';
// SquareEqual aliases
export { default as SquareEqualIcon } from '../icons/square-equal.js';
export { default as EqualSquareIcon } from '../icons/equal-square.js';
// SquareFunction aliases
export { default as SquareFunctionIcon } from '../icons/square-function.js';
export { default as FunctionSquareIcon } from '../icons/function-square.js';
// SquareKanban aliases
export { default as SquareKanbanIcon } from '../icons/square-kanban.js';
export { default as KanbanSquareIcon } from '../icons/kanban-square.js';
// SquareLibrary aliases
export { default as SquareLibraryIcon } from '../icons/square-library.js';
export { default as LibrarySquareIcon } from '../icons/library-square.js';
// SquareM aliases
export { default as SquareMIcon } from '../icons/square-m.js';
export { default as MSquareIcon } from '../icons/m-square.js';
// SquareMenu aliases
export { default as SquareMenuIcon } from '../icons/square-menu.js';
export { default as MenuSquareIcon } from '../icons/menu-square.js';
// SquareMousePointer aliases
export { default as SquareMousePointerIcon } from '../icons/square-mouse-pointer.js';
export { default as InspectIcon } from '../icons/inspect.js';
// SquareParking aliases
export { default as SquareParkingIcon } from '../icons/square-parking.js';
export { default as ParkingSquareIcon } from '../icons/parking-square.js';
// SquareMinus aliases
export { default as SquareMinusIcon } from '../icons/square-minus.js';
export { default as MinusSquareIcon } from '../icons/minus-square.js';
// SquareParkingOff aliases
export { default as SquareParkingOffIcon } from '../icons/square-parking-off.js';
export { default as ParkingSquareOffIcon } from '../icons/parking-square-off.js';
// SquarePercent aliases
export { default as SquarePercentIcon } from '../icons/square-percent.js';
export { default as PercentSquareIcon } from '../icons/percent-square.js';
// SquarePen aliases
export { default as SquarePenIcon } from '../icons/square-pen.js';
export { default as PenBoxIcon } from '../icons/pen-box.js';
export { default as EditIcon } from '../icons/edit.js';
export { default as PenSquareIcon } from '../icons/pen-square.js';
// SquarePi aliases
export { default as SquarePiIcon } from '../icons/square-pi.js';
export { default as PiSquareIcon } from '../icons/pi-square.js';
// SquarePilcrow aliases
export { default as SquarePilcrowIcon } from '../icons/square-pilcrow.js';
export { default as PilcrowSquareIcon } from '../icons/pilcrow-square.js';
// SquarePlay aliases
export { default as SquarePlayIcon } from '../icons/square-play.js';
export { default as PlaySquareIcon } from '../icons/play-square.js';
// SquarePlus aliases
export { default as SquarePlusIcon } from '../icons/square-plus.js';
export { default as PlusSquareIcon } from '../icons/plus-square.js';
// SquarePower aliases
export { default as SquarePowerIcon } from '../icons/square-power.js';
export { default as PowerSquareIcon } from '../icons/power-square.js';
// SquareScissors aliases
export { default as SquareScissorsIcon } from '../icons/square-scissors.js';
export { default as ScissorsSquareIcon } from '../icons/scissors-square.js';
// SquareSigma aliases
export { default as SquareSigmaIcon } from '../icons/square-sigma.js';
export { default as SigmaSquareIcon } from '../icons/sigma-square.js';
// SquareSlash aliases
export { default as SquareSlashIcon } from '../icons/square-slash.js';
export { default as SlashSquareIcon } from '../icons/slash-square.js';
// SquareSplitHorizontal aliases
export { default as SquareSplitHorizontalIcon } from '../icons/square-split-horizontal.js';
export { default as SplitSquareHorizontalIcon } from '../icons/split-square-horizontal.js';
// SquareSplitVertical aliases
export { default as SquareSplitVerticalIcon } from '../icons/square-split-vertical.js';
export { default as SplitSquareVerticalIcon } from '../icons/split-square-vertical.js';
// SquareUserRound aliases
export { default as SquareUserRoundIcon } from '../icons/square-user-round.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link SquareUserRound} instead. This alias will be removed in v1.0 */
default as UserSquare2Icon } from '../icons/user-square-2.js';
// SquareTerminal aliases
export { default as SquareTerminalIcon } from '../icons/square-terminal.js';
export { default as TerminalSquareIcon } from '../icons/terminal-square.js';
// SquareUser aliases
export { default as SquareUserIcon } from '../icons/square-user.js';
export { default as UserSquareIcon } from '../icons/user-square.js';
// SquareX aliases
export { default as SquareXIcon } from '../icons/square-x.js';
export { default as XSquareIcon } from '../icons/x-square.js';
// TestTubeDiagonal aliases
export { default as TestTubeDiagonalIcon } from '../icons/test-tube-diagonal.js';
export { default as TestTube2Icon } from '../icons/test-tube-2.js';
// TextSelect aliases
export { default as TextSelectIcon } from '../icons/text-select.js';
export { default as TextSelectionIcon } from '../icons/text-selection.js';
// TramFront aliases
export { default as TramFrontIcon } from '../icons/tram-front.js';
export { default as TrainIcon } from '../icons/train.js';
// TreePalm aliases
export { default as TreePalmIcon } from '../icons/tree-palm.js';
export { default as PalmtreeIcon } from '../icons/palmtree.js';
// TriangleAlert aliases
export { default as TriangleAlertIcon } from '../icons/triangle-alert.js';
export { default as AlertTriangleIcon } from '../icons/alert-triangle.js';
// TvMinimal aliases
export { default as TvMinimalIcon } from '../icons/tv-minimal.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link TvMinimal} instead. This alias will be removed in v1.0 */
default as Tv2Icon } from '../icons/tv-2.js';
// University aliases
export { default as UniversityIcon } from '../icons/university.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link University} instead. This alias will be removed in v1.0 */
default as School2Icon } from '../icons/school-2.js';
// UserRoundCheck aliases
export { default as UserRoundCheckIcon } from '../icons/user-round-check.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundCheck} instead. This alias will be removed in v1.0 */
default as UserCheck2Icon } from '../icons/user-check-2.js';
// UserRoundCog aliases
export { default as UserRoundCogIcon } from '../icons/user-round-cog.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundCog} instead. This alias will be removed in v1.0 */
default as UserCog2Icon } from '../icons/user-cog-2.js';
// UserRoundMinus aliases
export { default as UserRoundMinusIcon } from '../icons/user-round-minus.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundMinus} instead. This alias will be removed in v1.0 */
default as UserMinus2Icon } from '../icons/user-minus-2.js';
// UserRoundX aliases
export { default as UserRoundXIcon } from '../icons/user-round-x.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundX} instead. This alias will be removed in v1.0 */
default as UserX2Icon } from '../icons/user-x-2.js';
// UserRoundPlus aliases
export { default as UserRoundPlusIcon } from '../icons/user-round-plus.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRoundPlus} instead. This alias will be removed in v1.0 */
default as UserPlus2Icon } from '../icons/user-plus-2.js';
// UserRound aliases
export { default as UserRoundIcon } from '../icons/user-round.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UserRound} instead. This alias will be removed in v1.0 */
default as User2Icon } from '../icons/user-2.js';
// UsersRound aliases
export { default as UsersRoundIcon } from '../icons/users-round.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link UsersRound} instead. This alias will be removed in v1.0 */
default as Users2Icon } from '../icons/users-2.js';
// Utensils aliases
export { default as UtensilsIcon } from '../icons/utensils.js';
export { default as ForkKnifeIcon } from '../icons/fork-knife.js';
// UtensilsCrossed aliases
export { default as UtensilsCrossedIcon } from '../icons/utensils-crossed.js';
export { default as ForkKnifeCrossedIcon } from '../icons/fork-knife-crossed.js';
// WalletMinimal aliases
export { default as WalletMinimalIcon } from '../icons/wallet-minimal.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link WalletMinimal} instead. This alias will be removed in v1.0 */
default as Wallet2Icon } from '../icons/wallet-2.js';
// WandSparkles aliases
export { default as WandSparklesIcon } from '../icons/wand-sparkles.js';
export { 
/** @deprecated The name of this icon was changed because it didn't meet our guidelines anymore, use {@link WandSparkles} instead. This alias will be removed in v1.0 */
default as Wand2Icon } from '../icons/wand-2.js';
