// effect/dist/cjs/internal/schema/schemaId.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NonNaNSchemaId = exports.MinLengthSchemaId = exports.MinItemsSchemaId = exports.MaxLengthSchemaId = exports.MaxItemsSchemaId = exports.LessThanSchemaId = exports.LessThanOrEqualToSchemaId = exports.LessThanOrEqualToBigIntSchemaId = exports.LessThanBigIntSchemaId = exports.LengthSchemaId = exports.JsonNumberSchemaId = exports.ItemsCountSchemaId = exports.IntSchemaId = exports.GreaterThanSchemaId = exports.GreaterThanOrEqualToSchemaId = exports.GreaterThanOrEqualToBigIntSchemaId = exports.GreaterThanBigintSchemaId = exports.FiniteSchemaId = exports.DateFromSelfSchemaId = exports.BetweenSchemaId = exports.BetweenBigintSchemaId = void 0;
/** @internal */
const DateFromSelfSchemaId = exports.DateFromSelfSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/DateFromSelf");
/** @internal */
const GreaterThanSchemaId = exports.GreaterThanSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/GreaterThan");
/** @internal */
const GreaterThanOrEqualToSchemaId = exports.GreaterThanOrEqualToSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/GreaterThanOrEqualTo");
/** @internal */
const LessThanSchemaId = exports.LessThanSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/LessThan");
/** @internal */
const LessThanOrEqualToSchemaId = exports.LessThanOrEqualToSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/LessThanOrEqualTo");
/** @internal */
const IntSchemaId = exports.IntSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/Int");
/** @internal */
const NonNaNSchemaId = exports.NonNaNSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/NonNaN");
/** @internal */
const FiniteSchemaId = exports.FiniteSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/Finite");
/** @internal */
const JsonNumberSchemaId = exports.JsonNumberSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/JsonNumber");
/** @internal */
const BetweenSchemaId = exports.BetweenSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/Between");
/** @internal */
const GreaterThanBigintSchemaId = exports.GreaterThanBigintSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/GreaterThanBigint");
/** @internal */
const GreaterThanOrEqualToBigIntSchemaId = exports.GreaterThanOrEqualToBigIntSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/GreaterThanOrEqualToBigint");
/** @internal */
const LessThanBigIntSchemaId = exports.LessThanBigIntSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/LessThanBigint");
/** @internal */
const LessThanOrEqualToBigIntSchemaId = exports.LessThanOrEqualToBigIntSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/LessThanOrEqualToBigint");
/** @internal */
const BetweenBigintSchemaId = exports.BetweenBigintSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/BetweenBigint");
/** @internal */
const MinLengthSchemaId = exports.MinLengthSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/MinLength");
/** @internal */
const MaxLengthSchemaId = exports.MaxLengthSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/MaxLength");
/** @internal */
const LengthSchemaId = exports.LengthSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/Length");
/** @internal */
const MinItemsSchemaId = exports.MinItemsSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/MinItems");
/** @internal */
const MaxItemsSchemaId = exports.MaxItemsSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/MaxItems");
/** @internal */
const ItemsCountSchemaId = exports.ItemsCountSchemaId = /*#__PURE__*/Symbol.for("effect/SchemaId/ItemsCount");
